#!/bin/bash
#PBS -P pq84
#PBS -q normal 
#PBS -N Variant_calling_bcftools
#PBS -j oe
#PBS -m ae
#PBS -l walltime=48:00:00,mem=128GB,ncpus=16
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"
echo "Chromosomes 1 to 4"

echo "---------------------------------------"
echo "Define paths and load modules"
module load bcftools/1.12

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
INDEXTDIR="/g/data/pq84/malaria/data/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"
OUTDIR="/g/data/pq84/malaria/Pk_clusters/outputs/04_Variant_calling/bcftools"
INDIR="/g/data/pq84/malaria/Pk_clusters/outputs/03_Bam-pre/final_bam"

echo "---------------------------------------------------------------------------------------------------------------------"
echo "create input bam file list"
echo "---------------------------------------------------------------------------------------------------------------------"
cd $INDIR

ls *bam | cut -d ' ' -f 10 > input_bam_files.list

echo "---------------------------------------------------------------------------------------------------------------------"
echo "bcftools"
echo "---------------------------------------------------------------------------------------------------------------------"

bcftools mpileup --threads 16 -f $INDEXTDIR -b $INDIR/input_bam_files.list \
   -r ordered_PKNH_01_v2,ordered_PKNH_02_v2,ordered_PKNH_03_v2,ordered_PKNH_04_v2 \
   | bcftools call --threads 16 -m -Oz -a FORMAT/GQ,FORMAT/GP,INFO/PV4 -v -o $OUTDIR/bcftools_genotyped_1.vcf.gz

bcftools index --threads 16 -t -o $OUTDIR/bcftools_genotyped.vcf.gz.tbi $OUTDIR/bcftools_genotyped_1.vcf.gz

echo "---------------------------------------"
echo "Finished "
