#!/bin/bash
#PBS -P pq84
#PBS -q normal 
#PBS -N Merge_VCFs_from_bcftools
#PBS -j oe
#PBS -m ae
#PBS -l walltime=8:00:00,mem=80GB,ncpus=10 
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and load modules"
module load java/jdk-8.40 
PICARD="/g/data/pq84/bin/picard/build/libs/picard.jar"
GATK="/g/data/pq84/bin/GenomeAnalysisTK-3.8-1-0/GenomeAnalysisTK.jar"

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
cd /g/data/pq84/malaria/Pk_clusters/outputs/04_Variant_calling/bcftools

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Use GATK to merge VCF files containing the SAME samples BUT DIFFERENT contigs'
java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $PICARD \
    MergeVcfs \
    I=bcftools_genotyped_1.vcf.gz \
    I=bcftools_genotyped_2.vcf.gz \
    I=bcftools_genotyped_3.1.vcf.gz \
    I=bcftools_genotyped_3.2.vcf.gz \
    I=bcftools_genotyped_4.vcf.gz \
    O=bcftools_genotyped.vcf.gz

echo "---------------------------------------"
echo "Finished "