#!/bin/bash
#PBS -P pq84
#PBS -q normal 
#PBS -N Haplotype_call
#PBS -j oe
#PBS -m ae
#PBS -l walltime=12:00:00,mem=20GB,ncpus=10
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and load modules"
module load java/jdk-8.40 

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
OUTDIR="/g/data/pq84/malaria/Pk_clusters/outputs/04_Variant_calling/haplotype_call"
INDIR="/g/data/pq84/malaria/Pk_clusters/outputs/03_Bam-pre/final_bam"
PICARD="/g/data/pq84/bin/picard/build/libs/picard.jar"
GATK="/g/data/pq84/bin/GenomeAnalysisTK-3.8-1-0/GenomeAnalysisTK.jar"
INDEXTDIR="/g/data/pq84/malaria/data/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"
INTERVALS="/g/data/pq84/malaria/data/ref_genomes/PKA1H1/strain_A1_H.1.Icor.fasta.bed"

echo "---------------------------------------------------------------------------------------------------------------------"
echo "GATK"
echo "---------------------------------------------------------------------------------------------------------------------"
java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
    -T HaplotypeCaller \
    -ERC GVCF \
    --minPruning 3 \
    --maxNumHaplotypesInPopulation 200 \
    --max_alternate_alleles 3 \
    --variant_index_type LINEAR \
    --variant_index_parameter 128000 \
    -contamination 0.0 \
    -G Standard \
    -R $INDEXTDIR \
    -I $INDIR/SAMPLE.dupmarked.realigned.recal.bam \
    -o $OUTDIR/SAMPLE.dupmarked.realigned.recal.g.vcf.gz

echo "---------------------------------------------------------------------------------------------------------------------"
echo "FINISHED"
echo "---------------------------------------------------------------------------------------------------------------------"