---
title: "Exploring_popgen_metadata"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/jacobwestaway/Desktop/Menzies/Knowlesi/Pipeline/Pk_Pipeline/')
library(tidyverse)
library(janitor)
library(gridExtra)
library(pheatmap)
library(graph4lg)
library(maps)
library(sf)
library(rnaturalearth)
library(forcats)
```

# Combine cluster data with metadata, using the ZB indexes to get sample names

```{r,warning=F,message=F,eval=F}
popgen_metadata <- read_csv("data/metadata/cluster_metadata_popgen.csv") %>% 
  filter(grepl("PK", Sample)) %>% 
  mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
  left_join(
    readxl::read_xlsx("data/metadata/PK_Sabah_Sample_naming_indexes.xlsx") %>% 
      mutate(subjectid = as.character(subjectid)) %>%
      mutate(subjectid = ifelse(str_length(subjectid) == 1, paste0("00", subjectid), # if subjectid length = 1 paste 00
                                ifelse(str_length(subjectid) == 2, paste0("0", subjectid), # if not, then if subjectid length = 2 paste 0
                                       .$subjectid))) %>% # if not, keep value the same
      mutate(studycode = paste0(group, subjectid)) %>% # create studycode column that aligns with metadata
      left_join(
        readxl::read_xlsx("data/metadata/Epi_data_for_ZB_inventory_130422.xlsx", sheet = 2) %>% 
          select(studycode, date, pcfinal, sampletype_n, district, study_) %>% 
          rbind(
            readxl::read_xlsx("data/metadata/Epi_data_for_ZB_inventory_130422.xlsx", sheet = 4) %>% 
              select(studycode, datenrol, pcfinal, sampletype_n, district, study_) %>% 
              rename("date" = "datenrol")
            ) 
        ) %>% 
      rename(Sample = sampleid)
    ) 

```

# Exporting our metadata for data uplaod
```{r}
 readxl::read_xlsx("data/metadata/PK_Sabah_Sample_naming_indexes.xlsx") %>% 
      mutate(subjectid = as.character(subjectid)) %>%
      mutate(subjectid = ifelse(str_length(subjectid) == 1, paste0("00", subjectid), # if subjectid length = 1 paste 00
                                ifelse(str_length(subjectid) == 2, paste0("0", subjectid), # if not, then if subjectid length = 2 paste 0
                                       .$subjectid))) %>%
  mutate(studycode = paste0(group, subjectid)) %>%
  left_join(
        readxl::read_xlsx("data/metadata/Epi_data_for_ZB_inventory_130422.xlsx", sheet = 2) %>% 
          select(studycode, date, pcfinal, sampletype_n, district, study_) %>% 
          rbind(
            readxl::read_xlsx("data/metadata/Epi_data_for_ZB_inventory_130422.xlsx", sheet = 4) %>% 
              select(studycode, datenrol, pcfinal, sampletype_n, district, study_) %>% 
              rename("date" = "datenrol"))
            ) %>%
  arrange(sampleid) #%>% write_csv("tmp_ZB_data.csv")
```

# Distribution of severe cases across clusters

 - use summarise to get proportions of severe and non-sever for each cluster and then plot is col
 
```{r,warning=F,message=F,eval=F}
popgen_metadata %>% 
  mutate_if(is.character, as.factor) %>% 
  group_by(Cluster) %>% 
  summarise(Prop_severe = sum(severe == "Y")/n(), Sample_count = n())
```


# Distribution of severe cases across districts

```{r,warning=F,message=F,eval=F}
popgen_metadata %>% 
  mutate_if(is.character, as.factor) %>% 
  group_by(district) %>% 
  summarise(Prop_severe = sum(severe == "Y")/n(), Sample_count = n())
```


# Differences in parasitemia between clusters or districts?

```{r,warning=F,message=F,eval=F}
popgen_metadata %>% 
  mutate_if(is.character, as.factor) %>% 
  ggplot(aes(x = Cluster, y = parasitemia)) +
  geom_boxplot() +
  geom_point(aes(colour = district)) +
  scale_y_log10()
```

# Do mixed infections come from a specific district?


# Bar plot of ADMIXTURE output

```{r,warning=F,message=F,eval=F}
admixture_table <- read.table("data/ADMIXTURE/cleaned.3.Q") %>% 
  cbind(
    read_csv("data/metadata/cluster_metadata_popgen.csv") 
  )

admixture_table %>% 
  pivot_longer(1:3, names_to = "Ancestry", values_to = "Proportion") %>% 
  mutate(Cluster = ifelse(Ancestry == "V1", "Peninsular",
                          ifelse(Ancestry == "V2", "Mn", "Mf"))) %>% 
  mutate(Ancestry = Proportion) %>% 
  ggplot(aes(x = Sample, y = Ancestry, fill = Cluster)) +
  geom_col() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  scale_fill_manual(values = c("#440154FF", "#39568CFF", "#73D055FF"))
```

# Digging into mixed samples

```{r,warning=F,message=F,eval=F}
admixture_table %>% 
  filter(V1 > 0.005 & V2 > 0.005 |
           V1 > 0.005 & V3 > 0.005 |
           V2 > 0.005 & V3 > 0.005) %>% 
  pivot_longer(1:3, names_to = "Ancestry", values_to = "Proportion") %>% 
  mutate(Sample = str_remove(Sample, "_DKD.*")) %>%
  mutate(Cluster = ifelse(Ancestry == "V1", "Peninsular",
                          ifelse(Ancestry == "V2", "Mn", "Mf"))) %>% 
  mutate(Ancestry = Proportion) %>% 
  ggplot(aes(x = Sample, y = Ancestry, fill = Cluster)) +
  geom_col() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_manual(values = c("#440154FF", "#39568CFF", "#73D055FF"))
```

# What districts are the mixed samples from?

```{r,warning=F,message=F,eval=F}
admixture_table %>% 
  filter(V1 > 0.005 & V2 > 0.005 |
           V1 > 0.005 & V3 > 0.005 |
           V2 > 0.005 & V3 > 0.005) %>% 
  mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
  select(Sample) %>% 
  left_join(
    popgen_metadata %>% 
      select(Sample, district) %>% 
      filter(grepl("PK", Sample)) %>% 
      rbind(
        read_csv("data/metadata/Pk_clusters_metadata.csv") %>% 
          select(Sample, area) %>% 
          rename(district = area)
      )
  )
```

# Map with distribution of samples and clusters

```{r,warning=F,message=F,eval=F}
map_data <- popgen_metadata %>% 
  dplyr::select(Sample, district, Cluster) %>% 
  rename(Location = district) %>% 
  rbind(
    read_csv("data/metadata/cluster_metadata_popgen.csv") %>% 
      filter(!grepl("PK",Sample)) %>% 
      dplyr::select(Sample, Location, Cluster)
  ) %>% 
  mutate(Location = ifelse(Location == "Sungai_Siuit", "Sungai_Siput", .$Location))

#map_data %>%  # used this to get long an lat
 # select(Location) %>% 
  #unique() %>% 
#write_csv("data/metadata/map_data.csv")
  
map_data <- map_data %>% 
  left_join(
    read_csv("data/metadata/map_data.csv")
  ) %>% 
  na.omit() %>% 
  arrange(Location) %>% 
  left_join(
    admixture_table %>% 
      select(1:4) %>% 
      mutate(Mixed = ifelse(V1 > 0.005 & V2 > 0.005 | V1 > 0.005 & V3 > 0.005 | V2 > 0.005 & V3 > 0.005, "Yes", "No")) %>% 
      mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
      select(Sample, Mixed)
  ) %>% 
  na.omit()

# world <- ne_countries(scale = "medium", returnclass = "sf", country = c("Malaysia", "Indonesia"))
world <- ne_countries(scale = "medium", returnclass = "sf", continent = "Asia")

sf_use_s2(FALSE) # Spherical geometry (s2) switched off

ggplot() +
  geom_sf(data = world, col="transparent") +
  scale_x_continuous(limits = c(95, 120)) +
  scale_y_continuous(limits = c(-5, 10)) +
  coord_sf(expand = FALSE) +
  xlab("Longitude") + ylab("Latitude") +
  theme_linedraw() +
  theme_dark() +
  theme(plot.title = element_text(size=15, hjust = 0.5), plot.margin = unit(c(1,1,1,1), "cm")) +
  geom_jitter(data = map_data, aes(x = Long, y = Lat, colour = Cluster, shape = Mixed), size = 3, width = 0.4, height = 0.4) +
  scale_shape_manual(values = c(3, 17)) +
  scale_colour_manual(values = c("#440154FF", "#73D055FF", "#39568CFF"))

#ggsave("figure/Admix_map.png", map, dpi = 300)
```

```{r}
ggplot() +
  geom_sf(data = world, col = "transparent", fill = "white") +
  scale_x_continuous(limits = c(107, 120)) +
  scale_y_continuous(limits = c(-4.5, 7)) +
  coord_sf(expand = FALSE) +
  xlab("Longitude") + ylab("Latitude") +
  theme_linedraw() +
  theme_dark() +
  theme(plot.title = element_text(size=15, hjust = 0.5), plot.margin = unit(c(1,1,1,1), "cm")) 

#ggsave("Map_borneo.png", plot, dpi = 300)

malaysia <- ne_states(country = c("malaysia", "indonesia"), returnclass = "sf") %>% 
  mutate(name = ifelse(name != "Sabah" & name != "Sarawak", "Other", name))

ggplot() +
  geom_sf(data = malaysia, col = "transparent", mapping = aes(fill = name)) +
  scale_x_continuous(limits = c(107, 120)) +
  scale_y_continuous(limits = c(-4.5, 7)) +
  coord_sf(expand = FALSE) +
  xlab("Longitude") + ylab("Latitude") +
  theme_linedraw() +
  theme_dark() +
  theme(plot.title = element_text(size=15, hjust = 0.5), plot.margin = unit(c(1,1,1,1), "cm")) +
  scale_fill_manual(values = c(Sabah = "#440154FF", Sarawak = "#FDE725FF", Other = "#33638DFF"), name = "State")
```

# Locations of different Sabah samples

**Do the clusters come from specific districts?**

```{r,warning=F,message=F}
Sabah_map_data <- map_data %>% 
  filter(Long > 110 & Lat > 3.5)

ggplot() +
  geom_sf(data = world, col="transparent") +
  scale_x_continuous(limits = c(115, 120)) + 
  scale_y_continuous(limits = c(4.5, 7)) +
  coord_sf(expand = FALSE) +
  xlab("Longitude") + ylab("Latitude") +
  theme_linedraw() +
  theme_dark() +
  theme(plot.title = element_text(size = 15, hjust = 0.5), plot.margin = unit(c(1,1,1,1), "cm")) +
  geom_jitter(data = Sabah_map_data, aes(x = Long, y = Lat, colour = Cluster, shape = Mixed), size = 4, width = 0.05, height = 0.05) +
  scale_shape_manual(values = c(3, 17)) +
  scale_colour_manual(values = c("#440154FF", "#39568CFF"))
```


```{r,warning=F,message=F}
map_data %>% 
  filter(Long > 110 & Lat > 3.5) %>% 
  group_by(Location) %>% 
  dplyr::summarise(Mf = sum(Cluster == "Mf"),
                   Mn = sum(Cluster == "Mn")) %>% 
  arrange(desc(Mf))
```



# Combining information on genetic and geographic clusters

```{r,warning=F,message=F,eval=F}
admixture_table <- admixture_table %>% 
  arrange(desc(V3), desc(V2), desc(V1)) %>% 
  add_column(Sample2 = as.factor(1:nrow(.)))
 
mixed_plot <- admixture_table %>% 
  mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
  left_join(
    map_data %>% 
      select(Sample, Long, Lat)
  ) %>% 
  mutate(Geo_cluster = ifelse(Long < 105, "Peninsular", 
                               ifelse(Long > 105 & Long < 112.2, "Sarawak", 
                                      ifelse(Long > 112.2 & Long < 115, "Sarawak", "Sabah")))) %>% 
  mutate(Geo_cluster = ifelse(is.na(Geo_cluster), "Peninsular", .$Geo_cluster)) %>% 
  pivot_longer(1:3, names_to = "Ancestry", values_to = "Proportion") %>% 
  mutate(Cluster = ifelse(Ancestry == "V1", "Peninsular",
                          ifelse(Ancestry == "V2", "Mn", "Mf"))) %>% 
  mutate(Ancestry = Proportion) %>% 
  arrange(Geo_cluster, desc(Ancestry)) %>% 
  ggplot(aes(x = Sample2, y = Ancestry, fill = Cluster)) +
    geom_col() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    scale_fill_manual(values = c("#440154FF", "#73D055FF", "#39568CFF")) +
    facet_grid(~Geo_cluster, scales = "free", space = "free") +
    xlab("Sample")

ggsave("mixed_plot.png", width = 12, height = 4, dpi = 600, mixed_plot)
```

**Why does Kapit not have mixed samples?**

# Heatmap of IBD information for clusters

## Discordance

```{r}
heatmap_annotation <- admixture_table %>% 
  mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
  left_join(
    map_data %>% 
      select(Sample, Long, Lat)
  ) %>% 
  mutate(Geo_cluster = ifelse(Long < 105, "Peninsular", 
                               ifelse(Long > 105 & Long < 112.2, "Betong", 
                                      ifelse(Long > 112.2 & Long < 115, "Kapit", "Sabah")))) %>% 
  mutate(Geo_cluster = ifelse(is.na(Geo_cluster), "Peninsular", .$Geo_cluster)) %>% 
  select(Sample, Cluster, Geo_cluster) %>% 
  remove_rownames() %>% 
  column_to_rownames("Sample")
 
read.table("data/hmmIBD/Pk.hmm_fract.txt") %>%
  select(1,2,4) %>% 
  row_to_names(1)  %>% 
  mutate_if(is.factor, as.character) %>% 
  mutate_at(c("sample1", "sample2"), ~str_remove(., "_DK.*")) %>% 
  mutate(discordance = as.numeric(discordance)) %>% 
  df_to_pw_mat(from = "sample1", to = "sample2", value = "discordance") %>%
  pheatmap(show_rownames = F, show_colnames = F, 
           annotation_row = heatmap_annotation, annotation_col = heatmap_annotation)
```


## Fraction of variant sites called IBD

```{r}
read.table("data/hmmIBD/Pk.hmm_fract.txt") %>% 
  select(1,2,10) %>% 
  row_to_names(1)  %>% 
  mutate_if(is.factor, as.character) %>% 
  mutate_at(c("sample1", "sample2"), ~str_remove(., "_DK.*")) %>% 
  mutate(fract_sites_IBD = as.numeric(fract_sites_IBD)) %>% 
  df_to_pw_mat(from = "sample1", to = "sample2", value = "fract_sites_IBD") %>%
  pheatmap(show_rownames = F, show_colnames = F, 
           annotation_row = heatmap_annotation, annotation_col = heatmap_annotation)
```



# IBD sites in sample 32 compared to Peninsular samples 

```{r}
read.table("data/hmmIBD/PK_SB_DNA_032.hmm.txt") %>% 
  mutate(different = as.numeric(different)) %>%
  left_join(
    admixture_table %>% 
      rename(sample1 = Sample) %>% 
      select(4:6)
      ) %>%
  mutate(sample1 = paste(sample1, 1:n(), sep = "_")) %>% 
  pivot_longer(cols = start:end, names_to = "pos", values_to = "pos_value") %>% 
  filter(Cluster == "Peninsular") %>% 
  arrange(sample1) %>% 
  filter(chr == 3 | chr == 9) %>% 
  mutate(chr = paste0("Chromosome ", chr)) %>% 
  mutate(IBD = as.factor(different)) %>% 
  ggplot(aes(x = pos_value, y = IBD, colour = IBD, group = sample1)) +
  geom_point() +
  geom_line() +
  theme(legend.position = "none") +
  facet_grid(~chr, scales = "free", space = "free") +
  scale_colour_manual(values = c( "#73D055FF", "#440154FF")) +
  xlab("Position in chromosome")
```

# Table with IBD sample info

```{r}
read.table("data/hmmIBD/PK_SB_DNA_032.hmm.txt") %>% 
  mutate(different = as.numeric(different)) %>%
  left_join(
    admixture_table %>% 
      rename(sample1 = Sample) %>% 
      select(4:6)
      ) %>% 
  mutate(sample1 = paste(sample1, 1:n(), sep = "_")) %>% 
  filter(different == 0 & Cluster == "Peninsular") %>% 
  filter(chr == 3 | chr == 9)
```


# IBD sites in sample 32 compared to other Mf sample (make sure its not an error)


```{r}
read.table("data/hmmIBD/PK_SB_DNA_032.hmm.txt") %>% 
  select(1:6) %>% 
  mutate(different = as.numeric(different)) %>%
  left_join(
    admixture_table %>% 
      rename(sample1 = Sample) %>% 
      select(4:6) 
      ) %>%
  mutate(sample1 = paste(sample1, 1:n(), sep = "_")) %>% 
  pivot_longer(cols = start:end, names_to = "pos", values_to = "pos_value") %>% 
  filter(Cluster == "Mf") %>% 
  arrange(sample1) %>% 
  filter(chr == 3 | chr == 9) %>% 
  mutate(chr = paste0("Chromosome ", chr)) %>% 
  mutate(IBD = as.factor(different)) %>% 
  ggplot(aes(x = pos_value, y = IBD, colour = IBD, group = sample1)) +
  geom_point() +
  geom_line() +
  theme(legend.position = "none") +
  facet_grid(~chr, scales = "free", space = "free") +
  scale_colour_manual(values = c( "#73D055FF", "#440154FF")) +
  xlab("Position in chromosome")
```


# IBD sites in sample 32 compared to other Mn


```{r}
read.table("data/hmmIBD/PK_SB_DNA_032.hmm.txt") %>% 
  select(1:6) %>% 
  mutate(different = as.numeric(different)) %>%
  left_join(
    admixture_table %>% 
      rename(sample1 = Sample) %>% 
      select(4:6) 
      ) %>%
  mutate(sample1 = paste(sample1, 1:n(), sep = "_")) %>% 
  pivot_longer(cols = start:end, names_to = "pos", values_to = "pos_value") %>% 
  filter(Cluster == "Mn") %>% 
  arrange(sample1) %>% 
  filter(chr == 3 | chr == 9) %>% 
  mutate(chr = paste0("Chromosome ", chr)) %>% 
  mutate(IBD = as.factor(different)) %>% 
  ggplot(aes(x = pos_value, y = IBD, colour = IBD, group = sample1)) +
  geom_point() +
  geom_line() +
  theme(legend.position = "none") +
  facet_grid(~chr, scales = "free", space = "free") +
  scale_colour_manual(values = c( "#73D055FF", "#440154FF")) +
  xlab("Position in chromosome")
```

**ADMIXTURE shows no Mn in samples 32...**




# Heatmap of 'genotype file' checking for possible IBD

**NB** The input file is the file created for input into hmmIBD.
	
```{r}
IBD_heatmap_data <- read_tsv("data/hmmIBD/hmmIBD_temp.tsv") %>% 
  pivot_longer(3:ncol(.), names_to = "Sample", values_to = "Genotype") %>% 
  left_join(
    admixture_table %>% 
      select(4:6)
  ) %>% 
  filter(grepl("PK_SB_DNA_032", Sample) | Cluster == "Peninsular") 


IBD_heatmap_annotation <- admixture_table %>% 
  select(Sample, Cluster) %>% 
  column_to_rownames("Sample")

IBD_heatmap_data %>% 
  filter(POS > 657462 & POS < 668818) %>% 
  select(2:4) %>% 
  pivot_wider(names_from = "POS", values_from = "Genotype") %>% 
  column_to_rownames("Sample") %>% 
  pheatmap(show_rownames = T, show_colnames = F, annotation_row = IBD_heatmap_annotation, cluster_cols = F)
```



```{r}
IBD_heatmap_data <- read_tsv("data/hmmIBD/hmmIBD_temp2.tsv") %>% 
  pivot_longer(3:ncol(.), names_to = "Sample", values_to = "Genotype") %>% 
  left_join(
    admixture_table %>% 
      select(4:6)
  ) %>% 
  filter(grepl("PK_SB_DNA_032", Sample) | grepl("ERR3374033", Sample) | grepl("ERR3374037", Sample)
         | grepl("ERR3374040", Sample) | grepl("ERR3374051", Sample) | grepl("ERR3374057", Sample)) 

IBD_heatmap_data %>% 
  filter(POS > 1848114 & POS < 1850009) %>% 
  select(2:4) %>% 
  pivot_wider(names_from = "POS", values_from = "Genotype") %>% 
  column_to_rownames("Sample") %>% 
  pheatmap(show_rownames = T, show_colnames = F, annotation_row = IBD_heatmap_annotation, cluster_cols = F)
```



# Introgression with Mf and Mn Samples

## What samples are mixed?
```{r}
admixture_table %>% 
  filter(V1 > 0.005 & V2 > 0.005 |
           V1 > 0.005 & V3 > 0.005 |
           V2 > 0.005 & V3 > 0.005) %>% 
  mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
  select(Sample, Cluster) 
```


## On HPC - filter hmmIBD file for these samples

```{r}
read.table("Pk.hmm.txt", header =T) %>% 
  filter(sample2 == "ERR274221" | 
           sample2 == "ERR274222" |
           sample2 == "ERR985375" |
           sample2 == "ERR985378" |
           sample2 == "ERR985381" |
           sample2 == "ERR985384" |
           sample2 == "PK_SB_DNA_028_DKDL210002157-1a_HWHGKDSXY_L4" |
           sample2 == "PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4" |
           sample2 == "PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4" |
           sample2 == "PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4" ) %>% 
    write.table("Mixed_sample_hmm.txt", col.names = T, row.names = F, quote = F)

```




# IBD sites in mixed samples comapred to other samples


## Mn-majority mixed samples
```{r}
read.table("data/hmmIBD/Mixed_sample_hmm.txt", header = T) %>% 
  mutate(different = as.numeric(different)) %>%
  left_join(
    admixture_table %>% 
      rename(sample1 = Sample) %>% 
      dplyr::select(4:6)
      ) %>% 
  mutate(sample1 = paste(sample1, 1:n(), sep = "_")) %>% 
  pivot_longer(cols = start:end, names_to = "pos", values_to = "pos_value") %>% 
  filter(Cluster == "Mf" |            
           sample2 == "PK_SB_DNA_028_DKDL210002157-1a_HWHGKDSXY_L4" |
           sample2 == "PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4" |
           sample2 == "PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4" |
           sample2 == "PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4") %>% 
  arrange(sample1) %>% 
  filter(chr == 5) %>% 
  mutate(chr = paste0("Chromosome ", chr)) %>% 
  mutate(IBD = as.factor(different)) %>% 
  ggplot(aes(x = pos_value, y = IBD, colour = IBD, group = sample1)) +
  geom_point() +
  geom_line() +
  theme(legend.position = "none") +
  facet_grid(~chr, scales = "free", space = "free") +
  scale_colour_manual(values = c( "#73D055FF", "#440154FF")) +
  xlab("Position in chromosome") 
```










# Exploring samples with high MOI

```{r}
popgen_metadata %>% 
  filter(Sample == "PK_SB_DNA_006" | Sample == "PK_SB_DNA_032" | Sample == "PK_SB_DNA_039" | Sample == "PK_SB_DNA_048" |
         Sample == "PK_SB_DNA_051" | Sample == "PK_SB_DNA_055" | Sample == "PK_SB_DNA_083") %>% 
  select(Sample, parasitemia, date)
```

ERR2214850
ERR2214856
ERR3374041
ERR985386
ERR985395
ERR985396
ERR985397
ERR985405
ERR985410
ERR985417
PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4
PK_SB_DNA_032_DKDL210002161-1a_HWHGKDSXY_L4
PK_SB_DNA_039_DKDL210002168-1a_HWHGKDSXY_L4
PK_SB_DNA_048_DKDL210002177-1a_HWHGKDSXY_L4
PK_SB_DNA_051_DKDL210002180-1a_HWHGKDSXY_L4
PK_SB_DNA_055_DKDL210002184-1a_HWHGKDSXY_L4
PK_SB_DNA_083_DKDL210002212-1a_HWHGKDSXY_L4

```{r}
popgen_metadata %>% 
  select(Sample, district, Cluster) %>% 
  rename(Location = district) %>% 
  rbind(
    read_csv("data/metadata/cluster_metadata_popgen.csv") %>% 
      filter(!grepl("PK",Sample)) %>% 
      select(Sample, Location, Cluster)
  ) %>% 
  mutate(Location = ifelse(Location == "Sungai_Siuit", "Sungai_Siput", .$Location)) %>% 
  filter(Sample == "PK_SB_DNA_006" | Sample == "PK_SB_DNA_032" | Sample == "PK_SB_DNA_039" | Sample == "PK_SB_DNA_048" |
           Sample == "PK_SB_DNA_051" | Sample == "PK_SB_DNA_055" | Sample == "PK_SB_DNA_083" | Sample == "ERR2214850" | Sample == "ERR2214856" |
           Sample == "ERR3374041" | Sample == "ERR985386" | Sample == "ERR985395" | Sample == "ERR985396" | Sample == "ERR985397" | Sample == "ERR985405" |
           Sample == "ERR985410" | Sample == "ERR985417") %>% # the non-Sabah samples aren't in there because they didn't pass the MAF/geno filters
  select(Sample, Location, Cluster)
```

Other samples:
 - 8 from Kapit
 - 1 from Betong
 - 1 from Kuala Lipis 
 
ERR2214850 - Kapit - Mn
ERR2214856 - Kapit - Mn
ERR3374041 - Kuala Lipis - Mf
ERR985386 - Kapit - Mf
ERR985395 - Kapit - Mf
ERR985396 - Kapit - Mf
ERR985397 - Kapit - Mf
ERR985405 - Kapit - Mf
ERR985410 - Betong - Mn
ERR985417 - Kapit - Mn
 
cat /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_metadata.csv | grep 'ERR2214850\|ERR2214856\|ERR3374041\|ERR985386\|ERR985395\|ERR985396\|ERR985397\|ERR985405\|ERR985410\|ERR985417'
 


```{r}
world <- ne_countries(scale = "medium", returnclass = "sf", continent = "Asia") %>% 
  mutate(Samples = ifelse(name == "Malaysia", "Yes", 
                          ifelse(name == "Cambodia" | name == "Indonesia" | name == "Thailand" | name == "Myanmar" | name == "Vietnam", "No", "Limited cases"))) %>% 
  mutate(Samples = fct_relevel(Samples, c("Yes", "No", "Limited cases")))
  

map_plot <- ggplot() +
  geom_sf(data = world, col="transparent", mapping = aes(fill = Samples)) +
  scale_x_continuous(limits = c(90, 130), breaks = c(90, 100, 110, 120, 130)) +
  scale_y_continuous(limits = c(-15, 30)) +
  coord_sf(expand = FALSE) +
  xlab("Longitude") + ylab("Latitude") +
  theme_linedraw() +
  theme_dark() +
  theme(plot.title = element_text(size=15, hjust = 0.5), plot.margin = unit(c(1,1,1,1), "cm")) +
  scale_fill_manual(values = c("Limited cases" = "#440154FF", 
                               Yes = "#FDE725FF", 
                               No = "#33638DFF"), 
                    name = "Pk Samples")

ggsave("map_ACREME.png", map_plot)
```


```{r}

#ggsave("Map_borneo.png", plot, dpi = 300)

malaysia <- ne_states(country = c("malaysia", "indonesia"), returnclass = "sf") %>% 
  mutate(name = ifelse(name != "Sabah" & name != "Sarawak", "Other", name))

ggplot() +
  geom_sf(data = malaysia, col = "transparent", mapping = aes(fill = name)) +
  scale_x_continuous(limits = c(107, 120)) +
  scale_y_continuous(limits = c(-4.5, 7)) +
  coord_sf(expand = FALSE) +
  xlab("Longitude") + ylab("Latitude") +
  theme_linedraw() +
  theme_dark() +
  theme(plot.title = element_text(size=15, hjust = 0.5), plot.margin = unit(c(1,1,1,1), "cm")) +
  scale_fill_manual(values = c(Sabah = "#440154FF", Sarawak = "#FDE725FF", Other = "#33638DFF"), name = "State")
```


# SNP Barcodes


```{r}
SNP_barcode_table <- read_tsv("data/MSC/Introgression_genotype_tmp.tsv") %>%
  mutate(SNP = ifelse(SNP < 0, NA, .$SNP))
  
introgressed_samples <- read_tsv("data/MSC/introgressed_samples.tsv")


sample_order <- SNP_barcode_table %>% 
  arrange(Cluster, SAMPLE) %>% 
  select(SAMPLE, Cluster) %>% 
  unique() %>% 
  add_column(SAMPLE_2 = 1:nrow(.)) %>% 
  mutate(SAMPLE_2 = as.factor(SAMPLE_2))

SNP_barcode_table  <- SNP_barcode_table %>% 
  left_join(sample_order)
    
SNP_barcode_table %>% 
  filter(CHROM == "08" & POS > 950000	& POS < 959999) %>% 
  ggplot(aes(x = POS, y = SAMPLE_2, fill = Cluster, alpha = SNP, group = Cluster)) +
  geom_col() +
  scale_fill_manual(values = c("#440154FF", "#73D055FF", "#39568CFF")) +
  theme(axis.text.y=element_blank(), axis.ticks.y=element_blank(),
        axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
  scale_alpha(guide = 'none') +
  ylab("Sample") +
  xlab("Chromosome Position") +
  geom_text(aes(label = ifelse(SAMPLE %in% introgressed_samples$SAMPLE, "***", '')), size = 2, hjust = 2)

#ggsave("oocyte_window_snp_barcode.png", dpi = 300, width = 12, snp_barcode_plot)


# Function
snp_barcode_Mf <- function(WINDOW, CHROM_NUM, START, END){
  SNP_barcode_table %>% 
    filter(CHROM == paste0(CHROM_NUM) & POS > START	& POS < END) %>% 
    ggplot(aes(x = POS, y = SAMPLE_2, fill = Cluster, alpha = SNP, group = Cluster)) +
    geom_col() +
    scale_fill_manual(values = c("#440154FF", "#73D055FF", "#39568CFF")) +
    theme(axis.text.y=element_blank(), axis.ticks.y=element_blank(),
          axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
    scale_alpha(guide = 'none') +
    ylab("Sample") +
    xlab("Chromosome Position") +
    geom_text(aes(label = ifelse(SAMPLE %in% introgressed_samples$SAMPLE, "***", '')), size = 2, hjust = 2)
}

snp_barcode_Mf(826, "08", 950000, 959999)
```

```{r}
# Clustered plot - 826
SNP_barcode_table <- read_tsv("data/MSC/Introgression_genotype_tmp.tsv") %>%
  mutate(SNP = ifelse(SNP < 0, NA, .$SNP))

clust_data <- SNP_barcode_table %>% 
  #filter(CHROM == "08" & POS > 950000	& POS < 959999) %>% 
  filter(CHROM == "08" & POS > 930000	& POS < 979999) %>%
  select(CHROM, POS, SAMPLE, SNP) %>% 
  pivot_wider(names_from = SAMPLE, values_from = SNP) 
  
clust <- clust_data %>%
  select(-c(1:2)) %>% 
  t() %>% # each row is a sample and each column a SNP positon 
  dist() %>%
  hclust() #euclidean distance

#################################### Check order is correct - it is not
ggdendrogram(clust, rotate = TRUE, size = 2)
p <- ggdendrogram(clust, rotate = FALSE, size = 2)

ggplotly(p) 
#################################### Change sample order like we do to cluster the "clusters"

clust$labels[c(clust$order)]

sample_order <- as.data.frame(clust$labels[c(clust$order)]) %>% 
  rename(SAMPLE = "clust$labels[c(clust$order)]") %>% 
  filter(SAMPLE != "CHROM" & SAMPLE != "POS") %>% 
  add_column(SAMPLE_2 = 1:nrow(.)) %>% 
  mutate(SAMPLE_2 = as.factor(SAMPLE_2))

SNP_barcode_table <- SNP_barcode_table %>% 
  left_join(sample_order)

####################################

SNP_barcode_table %>% 
  #filter(CHROM == "08" & POS > 950000	& POS < 959999) %>% 
  filter(CHROM == "08" & POS > 930000	& POS < 979999) %>% 
  ggplot(aes(x = POS, y = SAMPLE_2, fill = Cluster, alpha = SNP, group = Cluster)) +
  geom_col() +
  scale_fill_manual(values = c("#440154FF", "#73D055FF", "#39568CFF")) +
  theme(axis.text.y=element_blank(), axis.ticks.y=element_blank(),
       axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
  scale_alpha(guide = 'none') +
  ylab("Sample") +
  xlab("Chromosome Position") +
  geom_text(aes(label = ifelse(SAMPLE %in% introgressed_samples$SAMPLE, "***", '')), size = 2, hjust = 2)

#ggsave("oocyte_window_snp_barcode_clustered.png", dpi = 300, width = 12, height = 20, snp_barcode_plot)
```


read_tsv("data/MSC/SNP_barcode_table.tsv")

```{r}
clustered_snp_barcode <- function(WINDOW, CHROM_NUM, START, END){
  SNP_barcode_table <- SNP_barcode_table %>%
    filter(CHROM == paste0(CHROM_NUM) & POS > START	& POS < END) %>% 
    mutate(SNP = ifelse(SNP < 0, NA, .$SNP))

  clust_data <- SNP_barcode_table %>% 
    select(CHROM, POS, SAMPLE, SNP) %>% 
    pivot_wider(names_from = SAMPLE, values_from = SNP) 
  
  clust <- clust_data %>%
    select(-c(1:2)) %>% 
    t() %>% # each row is a sample and each column a SNP positon 
    dist() %>%
    hclust() #euclidean distance
 
  sample_order <- as.data.frame(clust$labels[c(clust$order)]) %>% 
    rename(SAMPLE = "clust$labels[c(clust$order)]") %>% 
    filter(SAMPLE != "CHROM" & SAMPLE != "POS") %>% 
    add_column(SAMPLE_2 = 1:nrow(.)) %>% 
    mutate(SAMPLE_2 = as.factor(SAMPLE_2))

  SNP_barcode_table <- SNP_barcode_table %>% 
    left_join(sample_order)
  
  #introgressed_samples <- introgressed_windows_meta_Mn_to_Mf  %>% 
    #filter(WINDOW == 826) 
    
  SNP_barcode_table %>% 
    ggplot(aes(x = POS, y = SAMPLE_2, fill = Cluster, alpha = SNP, group = Cluster)) +
      geom_col() +
      scale_fill_manual(values = c("#440154FF", "#73D055FF", "#39568CFF")) +
      theme(axis.text.y=element_blank(), axis.ticks.y=element_blank(),
        axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
      scale_alpha(guide = 'none') +
      ylab("Sample") +
      xlab("Chromosome Position") +
      geom_text(aes(label = ifelse(SAMPLE %in% introgressed_samples$SAMPLE, "***", '')), size = 2, hjust = 2) 
}
```


