# Using plink to for explotaroy analyses

This markdown describes the population genetics analyses, inlcuding pre-processing steps.

Helpful links:
PLINK: https://www.biostars.org/p/271694/
PLINK: https://plink.readthedocs.io/en/latest/plink_ex/
MOI: https://bahlolab.github.io/moimix/vignettes/introduction.html#estimating-moi-with-f_ws

## Required tools/libraries for Analyses

```{bash}
module load bcftools/1.12
module load java/jdk-8.40 
module load R/4.1.0 
export R_LIBS_USER="/g/data/pq84/R"
export PATH=$PATH:/g/data/pq84/bin/plink2/

FASTA="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/ref_genomes/PKA1H1/fasta"
VCF="/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/04_Variant_calling/filtered/"
GATK="/g/data/pq84/bin/GenomeAnalysisTK-3.8-1-0/GenomeAnalysisTK.jar"
REGIONS_TO_MASK="/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/regions_to_mask.list"

cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses
```

## Filter for variants that pass filters and select chromosomes 1 to 14

```{bash,eval=F}
java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
    -T SelectVariants \
    -R $FASTA/strain_A1_H.1.Icor.fasta \
    -V $VCF/FILTERED_SNPs.vcf  \
    --excludeFiltered \
    --excludeIntervals PKNH_MIT_v2 --excludeIntervals new_API_strain_A1_H.1 \
    -o PK_consensus_filtered_pass.vcf
```


## Annotate VCF for downstream analysis

```{bash,eval=F}
#!/bin/bash
#PBS -P pq84
#PBS -q normalbw
#PBS -N Annotate_variants
#PBS -j oe
#PBS -m ae
#PBS -l walltime=12:00:00,mem=80GB,ncpus=10
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and load modules"
export PATH=$PATH:/g/data/pq84/bin/ensembl-vep

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
GFF="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/ref_genomes/PKA1H1/gff/updated_version"
FASTA="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"

cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses

echo "---------------------------------------------------------------------------------------------------------------------"
echo "Annotate SNPs"
vep -i PK_consensus_filtered_pass.vcf --gff $GFF/PlasmoDB-55_PknowlesiA1H1.gff.gz --fasta $FASTA -o PK_consensus_filtered_annotation.txt --force_overwrite --everything

echo "---------------------------------------"
echo "FINISHED"
```


## Calculate multiplicity of infection using moimix

To install MCMCpack on Gadi: https://opus.nci.org.au/display/Help/R 

Steps:
 - Mask problematic regions (GATK)
 - Convert VCF to GDS
 - Estimate BAF matrix 
    - estimate B-allele freq for each isolate. 
    - resulting bafmatrix stores estimates and genomic coordinates
 - Estimating MOI with binommix
    - estimated on per-isolate basis using binomial mixture model
    - to fit model - need a matrix of read counts and to set the value of k=2c where c is the number of colnal infections
 - Estimating MOI with Fws
    - alternative way of assessing clonality is to estmate Fws, with Fws < 0.95 indicating clonal infection (low MOI)

**NB.** Clean up VCF.
Add annotation to the ID column of your VCF file and subset to only biallelic site (assumption of moimix).
bcftools view PK_consensus_no_MOI.vcf -M2 | bcftools annotate --threads 10 --set-id +'%CHROM\_%POS\_%REF\_%FIRST_ALT' - > moimix/PK_consensus_biallelic.vcf



```{bash,eval=F}
cd moimix

java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V PK_consensus_biallelic.vcf \
        -XL $REGIONS_TO_MASK \
        -o PK_consensus_masked.vcf.gz

cdd 1
```

Run below in R
```{R,eval=F}
# Converting VCF to GDS for estMOI
library(tidyverse)
library(SeqArray)
seqVCF2GDS("moimix/PK_consensus_masked.vcf.gz", "moimix/PK_consensus_filtered_pass.gds", storage.option = "LZ4_RA")
isolates <- seqOpen("moimix/PK_consensus_filtered_pass.gds") # read into R
seqSummary(isolates)
sample.id <- seqGetData(isolates, "sample.id")
seqCheck(isolates)
str(seqGetData(isolates, "annotation/info/AC"))

head(seqGetData(isolates, "sample.id"))

# moimix analysis
library(moimix)
coords <- getCoordinates(isolates) # get genomic coordinates of all variants

## estimating BAF matrix 
isolate_baf <- bafMatrix(isolates)

baf_df <- isolate_baf$coords %>%
  cbind(
    as.data.frame(isolate_baf$baf_site) %>% 
      rename(baf = "isolate_baf$baf_site")
    ) %>%
  left_join(
    isolate_baf$baf_matrix %>%
      as.data.frame() %>%
      t() %>%
      as.data.frame() %>%
      rownames_to_column("variant.id") %>%
      mutate(variant.id = as.numeric(variant.id)) 
    ) %>%
  mutate(chromosome = str_remove(chromosome, ".*PKNH_")) %>%
  mutate(chromosome = str_remove(chromosome, "_v2"))

## Plot BAF for different Fws samples with ggplot
plot_baf <- function(DATA, SAMPLE){
DATA %>%
    ggplot(aes_string(x = "variant.id", y = SAMPLE, colour = "chromosome")) + # use aes string so that we can paste the str in from the function arguments - passing the sample in not as a string does not work
    geom_point() +
    scale_colour_manual(values = c(rep_len(c("#404788FF", "#1F968BFF"), length(unique(baf_df$chromosome))))) +
    xlab("Chromosome") +
    ylab("NRAF") + 
    theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        legend.position = "bottom", 
        legend.title = element_blank(), 
        panel.border = element_rect(colour = "black", fill = NA, size = 1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
    guides(colour = guide_legend(nrow = 1))
}

baf_plot <- plot_baf(baf_df, "ERR2214850")
ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/MAF-Fws_plots/ERR2214850_72.png", width = 14, dpi = 300, baf_plot)

## Plot BAF for different Fws samples using base
#png('/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/MAF-Fws_plots/Fws_99.png', width = 14, height = 6, units = "in", res = 300)
#plot(isolate_baf, "SRR2222335", ylab = "NRAF")
#dev.off()

## Estimate MOI with binommix
set.seed(2022)
counts <- alleleCounts(isolates)
m1 <- binommix(counts, sample.id = "ERR2214837", k = 2)
summary(m1)
param.estimates <- getTheta(m1)
param.estimates

## Plot MOI and export ######################

## Estiamting MOI with Fws
fws_all <- getFws(isolates) %>%
  as.data.frame() %>%
  rownames_to_column(var = "sample") %>%
  rename("Proportion" = ".")

## Plot MOI distribution and export ######################

## output MOI data
fws_all %>%  
  write_tsv("moimix/fws_MOI.tsv")

## output samples with HIGH MOI data
fws_all %>%  
  filter(Proportion < 0.85) %>% 
  select(sample) %>%
  write.table("moimix/fws_high_MOI.txt", col.names = F, row.names = F, quote = F)

## plot MOI - col
fws_plot <- fws_all %>%
  arrange(desc(Proportion)) %>%
  add_column(Sample = 1:nrow(.)) %>%
  mutate(MOI = ifelse(Proportion >= 0.98, "No MOI",
        ifelse(Proportion < 0.98 & Proportion > 0.85, "Low MOI", "High MOI"))) %>%
  ggplot(aes(x = Sample, y = Proportion, colour = MOI)) +
  geom_col() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

ggsave("moimix/fws_plot.png", width = 10, fws_plot)

## plot MOI - hist
fws_plot <- fws_all %>%
  arrange(desc(Proportion)) %>%
  add_column(Sample = 1:nrow(.)) %>%
  ggplot(aes(x = Proportion)) +
  geom_histogram() +
  ylab("Frequency")

ggsave("moimix/fws_hist_plot.png", fws_plot)

# Boxplots by location
fws_all <- fws_all %>%
  mutate(Sample = str_remove(sample, "_DK.*")) %>%
  select(-sample) %>%
  left_join(
    readxl::read_xlsx("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/PK_Sabah_Sample_naming_indexes.xlsx") %>% 
      select(1) %>%
      add_column(Location = "Sabah") %>%
      rename(Sample = sampleid) %>%
      add_column(Cluster = "Sabah") %>% 
      rbind(
        read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_metadata.csv") %>%
        select(1, 3, 5) %>% 
        rename(Location = area) %>%
        mutate(Cluster = str_remove(Group, "-Pk")) %>% 
        select(-Group) 
      ) %>%
      rbind(
        read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_peninsular_metadata.csv") %>%
        select(2, 8) %>%
        rename(Sample = ENA_accession_no_ES) %>%
        add_column(Cluster = "Peninsular") 
      ) %>%
    mutate(Location = str_replace(Location, " ", "_")) 
    ) %>%
    mutate(MOI = ifelse(Proportion >= 0.98, "No MOI", ifelse(Proportion < 0.98 & Proportion > 0.85, "Low MOI", "High MOI"))) %>%
    na.omit() %>%
    mutate(Cluster = ifelse(grepl("PK_SB_DNA_011", Sample) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", Sample) | 
                                grepl("PK_SB_DNA_043", Sample) |
                                grepl("PK_SB_DNA_016", Sample) |  
                                grepl("PK_SB_DNA_092", Sample) | 
                                grepl("PK_SB_DNA_030", Sample) | 
                                grepl("PK_SB_DNA_093", Sample) | 
                                grepl("PK_SB_DNA_042", Sample) | 
                                grepl("PK_SB_DNA_063", Sample), "Mn", .$Cluster)) %>% # if not, just use values from Cluster
    mutate(Cluster = ifelse(Cluster == "Sabah", "Mf", .$Cluster))

# Summary 
fws_all %>%
  group_by(Cluster) %>%
  summarise(Count = n(), Mean = mean(Proportion), Below_95 = sum(Proportion <= .95), Below_85 = sum(Proportion <= .85))



# Boxplot and point-plot by location
fws_plot <- fws_all %>%
  ggplot(aes(y = Location, x = Proportion, by = Location)) +
  geom_boxplot() +
  geom_point(aes(colour = Location)) +
  geom_line(aes(colour = Location)) +
  scale_color_viridis_d() +
  theme(legend.position = "none") +
  xlab("Fws")

ggsave("moimix/fws_lcoations_boxplot.png", width = 8, fws_plot)

fws_plot <- fws_all %>%
  ggplot(aes(y = Cluster, x = Proportion, by = Cluster)) +
  geom_boxplot() +
  geom_point(aes(colour = Cluster)) +
  geom_line(aes(colour = Cluster)) +
  scale_colour_manual(values = c("#440154FF", "#73D055FF", "#39568CFF")) +
  theme(legend.position = "none") +
  xlab("Fws")

ggsave("moimix/fws_cluster_boxplot.png", width = 6, height = 8, dpi = 600, fws_plot)
```

## Exclude samples with high MOI

```{bash,eval=F}
bcftools view PK_consensus_filtered_pass.vcf -s ^ERR2214850,ERR2214856,ERR3374041,ERR985386,ERR985395,ERR985396,ERR985397,ERR985405,ERR985410,ERR985417,PK_SB_DNA_006_DKDL210002135-1a_HWHGKDSXY_L4,PK_SB_DNA_032_DKDL210002161-1a_HWHGKDSXY_L4,PK_SB_DNA_039_DKDL210002168-1a_HWHGKDSXY_L4,PK_SB_DNA_048_DKDL210002177-1a_HWHGKDSXY_L4,PK_SB_DNA_051_DKDL210002180-1a_HWHGKDSXY_L4,PK_SB_DNA_055_DKDL210002184-1a_HWHGKDSXY_L4,PK_SB_DNA_083_DKDL210002212-1a_HWHGKDSXY_L4 \
  > PK_consensus_no_MOI.vcf
```

## Normalise VCF & convert filtered vcf to bcf

 - Ensure that multi-allelic calls are split (1st part)
 - Left-align indels (2nd part)
 - Set the ID field to a unique value: CHROM:POS:REF:ALT (3rd part)
 
 --indiv-sort - specify how samples will be sorted - use to be consistent when adding metadata downstream 

```{bash,eval=F}
bcftools view PK_consensus_no_MOI.vcf| \
  sed 's/AD,Number=R/AD,Number=./' | \
  bcftools norm -m-any | \
  bcftools norm --check-ref w -f $FASTA/strain_A1_H.1.Icor.fasta | \
  bcftools annotate -Ob -x ID -I +'%CHROM:%POS:%REF:%ALT:%ID' \
  > PK_consensus_filtered_pass.bcf
```


## Convert to PLINK format 

Generic PLINK function to produce generic PLINK files
First create sample_order file & regions_to_mask file 

The regions we masked are: https://plasmodb.org/plasmo/service/users/current/steps/429386433/reports/attributesTabular & https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5918592/ and have to follow this format: https://www.cog-genomics.org/plink/1.9/input#make_set.
To find other hypervariable regions that may need to be masked, you can plot data from Pk.bim (created in step below) to acquire SNP frequencies/density in a sliding window approach.

```{bash,eval=F}
bcftools view PK_consensus_filtered_pass.bcf | sed -n '/#CHROM/,$p' | head -n 1 | tr '\t' '\n' |  sed -n '/ERR2214837/,$p' | awk '{print $1 "\t" $1}' > sample_order.txt

plink --bcf PK_consensus_filtered_pass.bcf \
  --keep-allele-order \
  --vcf-idspace-to _ \
  --double-id \
  --allow-extra-chr 0 \
  --make-bed \
  --indiv-sort file sample_order.txt \
  --exclude 'range' regions_to_mask.txt \
  --allow-no-sex \
  --out Pk
```

################## Optional 
### For SNP density across the genome

```{R,eval=f}
# Load packages
library(tidyverse)

# Read in data
snp_data <- read_table("Pk.bim", col_names = F) %>%
    select(X2) %>%
    separate(X2, into = c("Contig", "Pos", "Major", "Minor", "Nothing"), sep = ":") %>%
    mutate(Pos = as.numeric(Pos)) %>%
    select(1:2) %>%
    arrange(Contig, Pos)

# Create window 
window_size <- 1000

snp_plot <- snp_data %>%
    mutate(Window = (floor(Pos/window_size) * window_size)+ (window_size/2)) %>%
    group_by(Contig, Window) %>%
    summarise(SNPs = n()) %>%
    ggplot(aes(x = Window, y = SNPs, colour = Contig)) +
    geom_point() +
    facet_wrap(~Contig) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
    xlab("Chromosome position")

ggsave("SNP_density.png", dpi = 600, width = 14, snp_plot)
```


## Add metadata to PLINK fam

*Had to write an R script to manually add metadata as PLINK was being painful*

```{bash,eval=F}
sed 's/ /,/g' Pk.fam > Pk.csv
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/pheno_to_plink_fam.R
sed 's/,/ /g' Pk.csv > Pk.fam
```

### pheno_to_plink_fam.R

```{R,eval=F}
# Packages
library(tidyverse)
library(janitor)
library(data.table)

# Meatdata
Location_data <- readxl::read_xlsx("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/PK_Sabah_Sample_naming_indexes.xlsx") %>% 
  select(1) %>%
  add_column(Location = "Sabah") %>%
  rename(Sample = sampleid) %>%
  add_column(Cluster = "Sabah") %>% 
  rbind(
    read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_metadata.csv") %>%
      select(1, 3, 5) %>% 
      rename(Location = area) %>%
      mutate(Cluster = str_remove(Group, "-Pk")) %>% 
      select(-Group) 
  ) %>%
  rbind(
    read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_peninsular_metadata.csv") %>%
      select(2, 8) %>%
      rename(Sample = ENA_accession_no_ES) %>%
      add_column(Cluster = "Peninsular") 
  ) %>%
  mutate(Location = str_replace(Location, " ", "_"))


Plink_fam <- read_csv("Pk.csv", col_names=FALSE) %>%
  select(-ncol(.)) %>%
  mutate(Sample = str_remove(X2, "_DKD.*")) %>% 
  left_join(Location_data) %>% 
  select(-Sample) 

write_csv(Plink_fam, "Pk.csv", col_names = FALSE)
```

## Remove duplicates

```{bash,eval=F}
cut -f 2 Pk.bim | sort | uniq -d > Pk.dups
```

## Calculate minor allele frequency, genotypic missingness

```{eval=F}
plink --bfile Pk \
  --freq \
  --missing \
  --allow-no-sex \
  --out Pk
```

 - N_MISS =Number of missing SNPs
 - N_GENO =Number of non-obligatory missing genotypes
 - F_MISS =Proportion of missing SNPs

## Plot distribution of MAF and genotypic missingness

```{bash,eval=F}
Rscript /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/Exploratory_analyses/PLINK/maf_dist.R
Rscript /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/Exploratory_analyses/PLINK/miss_dist.R
```

### maf_dist.R

```{R,eval=F}
# Load packages
library(tidyverse)

# Read in data
pk_maf <- read_table("Pk.frq", col_names=T) %>%
    mutate(CHR = str_remove(SNP, ":.*"))

# Density of MAF by SNP
pk_maf_density <- pk_maf  %>%
    ggplot(aes(x = MAF)) +
    geom_density(alpha=.3) +
    geom_vline(xintercept = 0.05, colour = "#1F968BFF") 

ggsave("Pk_maf_density.png", dpi = 600, pk_maf_density)

# Plot MAF by chr
pk_maf_chr <- pk_maf %>% 
    mutate(SNP = str_remove(SNP, ":.*"),
        SNP = str_remove(SNP, "ordered_"),
        SNP = str_remove(SNP, "new_"),
        SNP = str_remove(SNP, "_v2"),
        SNP = str_remove(SNP, "A1.*")) %>%
    group_by(SNP) %>%
    summarise(MAF = mean(MAF)) %>%
    ggplot(aes(x = SNP, y = MAF)) +
    geom_col() +
    theme(axis.text.x = element_text(angle = 45), legend.position = "none") +
    scale_fill_viridis_d() 

ggsave("Pk_maf_plot_chr.png", dpi = 600, pk_maf_chr)

# Summary stats

pk_maf %>% 
    mutate(SNP = str_remove(SNP, ":.*")) %>%
    group_by(SNP) %>%
    summarise(MAF_mean = mean(MAF),
            MAF_SD = sd(MAF),
            MAF_SE = (sd(MAF))/sqrt(n())) %>%
rbind(
    pk_maf %>%
        summarise(MAF_mean = mean(MAF),
                MAF_SD = sd(MAF),
                MAF_SE = (sd(MAF))/sqrt(n())) %>%
        add_column(SNP = "Total")
    ) %>%
    write_csv(col_names = T, "pk_maf_summary.csv")
```

### miss_dist.R

```{R,eval=F}
# Load packages
library(tidyverse)
library(data.table)

# Sample-based missing data
## Read in data
sample_mis <- read_table("Pk.imiss", col_names=T) 

# Density of miss by sample
sample_miss_density <- sample_mis %>% 
    rename(Sample = IID,
        Geno_Miss = F_MISS) %>%
    mutate(Sample = str_remove(Sample, "DKD.*"),
            Sample = str_remove(Sample, "PK_SB_DNA_")) %>%
    select(2:ncol(.)) %>%
    ggplot(aes(x = Geno_Miss)) +
    geom_density(alpha=.3) +
    geom_vline(xintercept = 0.25, colour = "#1F968BFF") 

ggsave("sample_miss_density_plot.png", width = 12, dpi = 600, sample_miss_density)

# Variant-based missing data
## Read in data
var_miss <- read_table("Pk.lmiss", col_names=T) %>%
    mutate(CHR = str_remove(SNP, ":.*")) %>%
    rename(Geno_Miss = F_MISS,
            Chr = CHR)

# Density of miss by variant
var_miss_density <- var_miss %>% 
    ggplot(aes(x = Geno_Miss)) +
    geom_density(alpha=.3) +
    geom_vline(xintercept = 0.25, colour = "#1F968BFF") 

ggsave("variant_miss_density_plot.png", width = 12, dpi = 600, var_miss_density)

# Summary stats

miss_summary <- var_miss %>%
    summarise(Miss_mean = mean(Geno_Miss),
                Miss_SD = sd(Geno_Miss),
                Miss_SE = (sd(Geno_Miss))/sqrt(n()))

write_csv(miss_summary, "pk_miss_summary.csv")
```

## Apply filters based on MAF, genotypic missingess and Fsw

 - first recode creates the intial .map and .ped files
 - recode generates new .ped and .map files
 - --make-bed generates .bed, .bim and .fam files
 - -mind filters per sample (missingness)
 - -geno filters per variant (missingness)

**FILTERED:**
 - exclude variants with > 25% missing alleles
 - exclude variants with minor allele frequencies > 5%
 - exclude samples with > 25% missing alleles

Values selected based on the distrbutions plotted previously.

MAF (=Minor Allele Frequency) = P(geno=1, Aa)x0.5 + P(geno=2, aa)

```{bash,eval=F}
plink --bfile Pk --recode --out Pk
plink --file Pk --maf 0.05 --geno 0.25 --mind 0.25 --make-bed --out cleaned
```


#########################################################################################################



# How do the Sabah samples fit into the previously defined clusters?

## Use PLINK to create PCA and MDS data, and a distance matrix (identity-by-state)

The different functions:
    --cluster is MDS
    --distance is identity by state
    --genome is identity by decent
        unbounded - "The underlying P(IBD=0/1/2) estimator sometimes yields numbers outside the range [0,1]; by default, these are clipped. The 'unbounded' modifier turns off this clipping."
        nudge - "Then, if PI_HAT2 < P(IBD=2), 'nudge' adjusts the final estimates to P(IBD=0) := (1-p2), P(IBD=1) := 2p(1-p), and P(IBD=2) := p2, where p is the current PI_HAT."

```{bash,eval=F}
plink --bfile cleaned --cluster --mds-plot 10 --out Pk
plink --bfile cleaned --pca --out Pk
plink --bfile cleaned --distance square --out Pk
plink --bfile cleaned --genome gz unbounded nudge --out Pk
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/ordination_plots.R
```

### ordination_plots.R

```{R,eval=F}
# Load packages
library(tidyverse)
library(ape)
library(ggtree)

# PCA

## Read in data
pca <- read_table("Pk.eigenvec", col_names=F) %>%
    select(-1) %>%
    mutate(X2 = str_remove(X2, "_DK.*")) %>%
    rename("sampleid" = "X2") %>%
    rename_at(vars(starts_with("X")), ~str_replace(., "X.*", paste0("PC", seq_along(.))))

eigenval <- scan("Pk.eigenval")

## Add metadata 
pca <- pca %>%
    left_join(read_csv("Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(sampleid = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>%
        mutate(sampleid = str_remove(sampleid, "_DK.*")))


## Percentage variance explained by each PC
pve_plot <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100) %>%
    ggplot(aes(PC, pve)) + 
    geom_bar(stat = "identity") + 
    ylab("Percentage variance explained") + 
    theme_light()

ggsave("Pk_clusters/Percentage_variance_explained.png", dpi=600, pve_plot)


## PCA - clusters
pve <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100)

pca_plot <- ggplot(pca, aes(PC1, PC2, colour = Cluster)) + 
    geom_point(size = 3) +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF", "#73D055FF")) +
    coord_equal() + 
    theme_light() + 
    xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + 
    ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))

ggsave("Pk_clusters/PCA.png", dpi = 600, pca_plot)


# MDS - clusters

## Read in data
mds <- read_table("Pk.mds", col_names=T) %>%
    select(-c("FID", "X14")) %>%
    mutate(IID = str_remove(IID, "_DK.*")) %>%
    rename("sampleid" = "IID") %>%
    rename_at(vars(starts_with("C")), ~str_replace(., "C", "MDS"))

## Add metadata 
mds <- mds %>%
    left_join(read_csv("Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(sampleid = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>%
        mutate(sampleid = str_remove(sampleid, "_DK.*")))

## Plot MDS
mds_plot <- ggplot(mds, aes(MDS1, MDS2, colour = Cluster)) + 
    geom_point(size = 3) +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF", "#73D055FF")) +
    coord_equal() + 
    theme_light()

ggsave("Pk_clusters/MDS.png", dpi=600, mds_plot)

# Neighbour-joining tree

## Create distance matrix from PLINK data and build NJT
NJT_ID <- read_table("Pk.dist.id", col_names=F) %>%
    as.data.frame() %>%
    mutate_all(~str_remove(., "_DKD.*")) 

NJT_matrix <- read_table("Pk.dist", col_names=NJT_ID$X1) %>%
    as.data.frame() %>%
    add_column(Row_Names = NJT_ID$X1) %>%
    column_to_rownames("Row_Names") %>%
    as.matrix()

NJT_tree <- nj(NJT_matrix)

## Plot tree
# options(ignore.negative.edge=TRUE)

# Read in initial metadata
NJT_metadata <- read_csv("Pk.csv", col_names = FALSE) %>%
    select(1, 6, 7) %>%
    rename(Sample = X1) %>%
    rename(Location = X6) %>%
    rename(Cluster = X7) %>%
    mutate(Sample = str_remove(Sample, "_DKD.*")) %>%
    mutate_if(is.character, as.factor) %>%
    mutate(Region = ifelse(Location == "Sabah" | Location == "Betong" | Location == "Kapit" | Location == "Sarikei", "Borneo", "Peninsular")) 

# Update metadata with more detailed location - districts - where we split up the Sabah samples (which represent OUR data), and also the clinic samples
NJT_metadata <- NJT_metadata %>%
    left_join(
    readxl::read_xlsx("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/PK_Sabah_Sample_naming_indexes.xlsx") %>% # index data
        mutate(subjectid = as.character(subjectid)) %>%
        mutate(subjectid = ifelse(str_length(subjectid) == 1, paste0("00", subjectid), # if subjectid length = 1 paste 00
                     ifelse(str_length(subjectid) == 2, paste0("0", subjectid), # if not, then if subjectid length = 2 paste 0
                     .$subjectid))) %>% # if not, keep value the same
    mutate(studycode = paste0(group, subjectid)) %>% # create studycode column that aligns with metadata
    left_join( # left join the epi data to the index data so that we only get epi data for the samples that we have
    read_tsv("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/Epi_data_ZB_clean.tsv")) %>%
    select(sampleid, district) %>%
    rename(Sample = sampleid) %>%
    rename(District = district) %>%
    mutate(District = str_replace(District, " ", "_"))
    ) %>% 
    mutate_if(is.factor, as.character) %>%
    mutate(District = ifelse(is.na(District), .$Location, .$District)) %>% 
    mutate(District = ifelse(Sample == "SRR2221468", "Hackeri", 
        ifelse(Sample == "SRR2222335", "H(AW)", 
        ifelse(Sample == "SRR2225467", "Malayan", 
        ifelse(Sample == "SRR2225571", "MR4-H", 
        ifelse(Sample == "SRR2225573", "Philippine",
        ifelse(Sample == "SRR3135172", "YH1", .$District))))))) %>%
    mutate(District = str_replace(District, "Kinaalu", "Kinabalu"))

options(ignore.negative.edge=TRUE)

# Cluster - Sabah samples vs previously defined clusters
NJT_tree_plot <- ggtree(NJT_tree, layout="circular", size = 0.5, aes(colour = Cluster)) %<+% NJT_metadata +
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank()) +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF", "#73D055FF"))
    
ggsave("Pk_clusters/NJT_tree_cluster_rooted.png", dpi = 600, height = 8, width = 16, NJT_tree_plot)

NJT_tree_plot <- ggtree(NJT_tree, layout="daylight", size = 0.5, aes(colour = Cluster)) %<+% NJT_metadata +
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank()) +
    scale_color_manual(values = c("#73D055FF", "#440154FF", "#39568CFF", "#1F968BFF"))
    
ggsave("Pk_clusters/NJT_tree_cluster_unrooted.png", dpi = 600, height = 15, width = 15, limitsize = FALSE, NJT_tree_plot)

#Labels
NJT_tree_plot <- ggtree(NJT_tree, layout="circular", size = 0.5, aes(colour = Cluster)) %<+% NJT_metadata +
    theme(legend.position = "right", 
    legend.title = element_blank(), 
    legend.key = element_blank()) +
    geom_tiplab(size = 2) +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF", "#73D055FF"))
    
ggsave("Pk_clusters/NJT_tree_labelled_rooted.png", dpi = 600, height = 20, width = 20, limitsize = FALSE, NJT_tree_plot)

NJT_tree_plot <- ggtree(NJT_tree, layout="daylight", size = 0.5, aes(colour = Cluster)) %<+% NJT_metadata +
    theme(legend.position = "right", 
    legend.title = element_blank(), 
    legend.key = element_blank()) +
    geom_tiplab(size = 2) +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF", "#73D055FF"))

ggsave("Pk_clusters/NJT_tree_labelled_unrooted.png", dpi = 600, height = 20, width = 20, limitsize = FALSE, NJT_tree_plot)
```



#########################################################################################################



# How do the Sabah samples fit into the Mn and Mf clusters (within Malaysian Borneo)?

## Here we calculate the distances based on a subset of the data by excluding Peninsular samples, so that we're only investigating Mn and Mf clusters.

```{bash,eval=F}
cat Pk.fam | grep 'Peninsular' > exclude_samples.txt

plink --bfile cleaned --remove exclude_samples.txt --cluster --mds-plot 10 --out Pk
plink --bfile cleaned --remove exclude_samples.txt --pca --out Pk
plink --bfile cleaned --remove exclude_samples.txt --distance square --out Pk
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_VS_Mn/ordination_plots_Mf_VS_Mn.R
```

## Fst statistic - Mn vs Mf

First create a file that contains the clusters you want use for the Fst. You can ammend the Pk.fam file (or metadata), but the file needs to be structured in PLINKs weird FID-IID-Cluster format.

Steps:
    - Apply filter (as done previously)
    - Attach metadata to fam file (as done previously)
    - Define clusters using metadata
    - Calculate and plot Fst for a-priori clusters (sample_clusters.txt)

```{bash,eval=F}
plink --file Pk --maf 0.05 --geno 0.25 --mind 0.25 --make-bed --out cleaned
sed 's/ /,/g' cleaned.fam > Pk.csv
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/pheno_to_plink_fam.R
sed 's/,/ /g' Pk.csv > Pk.fam

Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_VS_Mn/sample_clusters.R
sed 's/,/ /g' sample_clusters.csv > sample_clusters.txt
rm -f sample_clusters.csv

plink --bfile cleaned \
  --within sample_clusters.txt \
  --remove exclude_samples.txt \
  --fst \
  --allow-no-sex \
  --out Pk

Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_VS_Mn/Fst_plot.R
```

### sample_clusters.R script for defining clusters to compare for Fst

```{R,eval=F}
library(tidyverse)

read_csv("Pk.csv", col_names = F) %>% 
    mutate(Fst_cluster = ifelse(grepl("PK_SB_DNA_011", X1) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", X1) | 
                                grepl("PK_SB_DNA_043", X1) |
                                grepl("PK_SB_DNA_016", X1) |  
                                grepl("PK_SB_DNA_092", X1) | 
                                grepl("PK_SB_DNA_030", X1) | 
                                grepl("PK_SB_DNA_093", X1) | 
                                grepl("PK_SB_DNA_042", X1) | 
                                grepl("PK_SB_DNA_063", X1), "Mn", .$X7)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Fst_cluster = ifelse(Fst_cluster == "Sabah", "Mf", .$Fst_cluster)) %>% # if its the remaining Sabah samples, make them Mn, else keep them the same
    select(1, 2, 8) %>%
    filter(Fst_cluster != "Peninsular") %>%
    write_csv("sample_clusters.csv", col_names = F)
```


### Plot Fst 

Fst_plot.R

```{R,eval=F}
library(tidyverse)

Fst_matrix <- read_table("Pk.fst", col_names=T) %>%
    mutate(CHR = str_remove(SNP, "ordered_PKNH_")) %>%
    mutate(CHR = str_remove(CHR, "_v2.*"))

# Sliding window
window_size <- 10000

Fst_plot_window <- Fst_matrix %>%
    filter(FST != "nan") %>%
    mutate(FST = as.numeric(FST)) %>%
    mutate(Window = (floor(POS/window_size) * window_size)+ (window_size/2)) %>%
    group_by(CHR, Window) %>%
    summarise(Window_Fst = mean(FST), count = n()) %>% # caculate the mean Fst for each window in each chr
    filter(count > 10) %>% # remove windows with low counts
    ggplot(aes(x = Window, y = Window_Fst, colour = CHR)) +
    geom_point() +
    facet_wrap(~CHR) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
    ylab("Fst") +
    xlab("Windows (10000)")  +
    scale_color_viridis_d("Chr") +
    geom_hline(yintercept = 0.4, colour = "#1F968BFF")

ggsave("Pk_Mn_vs_Mf/Fst_sliding_window_plot.png", dpi = 600, width = 14, Fst_plot_window)

Fst_plot <- Fst_matrix %>%
    filter(FST != "nan") %>%
    mutate(FST = as.numeric(FST)) %>%
    ggplot(aes(x = POS/1000000, y = FST, colour = CHR)) +
    geom_point() +
    #theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    ylab("Fst") +
    xlab("Chromsome position (Mb)")  +
    scale_color_viridis_d("Chr") +
    facet_wrap(~CHR)

ggsave("Pk_Mn_vs_Mf/Fst_manhattan.png", dpi = 600, width = 20, Fst_plot)

Fst_plot <- Fst_matrix %>%
    filter(FST != "nan") %>%
    filter(CHR == "14") %>%
    mutate(FST = as.numeric(FST)) %>%
    ggplot(aes(x = POS, y = FST)) +
    geom_point() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
    ylab("Fst") +
    xlab("SNPs")  +
    scale_color_viridis_d("Chr") 

ggsave("Pk_Mn_vs_Mf/Fst_manhattan_14.png", dpi = 600, width = 14, Fst_plot)
```



# Variant selection - [rehh](https://cran.r-project.org/web/packages/rehh/vignettes/rehh.html#LoadData)

Since the whole file is read in for rehh, even though you can only look at a single chr at a time, it is advisable to split large data sets into chromosome-specific files.
In addition, computing Rsb requires two data frames containing each the inES statistics of one population as obtained by the scan_hh() function - THUS - need to create separate VCF files for the different clusters that we want to compare.

Steps:
    - create a lsit of the chromosomes
    - get a list of samples to be included (-sn) in the rehh analysis (from Mn of Mf)
    - subset VCF into a VCF for individual chromosomes using SelectVariants 
    - then use sample_clusters.txt created previously for the Fst calculations to create a list of 
    - SelectVariants is then used again to subset each of the chromosome-specific VCF files into the two clusters
    -XL is used to filter out problematic regions (SICAvar, Kir and sub-telomeric regions - as previously defined)

**Need to ammend the sn and xl line manually based on outputs of "cat"**

```{bash}
grep '##contig=<ID=ordered' PK_consensus_no_MOI.vcf| sed 's/##contig=<ID=//g' | sed 's/,.*//g' > rehh/chromosome.txt

cat sample_clusters.txt | awk '{print "-sn " $1}' | tr '\n' ' '

REGIONS_TO_MASK="/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/regions_to_mask.list"

for i in $(cat rehh/chromosome.txt)
    do 
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V PK_consensus_no_MOI.vcf \
        -L $i \
        -sn ERR2214837 -sn ERR2214838 -sn ERR2214839 -sn ERR2214840 -sn ERR2214841 -sn ERR2214842 -sn ERR2214843 -sn ERR2214844 -sn ERR2214845 -sn ERR2214846 -sn ERR2214847 -sn ERR2214848 -sn ERR2214849 -sn ERR2214851 -sn ERR2214852 -sn ERR2214853 -sn ERR2214854 -sn ERR2214855 -sn ERR2214857 -sn ERR274221 -sn ERR274222 -sn ERR274224 -sn ERR274225 -sn ERR366425 -sn ERR366426 -sn ERR985372 -sn ERR985373 -sn ERR985374 -sn ERR985375 -sn ERR985376 -sn ERR985377 -sn ERR985378 -sn ERR985379 -sn ERR985380 -sn ERR985381 -sn ERR985382 -sn ERR985383 -sn ERR985384 -sn ERR985385 -sn ERR985387 -sn ERR985388 -sn ERR985389 -sn ERR985390 -sn ERR985391 -sn ERR985392 -sn ERR985393 -sn ERR985394 -sn ERR985398 -sn ERR985399 -sn ERR985400 -sn ERR985401 -sn ERR985402 -sn ERR985403 -sn ERR985404 -sn ERR985406 -sn ERR985407 -sn ERR985408 -sn ERR985409 -sn ERR985411 -sn ERR985412 -sn ERR985413 -sn ERR985415 -sn ERR985416 -sn ERR985418 -sn ERR985419 -sn PK_SB_DNA_001_DKDL210002130-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_002_DKDL210002131-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_003_DKDL210002132-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_005_DKDL210002134-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_008_DKDL210002137-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_009_DKDL210002138-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_010_DKDL210002139-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_011_DKDL210002140-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_013_DKDL210002142-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_014_DKDL210002143-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_015_DKDL210002144-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_016_DKDL210002145-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_021_DKDL210002150-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_023_DKDL210002152-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_028_DKDL210002157-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_029_DKDL210002158-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_030_DKDL210002159-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_031_DKDL210002160-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_034_DKDL210002163-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_035_DKDL210002164-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_036_DKDL210002165-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_038_DKDL210002167-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_040_DKDL210002169-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_043_DKDL210002172-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_044_DKDL210002173-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_045_DKDL210002174-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_046_DKDL210002175-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_047_DKDL210002176-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_049_DKDL210002178-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_052_DKDL210002181-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_053_DKDL210002182-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_054_DKDL210002183-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_057_DKDL210002186-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_067_DKDL210002196-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_068_DKDL210002197-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_069_DKDL210002198-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_070_DKDL210002199-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_072_DKDL210002201-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_073_DKDL210002202-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_074_DKDL210002203-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_077_DKDL210002206-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_081_DKDL210002210-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_082_DKDL210002211-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_086_DKDL210002215-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_087_DKDL210002216-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_088_DKDL210002217-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_089_DKDL210002218-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_090_DKDL210002219-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_091_DKDL210002220-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_092_DKDL210002221-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_094_DKDL210002223-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o rehh/${i}.vcf.gz
    done 

cat sample_clusters.txt | grep 'Mn' | awk '{print "-sn " $1}' | tr '\n' ' '
cat sample_clusters.txt | grep 'Mn' | awk '{print "-xl_sn " $1}' | tr '\n' ' '


cd rehh
ls *vcf.gz > vcf_list.txt


for i in $(cat vcf_list.txt)
    do  
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V $i \
        -sn ERR2214837 -sn ERR2214838 -sn ERR2214839 -sn ERR2214840 -sn ERR2214841 -sn ERR2214842 -sn ERR2214843 -sn ERR2214844 -sn ERR2214845 -sn ERR2214846 -sn ERR2214847 -sn ERR2214848 -sn ERR2214849 -sn ERR2214851 -sn ERR2214852 -sn ERR2214853 -sn ERR2214854 -sn ERR2214855 -sn ERR2214857 -sn ERR274224 -sn ERR274225 -sn ERR366425 -sn ERR985411 -sn ERR985412 -sn ERR985413 -sn ERR985415 -sn ERR985416 -sn ERR985418 -sn ERR985419 -sn PK_SB_DNA_011_DKDL210002140-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_016_DKDL210002145-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_030_DKDL210002159-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_043_DKDL210002172-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_091_DKDL210002220-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_092_DKDL210002221-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o ${i%.vcf.gz}_Mn.vcf.gz
    done

for i in $(cat vcf_list.txt)
    do  
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V $i \
        -xl_sn ERR2214837 -xl_sn ERR2214838 -xl_sn ERR2214839 -xl_sn ERR2214840 -xl_sn ERR2214841 -xl_sn ERR2214842 -xl_sn ERR2214843 -xl_sn ERR2214844 -xl_sn ERR2214845 -xl_sn ERR2214846 -xl_sn ERR2214847 -xl_sn ERR2214848 -xl_sn ERR2214849 -xl_sn ERR2214851 -xl_sn ERR2214852 -xl_sn ERR2214853 -xl_sn ERR2214854 -xl_sn ERR2214855 -xl_sn ERR2214857 -xl_sn ERR274224 -xl_sn ERR274225 -xl_sn ERR366425 -xl_sn ERR985411 -xl_sn ERR985412 -xl_sn ERR985413 -xl_sn ERR985415 -xl_sn ERR985416 -xl_sn ERR985418 -xl_sn ERR985419 -xl_sn PK_SB_DNA_011_DKDL210002140-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_016_DKDL210002145-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_030_DKDL210002159-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_043_DKDL210002172-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_091_DKDL210002220-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_092_DKDL210002221-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o ${i%.vcf.gz}_Mf.vcf.gz
    done

Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_VS_Mn/rehh.R
```

Some of the different stats that can be calculated with rehh:
    - EHH - Homozygosity (EHH) is defined as the probability that two randomly chosen chromosomes, carrying the core allele, are homozygous over a given surrounding chromosomal region.
    - iHH - EHH starts at 1 and decays to 0 with increasing distance of t from the focal marker s. For a given core allele, the integrated EHH (iHH) is defined as the area under the EHH curve which, in turn, is defined by the EHH values and associated chromosomal positions.
    - EHHS - quantity is aimed to reflect the probability that any two randomly chosen chromosomes from a population are homozygous over a given surrounding chromosomal region of a focal marker. 
    - iES - Like EHH, EHHS has its maximum at the focal marker and decays to 0 with increasing distance from the focal marker. For a given focal marker, analogously to iHH, iES is defined as the integrated EHHS.

```{R,eval=F}
# Packages
library(rehh)
library(tidyverse)
library(data.table)
library(R.utils)

# WHOLE GENOME iHS

## need to read in chr files independently and build up an object resulting from the scan_hh outputs
### first create an empty df to be appended in the loop and a list of chr names
### the loop then creates the vcf file name (hap_file), and reads in the file to hh, which is then parsed into the scan_hh function to produce an object i that is appended to the growing wgscan object
### the resulting wgscan produces the same results as the scan_hh object above
### this can then be parsed into ihh2ihs

chromsomes_files <- c("ordered_PKNH_01_v2", "ordered_PKNH_02_v2", "ordered_PKNH_03_v2", "ordered_PKNH_04_v2", "ordered_PKNH_05_v2",
    "ordered_PKNH_06_v2", "ordered_PKNH_07_v2", "ordered_PKNH_08_v2", "ordered_PKNH_09_v2", "ordered_PKNH_10_v2", "ordered_PKNH_11_v2",
    "ordered_PKNH_12_v2", "ordered_PKNH_13_v2", "ordered_PKNH_14_v2")

wgscan <- tibble("CHR" = NA, "POSITION" = NA, "FREQ_A" = NA, "FREQ_D" = NA, "NHAPLO_A" = NA, "NHAPLO_D" = NA, "IHH_A" = NA, "IHH_D" = NA, "IES" = NA, "INES" = NA)

for(i in chromsomes_files) {
    # haplotype file name for each chromosome
    hap_file = paste(i, ".vcf.gz", sep = "") # filename pattern
    # create internal representation
    hh <- data2haplohh(hap_file = hap_file, 
                    chr.name = i, 
                    polarize_vcf = FALSE, # if the AA key is absent
                    min_perc_geno.mrk = 100, # discard markers genotyped on < 100% of haplotypes
                    min_maf = 0.05, # discard markers with a minor allele frequency of < 0.05)
                    vcf_reader = "data.table") # use this package
    scan <- scan_hh(hh) # perform scan on a single chromosome (calculate iHH/iES values)
    wgscan <- wgscan %>% # append the wgscan object every round with additional data
    rbind(scan)
}

wgs_ihs <- wgscan %>%
    na.omit() %>%
    ihh2ihs(min_maf = 0.05, # default
            freqbin = 0.025 # default
            )

### Output is a list with two elements:
### ihs: a data frame with iHS and corresponding p-value piHS (p-value in a âˆ’log10 scale assuming the iHS are normally distributed under the neutral hypothesis)
### frequency.class: a data frame summarizing the derived allele frequency bins used for standardization and mean and standard deviation of the un-standardized values
wgs_ihs$ihs
wgs_ihs$frequency.class


# Scan for candidate regions 
## First update iHS table so that its compatible
wgs_ihs_2 <- wgs_ihs
wgs_ihs_2$ihs <- wgs_ihs_2$ihs %>%
    mutate(CHR = str_remove(CHR, "ordered_PKNH_")) %>%
    mutate(CHR = str_remove(CHR, "_v2")) %>%
    mutate(CHR = as.numeric(CHR)) %>%
    as.data.frame()

candidate_regions <- calc_candidate_regions(wgs_ihs_2,
                                 threshold = 4,
                                 pval = TRUE,
                                 window_size = 10000,
                                 overlap = 1000,
                                 min_n_extr_mrk = 3) %>%
                                 add_column(Stat = "iHS") 


## Plot iHS and iHS p-values

iHS <- wgs_ihs$ihs %>%
    mutate(CHR = str_remove(CHR, "ordered_PKNH_")) %>%
    mutate(CHR = str_remove(CHR, "_v2")) %>%
    ggplot(aes(x = CHR, y = IHS, colour = CHR)) +
    geom_jitter() + 
    theme(legend.position = "none") +
    scale_colour_viridis_d() +
    scale_x_discrete() +
    xlab("Chromsome") +
    ylab("iHS") 

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/iHS.png", dpi = 600, width = 10, iHS)

iHS_pvalue <- wgs_ihs$ihs %>%
    mutate(CHR = str_remove(CHR, "ordered_PKNH_")) %>%
    mutate(CHR = str_remove(CHR, "_v2")) %>%
    ggplot(aes(x = CHR, y = LOGPVALUE, colour = CHR)) +
    geom_jitter() + 
    geom_hline(yintercept = 4, linetype = 2, colour = "#1F968BFF") + # adjusted with bonferonni
    theme(legend.position = "none") +
    scale_colour_viridis_d() +
    xlab("Chromosomes") +
    ylab("iHS p-value")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/iHS_pvalue.png", dpi = 600, width = 10, iHS_pvalue)


## Plotting using rehh

png('/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/iHS_rehh.png')
manhattanplot(wgs_ihs)
dev.off()

png('/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/iHS_p_rehh.png')
manhattanplot(wgs_ihs, pval = TRUE, threshold = 4)
dev.off()


## QQ-Plot
png('/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/iHS_QQ.png')
distribplot(wgs_ihs$ihs$IHS, 
            xlab = "iHS", 
            qqplot = TRUE)
dev.off()

#######################################################################################


# Comparing EHH between clusters

# CALCULATE Rsb  and XP-EHH

## pairwise population statistic 
## ratio of EHHS between populations

chromsomes_files <- c("ordered_PKNH_01_v2", "ordered_PKNH_02_v2", "ordered_PKNH_03_v2", "ordered_PKNH_04_v2", "ordered_PKNH_05_v2",
    "ordered_PKNH_06_v2", "ordered_PKNH_07_v2", "ordered_PKNH_08_v2", "ordered_PKNH_09_v2", "ordered_PKNH_10_v2", "ordered_PKNH_11_v2",
    "ordered_PKNH_12_v2", "ordered_PKNH_13_v2", "ordered_PKNH_14_v2")

### Mn
wgscan_Mn <- tibble("CHR" = NA, "POSITION" = NA, "FREQ_A" = NA, "FREQ_D" = NA, "NHAPLO_A" = NA, "NHAPLO_D" = NA, "IHH_A" = NA, "IHH_D" = NA, "IES" = NA, "INES" = NA)

for(i in chromsomes_files) {
    # haplotype file name for each chromosome
    hap_file = paste(i, "_Mn.vcf.gz", sep = "") # filename pattern
    # create internal representation
    hh <- data2haplohh(hap_file = hap_file, 
                    chr.name = i, 
                    polarize_vcf = FALSE, # if the AA key is absent
                    min_perc_geno.mrk = 100, # discard markers genotyped on < 100% of haplotypes
                    min_maf = 0.05, # discard markers with a minor allele frequency of < 0.05)
                    vcf_reader = "data.table") # use this package
    scan <- scan_hh(hh) # perform scan on a single chromosome (calculate iHH/iES values)
    wgscan_Mn <- wgscan_Mn %>% # append the wgscan object every round with additional data
    rbind(scan)
}

### Mf
wgscan_Mf <- tibble("CHR" = NA, "POSITION" = NA, "FREQ_A" = NA, "FREQ_D" = NA, "NHAPLO_A" = NA, "NHAPLO_D" = NA, "IHH_A" = NA, "IHH_D" = NA, "IES" = NA, "INES" = NA)

for(i in chromsomes_files) {
    # haplotype file name for each chromosome
    hap_file = paste(i, "_Mf.vcf.gz", sep = "") # filename pattern
    # create internal representation
    hh <- data2haplohh(hap_file = hap_file, 
                    chr.name = i, 
                    polarize_vcf = FALSE, # if the AA key is absent
                    min_perc_geno.mrk = 100, # discard markers genotyped on < 100% of haplotypes
                    min_maf = 0.05, # discard markers with a minor allele frequency of < 0.05)
                    vcf_reader = "data.table") # use this package
    scan <- scan_hh(hh) # perform scan on a single chromosome (calculate iHH/iES values)
    wgscan_Mf <- wgscan_Mf %>% # append the wgscan object every round with additional data
    rbind(scan)
}

# Rsb
Rsb <- ines2rsb(
        scan_pop1 = wgscan_Mn,
        scan_pop2 = wgscan_Mf,
        popname1 = "Mn",
        popname = "Mf"
    )

# XP-EHH
XP_EHH <- ies2xpehh(
        scan_pop1 = wgscan_Mn,
        scan_pop2 = wgscan_Mf,
        popname1 = "Mn",
        popname = "Mf"
    )


# Scan for candidate regions 
candidate_regions_Rsb <- calc_candidate_regions(Rsb,
                                 threshold = 4,
                                 pval = TRUE,
                                 window_size = 10000,
                                 overlap = 1000,
                                 min_n_extr_mrk = 3) %>%
                                 add_column(Stat = "Rsb")

candidate_regions_XP_EHH <- calc_candidate_regions(XP_EHH,
                                 threshold = 4,
                                 pval = TRUE,
                                 window_size = 10000,
                                 overlap = 1000,
                                 min_n_extr_mrk = 3) %>%
                                 add_column(Stat = "XP_EHH")

## Save candidate regions for all stats
candidate_regions %>%
    rbind(candidate_regions_Rsb) %>%
    rbind(candidate_regions_XP_EHH) %>%
    write_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/candidate_regions.csv")


# Plot Rsb and p-value
Rsb_plot <- Rsb %>%
    na.omit() %>%
    mutate(CHR = str_remove(CHR, "ordered_PKNH_")) %>%
    mutate(CHR = str_remove(CHR, "_v2")) %>%
    rename("Rsb" = "RSB_Mn_Mf") %>%
    ggplot(aes(x = CHR, y = Rsb, colour = CHR)) +
    geom_jitter() + 
    theme(legend.position = "none") +
    scale_colour_viridis_d() +
    xlab("Chromsome") +
    ylab("Rsb") 

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/Rsb.png", dpi = 600, width = 10, Rsb_plot)

Rsb_pvalue <- Rsb %>%
    na.omit() %>%
    mutate(CHR = str_remove(CHR, "ordered_PKNH_")) %>%
    mutate(CHR = str_remove(CHR, "_v2")) %>%
    ggplot(aes(x = CHR, y = LOGPVALUE, colour = CHR)) +
    geom_jitter() + 
    geom_hline(yintercept = 4, linetype = 2, colour = "#1F968BFF") + # adjusted with bonferonni
    theme(legend.position = "none") +
    scale_colour_viridis_d() +
    xlab("Chromosomes") +
    ylab("Rsb p-value")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/Rsb_pvalue.png", dpi = 600, width = 10, Rsb_pvalue)

#Plot XP-EHH and p-value

XP_EHH_plot <- XP_EHH %>%
    na.omit() %>%
    mutate(CHR = str_remove(CHR, "ordered_PKNH_")) %>%
    mutate(CHR = str_remove(CHR, "_v2")) %>%
    ggplot(aes(x = CHR, y = XPEHH_Mn_Mf, colour = CHR)) +
    geom_jitter() + 
    theme(legend.position = "none") +
    scale_colour_viridis_d() +
    xlab("Chromsome") +
    ylab("XP_EHH") 

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/XP_EHH.png", dpi = 600, width = 10, XP_EHH_plot)

XP_EHH_pvalue <- XP_EHH %>%
    na.omit() %>%
    mutate(CHR = str_remove(CHR, "ordered_PKNH_")) %>%
    mutate(CHR = str_remove(CHR, "_v2")) %>%
    ggplot(aes(x = CHR, y = LOGPVALUE, colour = CHR)) +
    geom_jitter() + 
    geom_hline(yintercept = 4, linetype = 2, colour = "#1F968BFF") + # adjusted with bonferonni
    theme(legend.position = "none") +
    scale_colour_viridis_d() +
    xlab("Chromosomes") +
    ylab("XP_EHH p-value")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/XP_EHH_pvalue.png", dpi = 600, width = 10, XP_EHH_pvalue)

# Rsb vs XP-EHH

Rsb_XP_comaprison <- Rsb %>%
    left_join(
        XP_EHH %>%
            select(1:3)
        ) %>%
    na.omit() %>%
    ggplot(aes(x = RSB_Mn_Mf, y = XPEHH_Mn_Mf, colour = CHR)) +
    geom_point() + 
    theme(legend.position = "none") +
    scale_colour_viridis_d() +
    xlab("Rsb") +
    ylab("XP_EHH") +
    geom_abline()

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk_Mn_vs_Mf/XP_Rsb_comparison.png", dpi = 600, width = 10, Rsb_XP_comaprison)
```




#########################################################################################################




# Comparing Mn and Mf - just for Sabah samples

## Here we subset calculate the distances based on a subset of the data by only including Sabah samples.

```{bash,eval=F}
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses

cat Pk.fam | grep 'Sabah' > include_samples.txt

plink --bfile cleaned --keep include_samples.txt --cluster --mds-plot 10 --out Pk
plink --bfile cleaned --keep include_samples.txt --pca --out Pk
plink --bfile cleaned --keep include_samples.txt --distance square --out Pk
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_VS_Mn_SABAH_ONLY/ordination_plots_Mf_VS_Mn_SABAH_ONLY.R
```

## Fst statistic - Mn vs Mf within Sabah

```{bash,eval=F}
plink --file Pk --maf 0.05 --geno 0.25 --mind 0.25 --make-bed --out cleaned
sed 's/ /,/g' cleaned.fam > Pk.csv
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/pheno_to_plink_fam.R
sed 's/,/ /g' Pk.csv > Pk.fam

Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_VS_Mn_SABAH_ONLY/sample_clusters_Mf_VS_Mn_SABAH_ONLY.R
sed 's/,/ /g' sample_clusters.csv > sample_clusters.txt
rm -f sample_clusters.csv

plink --bfile cleaned \
  --within sample_clusters.txt \
  --keep include_samples.txt \
  --fst \
  --allow-no-sex \
  --out Pk

Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_VS_Mn_SABAH_ONLY/Fst_plot_Mf_VS_Mn_SABAH_ONLY.R
```


# Variant selection - between Mn and Mf within Sabah

**Need to ammend the sn and xl line manually based on outputs of "cat"**

```{bash}
grep '##contig=<ID=ordered' PK_consensus_no_MOI.vcf| sed 's/##contig=<ID=//g' | sed 's/,.*//g' > rehh/chromosome.txt

cat sample_clusters.txt | awk '{print "-sn " $1}' | tr '\n' ' '
rm -f rehh/*vcf.gz*

for i in $(cat rehh/chromosome.txt)
    do 
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V PK_consensus_no_MOI.vcf \
        -L $i \
        -sn PK_SB_DNA_001_DKDL210002130-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_002_DKDL210002131-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_003_DKDL210002132-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_005_DKDL210002134-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_008_DKDL210002137-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_009_DKDL210002138-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_010_DKDL210002139-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_011_DKDL210002140-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_013_DKDL210002142-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_014_DKDL210002143-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_015_DKDL210002144-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_016_DKDL210002145-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_021_DKDL210002150-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_023_DKDL210002152-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_028_DKDL210002157-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_029_DKDL210002158-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_030_DKDL210002159-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_031_DKDL210002160-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_034_DKDL210002163-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_035_DKDL210002164-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_036_DKDL210002165-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_038_DKDL210002167-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_040_DKDL210002169-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_043_DKDL210002172-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_044_DKDL210002173-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_045_DKDL210002174-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_046_DKDL210002175-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_047_DKDL210002176-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_049_DKDL210002178-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_052_DKDL210002181-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_053_DKDL210002182-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_054_DKDL210002183-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_057_DKDL210002186-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_067_DKDL210002196-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_068_DKDL210002197-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_069_DKDL210002198-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_070_DKDL210002199-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_072_DKDL210002201-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_073_DKDL210002202-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_074_DKDL210002203-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_077_DKDL210002206-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_081_DKDL210002210-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_082_DKDL210002211-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_086_DKDL210002215-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_087_DKDL210002216-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_088_DKDL210002217-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_089_DKDL210002218-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_090_DKDL210002219-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_091_DKDL210002220-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_092_DKDL210002221-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_094_DKDL210002223-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o rehh/${i}.vcf.gz
    done 

cat sample_clusters.txt | grep 'Mn' | awk '{print "-sn " $1}' | tr '\n' ' '
cat sample_clusters.txt | grep 'Mn' | awk '{print "-xl_sn " $1}' | tr '\n' ' '


cd rehh
ls *vcf.gz > vcf_list.txt

for i in $(cat vcf_list.txt)
    do  
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V $i \
        -sn PK_SB_DNA_011_DKDL210002140-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_016_DKDL210002145-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_030_DKDL210002159-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_043_DKDL210002172-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_091_DKDL210002220-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_092_DKDL210002221-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o ${i%.vcf.gz}_Mn.vcf.gz
    done

for i in $(cat vcf_list.txt)
    do  
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V $i \
        -xl_sn PK_SB_DNA_011_DKDL210002140-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_016_DKDL210002145-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_030_DKDL210002159-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_043_DKDL210002172-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_091_DKDL210002220-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_092_DKDL210002221-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o ${i%.vcf.gz}_Mf.vcf.gz
    done


Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_VS_Mn_SABAH_ONLY/rehh_Mf_VS_Mn_SABAH_ONLY.R
```



#########################################################################################################




# Comparing Sabah vs Kapit vs Betong

## Here we subset calculate the distances based on a subset of the data by only including samples from Sabah, Kapit and Betong.

```{bash,eval=F}
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses

cat Pk.fam | grep 'Sabah\|Betong\|Kapit' > include_samples.txt

plink --bfile cleaned --keep include_samples.txt --cluster --mds-plot 10 --out Pk
plink --bfile cleaned --keep include_samples.txt --pca --out Pk
plink --bfile cleaned --keep include_samples.txt --distance square --out Pk
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Kapit_Betong_Sabah/ordination_plots_Kapit_Betong_Sabah.R
```

**Did NOT progress beyond clutering and trees as it did not look interesting.**


# Comparing districts within Sabah

**Attempted to create clusters using geographic distribution of collected samples, but no patterns emerged.**


#########################################################################################################



# Exploring clustering with identity by descent using hmmIBD

Here we explore clustering using IBD instead of IBS. hmmIBD uses a  hidden Markov model to detect segments of shared ancestry in genetics sequences

## Convert VCF to correct format for [hmmIBD](https://github.com/glipsnort/hmmIBD)

Filtering
    - need to first create a list of samples to exclude - use PLINK cleaned.fam data to do an inverse fgrep of the original file list in Pk.fam
    - filter out variants based on MAF and missingess - from BCF file

hmmIBD format:
 - tab-delimited text file, with one SNP per line. 
 - First two columns are chromosome and position, followed by one sample per column. 
 - Header line, with sample names. 
 - Genotypes are coded by number: 
    - -1 for missing data
    - 0 for the first allele
    - 1 for the second, etc. 
 - SNPs and indels can be treated as equal.
 - Variants must be in chromosome and position order.
 - Variants can have between two and eight alleles.


```{bash,eval=F}
mkdir hmmIBD
awk '{print $1}' cleaned.fam > exclude_tmp1.txt
bcftools view PK_consensus_no_MOI.vcf | grep '#CHROM' | tr '\t' '\n' | sed -n '/ERR2214837/,$p' > exclude_tmp2.txt
fgrep -v -f exclude_tmp1.txt exclude_tmp2.txt > hmmIBD/exclude.txt
rm -f exclude_tmp1.txt exclude_tmp2.txt

awk '{print $2}' cleaned.bim | sed 's/:/\t/g' | awk '{print $1"\t"$2}' | uniq -u > hmmIBD/grep_patterns.tsv

Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/hmmIBD_genotype_file.R 
cd hmmIBD
export PATH=$PATH:/g/data/pq84/bin/hmmIBD/   
hmmIBD -i hmmIBD.tsv -b exclude.txt -o Pk
```

```{R,eval=F}
# Packages
library(dplyr)
library(readr)
library(stringr)
library(data.table)

# Read in and wrangle VCF

read_tsv("hmmIBD/grep_patterns.tsv", col_names = c("CHROM", "POS")) %>%
        left_join(
                read_table("PK_consensus_no_MOI.vcf", skip = 87) %>%
                rename("CHROM" = `#CHROM`) # this first section selects only variants that passed the PLINK filters we applied
        ) %>%
        mutate(CHROM = str_remove(CHROM, "ordered_PKNH_"),
        CHROM = str_remove(CHROM, "_v2")) %>%
        mutate_at(c(10:ncol(.)), ~str_remove(., ":.*")) %>% 
        mutate_at(c(10:ncol(.)), ~ifelse(. %like% "1/1" | . %like% "1/0" | . %like% "0/1" , "1", .)) %>% 
        mutate_at(c(10:ncol(.)), ~ifelse(. %like% "2/2" | . %like% "2/0" | . %like% "0/2" , "2", .)) %>% 
        mutate_at(c(10:ncol(.)), ~ifelse(. %like% "3/3" | . %like% "3/0" | . %like% "0/3" , "3", .)) %>% 
        mutate_at(c(10:ncol(.)), ~ifelse(. %like% "4/4" | . %like% "4/0" | . %like% "0/4" , "4", .)) %>% 
        mutate_at(c(10:ncol(.)), ~ifelse(. %like% "0/0" , "0", .)) %>% # 0 for reference allele
        mutate_at(c(10:ncol(.)), ~ifelse(. %like% "./." , "-1", .)) %>% # for missing 
        select(-c(3:9)) %>%
        arrange(CHROM, POS) %>%
        write_tsv("hmmIBD/hmmIBD.tsv")   
```

## Claculate pairwise statistics from hmmIBD output (thanks Edwin!)

File: 06_Analyses/hmmIBD_stats.R

This code was provided by Edwin.

```{R,eval=F}
# Packages
library(dplyr)
library(readr)
library(stringr)
library(data.table)

# Define Edwins Function
.add.label.to.IBD <-
  function(IBD, metadata, label.col, sample.col = "Sample") {
    metadata <- metadata[, c(sample.col, label.col)]
    
    IBD <- merge(
      IBD,
      metadata,
      by.x = "sample1",
      by.y = sample.col,
      all.x = TRUE,
      sort = FALSE
    )
    colnames(IBD)[length(IBD)] <- paste0(label.col, 1)
    
    IBD <- merge(
      IBD,
      metadata,
      by.x = "sample2",
      by.y = sample.col,
      all.x = TRUE,
      sort = FALSE
    )
    colnames(IBD)[length(IBD)] <- paste0(label.col, 2)
    
    IBD
  }


.create.pairwise.matrix <- function(metadata, label.col) {
  unique.labels <- sort(unique(metadata[, label.col]))
  labels.length <- length(unique.labels)
  matrix(
    nrow = labels.length,
    ncol = labels.length,
    dimnames = list(unique.labels, unique.labels)
  )
}


.populate.pairwise.matrix.with.statistic <-
  function(pairwise.matrix,
           IBD,
           label.col,
           statistic) {
    label.cols1 <- paste0(label.col, 1)
    label.cols2 <- paste0(label.col, 2)
    labels <- rownames(pairwise.matrix)
    
    for (i in 1:nrow(pairwise.matrix)) {
      label1 <- labels[i]
      
      for (j in 1:ncol(pairwise.matrix)) {
        if (i < j)
          break
        label2 <- labels[j]
        
        e11 <- IBD[, label.cols1] == label1
        e12 <- IBD[, label.cols1] == label2
        e21 <- IBD[, label.cols2] == label1
        e22 <- IBD[, label.cols2] == label2
        
        pairwise.matrix[i, j] <-
          statistic(IBD[(e11 &
                           e22) | (e12 & e21), "fract_sites_IBD"]) 
      }
    }
    
    pairwise.matrix
  }


calculate.pairwise.matrix <-
  function(IBD,
           metadata,
           label.col,
           statistic,
           sample.col = "Sample") {
    IBD <-
      .add.label.to.IBD(IBD, metadata, label.col, sample.col = sample.col)
    
    pairwise.matrix <-
      .create.pairwise.matrix(IBD, paste0(label.col, 1))
    
    .populate.pairwise.matrix.with.statistic(pairwise.matrix, IBD, label.col, statistic)
  }


save.pairwise.matrix <- function(pairwise.matrix, file) {
  write.table(pairwise.matrix,
              file,
              quote = FALSE,
              sep = "\t",
              na = "")
}


.add.sentinel.to.upper.triangle <- function(pairwise.matrix) {
  random <- rnorm(1)
  while (any(pairwise.matrix == random, na.rm = TRUE))
    random <- rnorm(1)
  
  pairwise.matrix[upper.tri(pairwise.matrix)] <- random
  
  pairwise.matrix
}


.transform.pairwise.matrix.to.table <- function(pairwise.matrix) {
  pairwise.matrix <- .add.sentinel.to.upper.triangle(pairwise.matrix)
  
  randoms <- pairwise.matrix[upper.tri(pairwise.matrix)]
  random <- randoms[1]
  if (any(is.na(randoms)) ||
      !all(randoms == random))
    warning("Matrix upper triangle does not contain sentinels.")
  
  pairwise.table <- as.data.frame(as.table(pairwise.matrix))
  pairwise.table[!(pairwise.table[, "Freq"] %in% random), ]
}


.concatenate.pairwise.labels <-
  function(pairwise.table, sep = "-") {
    labels <-
      do.call(paste, c(pairwise.table[, c("Var1", "Var2")], sep = sep))
    as.data.frame(cbind(labels, pairwise.table[, "Freq"]))
  }


calculate.pairwise.table <- function(IBD,
                                     metadata,
                                     label.col,
                                     statistics,
                                     sample.col = "Sample") {
  pairwise.tables <- NULL
  # safeguard against non-vector
  statistics <- c(statistics)
  stats <- names(statistics)
  
  
  for (i in seq_along(statistics)) {
    pairwise.matrix <- calculate.pairwise.matrix(IBD,
                                                 metadata,
                                                 label.col,
                                                 statistics[[i]],
                                                 sample.col = sample.col)
    pairwise.table <-
      .transform.pairwise.matrix.to.table(pairwise.matrix)
    
    pairwise.table <- .concatenate.pairwise.labels(pairwise.table)
    if (is.null(stats)) {
      names(pairwise.table) <- c(label.col, LETTERS[i])
    } else {
      names(pairwise.table) <- c(label.col, stats[i])
    }
    
    if (is.null(pairwise.tables)) {
      pairwise.tables <- pairwise.table
      next
    }
    
    pairwise.tables <-
      merge(pairwise.tables,
            pairwise.table,
            by = label.col,
            all = TRUE)
  }
  
  pairwise.tables
}


save.pairwise.table <- function(pairwise.table, file) {
  write.table(
    pairwise.table,
    file,
    quote = FALSE,
    sep = "\t",
    row.names = FALSE
  )
}


statistics <- c(min, function(x)
  quantile(x, probs = 0.25),
  median, mean, function(x)
    quantile(x, probs = 0.75), max)

names(statistics) <- c("Min.", "1st Qu.", "Median", "Mean",
                       "3rd Qu.", "Max.")


# Pk Clusters in Malaysia

## stat medians
metadata <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(Sample = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>% 
        as.data.frame()

### add a priori classfication of Sabah clusters based on trees

metadata <- metadata %>% # mutate it below to add the clusters based on IBS plots
    mutate(Cluster = ifelse(grepl("PK_SB_DNA_011", Sample) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", Sample) | 
                                grepl("PK_SB_DNA_043", Sample) |
                                grepl("PK_SB_DNA_016", Sample) |  
                                grepl("PK_SB_DNA_092", Sample) | 
                                grepl("PK_SB_DNA_030", Sample) | 
                                grepl("PK_SB_DNA_093", Sample) | 
                                grepl("PK_SB_DNA_042", Sample) | 
                                grepl("PK_SB_DNA_063", Sample), "Mn", .$Cluster)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Cluster = ifelse(Cluster == "Sabah", "Mf", .$Cluster)) # if its the remaining Sabah samples, make them Mn, else keep them the same

Pk.IBD <- read_table("Pk.hmm_fract.txt")
Pk.pairwise.median <- calculate.pairwise.matrix(Pk.IBD, metadata, "Cluster", median)
save.pairwise.matrix(Pk.pairwise.median, "Pk_pairwise_IBD_median.txt")

## stat summary
Pk.pairwise.summary <- calculate.pairwise.table(Pk.IBD, metadata, "Cluster", statistics)
save.pairwise.table(Pk.pairwise.summary, "Pk_pairwise_summary.txt")
```


## Plot pairwise statistics from hmmIBD output (thanks Edwin!)

File: 06_Analyses/hmmIBD_plots.R 

This code was again provided by Edwin.

```{R,eval=F}
library(igraph)

.reorder.metadata <-
  function(IBD, metadata, sample.col = "Sample") {
    samples <- unique(c(IBD[, "sample1"], IBD[, "sample2"]))
    sample.order <- match(metadata[, sample.col], samples)
    actual.length <- length(na.omit(sample.order))
    
    metadata <- metadata[order(sample.order), ]
    metadata[1:actual.length, ]
  }


.get.labels <-
  function(metadata,
           label.col,
           label.palette = NULL) {
    labels <- metadata[, label.col]
    labels[is.na(labels)] <- NaN
    
    labels <- factor(labels)
    
    if (is.null(label.palette)) {
      labels
    } else {
      factor(labels, levels = names(label.palette))
    }
  }


.generate.label.palette <-
  function(labels) {
    label.names <- levels(labels)
    label.palette <- # c("#440154FF", "#73D055FF", "#39568CFF") # for clusters
      viridis(length(label.names), option = ifelse(length(label.names) < 5, "D", "H")) # for districts
    names(label.palette) <- label.names
    
    label.palette
  }


.generate.label.colours <-
  function(labels, label.palette)
    label.palette[labels]


.create.edgelist <-
  function(IBD)
    IBD[, c("sample1", "sample2", "fract_sites_IBD")]


.create.vertices <-
  function(metadata, sample.col = "Sample") {
    vertices <- data.frame(metadata[, sample.col])
    names(vertices) <- sample.col
    
    vertices
  }


.internal.plot.IBD <- function(IBD.graph,
                               unlabelled = TRUE,
                               coords = layout_nicely,
                               percent.cutoff = NA) {
  if (unlabelled) {
    plot(
      IBD.graph,
      layout = coords,
      vertex.size = 4,
      vertex.label = NA,
      main = paste0("IBD >=", percent.cutoff, "%")
    )
  } else {
    plot(
      IBD.graph,
      layout = coords,
      vertex.size = 4,
      vertex.label.cex = 0.3,
      main = paste0("IBD >=", percent.cutoff, "%")
    )
  }
}


.plot.IBD <-
  function(file,
           IBD.graph,
           unlabelled = TRUE,
           label.palette = NULL,
           labels = NULL,
           legend.title = NA,
           coords = layout_nicely,
           percent.cutoff = NA) {
    pdf(file)
    par(mar = rep.int(1, 4) + 0.1)
    
    .internal.plot.IBD(
      IBD.graph,
      unlabelled = unlabelled,
      coords = layout_nicely,
      percent.cutoff = percent.cutoff
    )
    
    if (!is.null(label.palette))
      legend(
        "bottomleft",
        legend = levels(labels),
        fill = label.palette,
        title = legend.title,
        cex = 0.6
      )
    
    dev.off()
  }


.plot.IBDs <- function(prefix,
                       IBD.graph,
                       label.palette,
                       labels,
                       legend.title,
                       coords,
                       percent.cutoff) {
  labelled.file <-
    paste0(prefix, "_labelled_IBD", percent.cutoff, ".pdf")
  .plot.IBD(
    labelled.file,
    IBD.graph,
    unlabelled = FALSE,
    label.palette = label.palette,
    labels = labels,
    legend.title = legend.title,
    coords = coords,
    percent.cutoff = percent.cutoff
  )
  
  unlabelled.file <-
    paste0(prefix, "_unlabelled_IBD", percent.cutoff, ".pdf")
  .plot.IBD(
    unlabelled.file,
    IBD.graph,
    unlabelled = TRUE,
    label.palette = label.palette,
    labels = labels,
    legend.title = legend.title,
    coords = coords,
    percent.cutoff = percent.cutoff
  )
}


plot.IBD <- function(file,
                     IBD.cutoffs,
                     IBD,
                     metadata,
                     label.col,
                     unlabelled = TRUE,
                     label.palette = NULL,
                     legend.title = NA,
                     sample.col = "Sample") {
  edgelist <- .create.edgelist(IBD)
  
  metadata <- .reorder.metadata(IBD, metadata)
  
  vertices <- .create.vertices(metadata)
  
  sqrt.n.plots <- ceiling(sqrt(length(IBD.cutoffs)))
  size <- 7 * sqrt.n.plots
  
  legend.position <- sqrt.n.plots ^ 2 - sqrt.n.plots + 1
  legend.plot <- FALSE
  
  pdf(file, width = size, height = size)
  par(mar = rep.int(1, 4) + 0.1,
      mfrow = c(sqrt.n.plots, sqrt.n.plots))
  
  if (is.null(label.palette)) {
    labels <- .get.labels(metadata, label.col)
    label.palette <- .generate.label.palette(labels)
    
  } else {
    labels <-
      .get.labels(metadata, label.col, label.palette = label.palette)
  }
  
  label.colours <-
    .generate.label.colours(labels, label.palette)
  
  for (i in seq_along(IBD.cutoffs)) {
    d <- edgelist[edgelist[, "fract_sites_IBD"] >= IBD.cutoffs[i], ]
    
    IBD.graph <-
      graph_from_data_frame(d, directed = FALSE, vertices = vertices)
    
    coords <- layout_(IBD.graph, nicely())
    percent.cutoff <- round(IBD.cutoffs[i] * 100, digits = 2)
    
    IBD.graph <-
      set_vertex_attr(IBD.graph, "color", value = label.colours)
    
    .internal.plot.IBD(IBD.graph,
                       unlabelled,
                       coords,
                       percent.cutoff)
    
    if (i == legend.position) {
      legend(
        "bottomleft",
        legend = levels(labels),
        fill = label.palette,
        title = legend.title,
        cex = 0.6
      )
      
      legend.plot <- TRUE
    }
  }
  
  if (!legend.plot) {
    while (i != legend.position) {
      plot.new()
      i <- i + 1
    }
    
    legend(
      "bottomleft",
      legend = levels(labels),
      fill = label.palette,
      title = legend.title,
      cex = 0.6
    )
  }
  
  dev.off()
}


plot.IBDs <- function(prefixes,
                      IBD.cutoffs,
                      IBD,
                      metadata,
                      label.cols,
                      label.palettes = NULL,
                      legend.titles = NA,
                      sample.col = "Sample") {
  edgelist <- .create.edgelist(IBD)
  
  metadata <- .reorder.metadata(IBD, metadata)
  
  vertices <- .create.vertices(metadata)
  
  for (IBD.cutoff in IBD.cutoffs) {
    d <- edgelist[edgelist[, "fract_sites_IBD"] >= IBD.cutoff, ]
    
    IBD.graph <-
      graph_from_data_frame(d, directed = FALSE, vertices = vertices)
    
    coords <- layout_(IBD.graph, nicely())
    percent.cutoff <- round(IBD.cutoff * 100, digits = 2)
    
    for (label.col in label.cols) {
      label.palette <- label.palettes[[label.col]]
      
      if (is.null(label.palette)) {
        labels <- .get.labels(metadata, label.col)
        label.palette <- .generate.label.palette(labels)
        
      } else {
        labels <-
          .get.labels(metadata, label.col, label.palette = label.palette)
      }
      
      label.colours <-
        .generate.label.colours(labels, label.palette)
      
      IBD.graph <-
        set_vertex_attr(IBD.graph, "color", value = label.colours)
      
      legend.title <- legend.titles[[label.col]]
      
      prefix <- prefixes[[label.col]]
      
      .plot.IBDs(prefix,
                 IBD.graph,
                 label.palette,
                 labels,
                 legend.title,
                 coords,
                 percent.cutoff)
    }
  }
}


read.pairwise.matrix <- function(file) {
  as.matrix(read.delim(file, check.names = FALSE))
}


IBD.cutoffs <- c(0.125, 0.25, 0.5)

# include +-5% IBD
relaxed.IBD.cutoffs <- c(IBD.cutoffs, 1)
relaxed.IBD.cutoffs <- relaxed.IBD.cutoffs * 0.95


## Pk
library(igraph)
library(tidyverse)
library(MetamapsDB)
library(viridis)

Pk.IBD.file <- "Pk.hmm_fract.txt"
Pk.IBD <- read.delim(Pk.IBD.file)

metadata <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(Sample = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>% 
        as.data.frame() 

metadata <- metadata %>% 
    mutate(Cluster = ifelse(grepl("PK_SB_DNA_011", Sample) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", Sample) | 
                                grepl("PK_SB_DNA_043", Sample) |
                                grepl("PK_SB_DNA_016", Sample) |  
                                grepl("PK_SB_DNA_092", Sample) | 
                                grepl("PK_SB_DNA_030", Sample) | 
                                grepl("PK_SB_DNA_093", Sample) | 
                                grepl("PK_SB_DNA_042", Sample) | 
                                grepl("PK_SB_DNA_063", Sample), "Mn", .$Cluster)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Cluster = ifelse(Cluster == "Sabah", "Mf", .$Cluster)) # if its the remaining Sabah samples, make them Mn, else keep them the same

# add state and district info
metadata <- metadata %>%
  mutate(State = ifelse(Cluster == "Peninsular", "Peninsular", Location)) %>%
  mutate(State = ifelse(State != "Sabah" & State != "Peninsular", "Sarawak", State)) %>%
  mutate(Sample2 = Sample) %>%
  mutate(Sample = str_remove(Sample2, "_DK.*")) %>%
  left_join(
    read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters/districts_metadata.csv") %>%
    select(Sample, district)
  ) %>%
  mutate(district = ifelse(is.na(district), Location, district)) %>%
  select(-Sample) %>%
  rename(Sample = Sample2)

# Plot Fraction of sites that are IBD
IBD_meta_combined <- Pk.IBD %>%
    as.data.frame() %>%
    select(sample1, sample2, fract_sites_IBD) %>%
    left_join(
      metadata %>%
        rename(sample1 = Sample) %>%
        rename(cluster1 = Cluster) %>%
        rename(district1 = district) %>% 
        rename(state1 = State), 
      by = "sample1"
    ) %>% 
    left_join(
      metadata %>%
        rename(sample2 = Sample) %>%
        rename(cluster2 = Cluster) %>%
        rename(district2 = district) %>% 
        rename(state2 = State), 
      by = "sample2" 
    )
  
## Fraction IBD within and between clusters
IBD_fract_plot <- IBD_meta_combined %>%
  unite("Clusters", c("cluster1", "cluster2"), sep = "-") %>% 
  ggplot(aes(x = Clusters, y = fract_sites_IBD, colour = Clusters)) +
  geom_boxplot() +
  theme(legend.position = "none", 
    legend.title = element_blank(), 
    panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
  coord_flip() +
  scale_color_viridis_d() +
  ylab("Fraction of IBD sites")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD/fract_IBD_clusters.png", dpi = 300, height = 7, width = 7, IBD_fract_plot)

## Fraction IBD within and between districts
IBD_fract_plot <- IBD_meta_combined %>%
  unite("Districts", c("district1", "district2"), sep = "-") %>% 
  ggplot(aes(x = Districts, y = fract_sites_IBD, colour = Districts)) +
  geom_boxplot() +
  theme(legend.position = "none", 
    legend.title = element_blank(), 
    panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
  coord_flip() +
  scale_color_viridis_d() +
  ylab("Fraction of IBD sites")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD/fract_IBD_locations.png", height = 30, dpi = 300, IBD_fract_plot)


# Base plots
Pk.label.cols <- c("Cluster")
Pk.legend.titles <- list("Cluster" = "Cluster")
Pk.prefixes <- metadata %>% 
  select(Cluster) %>%
  unique() %>% 
  as.list()

plot.IBDs(
  Pk.prefixes,
  IBD.cutoffs,
  Pk.IBD,
  metadata,
  Pk.label.cols,
  legend.titles = Pk.legend.titles
)

Pk.median.file <- "Pk_pairwise_IBD_median.txt"
Pk.median <- read.pairwise.matrix(Pk.median.file)

Pk.IBD.cutoffs <- c(fivenum(Pk.median), relaxed.IBD.cutoffs)

IBDs.plot.file <- "Pk_major_IBDs.pdf"

plot.IBD(
  IBDs.plot.file,
  Pk.IBD.cutoffs,
  Pk.IBD,
  metadata,
  "Cluster",
  legend.title = "Cluster"
)

# Base plots - districts
Pk.label.cols <- c("district")
Pk.legend.titles <- list("district" = "district")
Pk.prefixes <- metadata %>% 
  select(district) %>%
  unique() %>% 
  as.list()

plot.IBDs(
  Pk.prefixes,
  IBD.cutoffs,
  Pk.IBD,
  metadata,
  Pk.label.cols,
  legend.titles = Pk.legend.titles
)

IBDs.plot.file <- "Major_IBDs_district.pdf"

plot.IBD(
  IBDs.plot.file,
  Pk.IBD.cutoffs,
  Pk.IBD,
  metadata,
  "district",
  legend.title = "Districts"
)


# Base plots - states - might want to change the palette for fewer levels
Pk.label.cols <- c("States")
Pk.legend.titles <- list("States" = "States")
Pk.prefixes <- metadata %>% 
  select(State) %>%
  unique() %>% 
  as.list()

plot.IBDs(
  Pk.prefixes,
  IBD.cutoffs,
  Pk.IBD,
  metadata,
  Pk.label.cols,
  legend.titles = Pk.legend.titles
)


IBDs.plot.file <- "Major_IBDs_States.pdf"

plot.IBD(
  IBDs.plot.file,
  Pk.IBD.cutoffs,
  Pk.IBD,
  metadata,
  "State",
  legend.title = "States"
)

# Identify the two outlier at 3.42% - that connect the two clusters
## Filter for state to state, calc mean and arrange in desc order
IBD_meta_combined %>%
  filter(cluster1 == "Mf" & cluster2 == "Mn") %>%
  group_by(sample1) %>%
  summarise(mean = mean(fract_sites_IBD)) %>%
  arrange(desc(mean)) %>%
  head(n=12)

IBD_meta_combined %>%
  filter(cluster1 == "Mf" & cluster2 == "Mn") %>%
  filter(fract_sites_IBD >= 0.034) %>%
  arrange(desc(fract_sites_IBD)) # the top two samples should be two in the plot that are closest to Mn
```

#########################################################################################################


# ADMIXTURE

ADMIXTURE is a clustering software that infers populations and ancestries. It requires unlinked SNPs in PLINK format, specifically the .bed (binary allelic genotype table) file as input. Here we used the "cleaned.bim" file previously created (filtered for MAF and GM).

**NB** ADMIXTURE does not accept chromosome names that are not human chromosomes and so make sure the first column is 0.

Steps:
    - identify the best K value = the number of ancestral populations you believe to be present. From the previous analyses it looks like there is 3 (Mn, Mf and Peninsular), however, you can run several and then compare.
    - collect the cv errors 
    - plot the results in R

```{bash,eval=F}
export PATH=$PATH:/g/data/pq84/bin/admixture_linux-1.3.0/
cd ADMIXTURE

for i in {3..5}
    do admixture --cv /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/cleaned.bed $i > log${i}.out
done

awk '/CV/ {print $3,$4}' *out | cut -c 4,7-20 > Pk.cv.error
```

**NB** I had some issues getting a couple of the packages up and running on Gadi and so I ran this on my local.

```{R,eval=F} 
# Packages
library(maps)
library(sf)
library(rnaturalearth)
library(tidyverse)
library(janitor)

# Bar plot of ADMIXTURE output

admixture_table <- read.table("data/ADMIXTURE/cleaned.3.Q") %>% 
  cbind(
    read_csv("data/metadata/cluster_metadata_popgen.csv")
  )

admixture_table %>% 
  pivot_longer(1:3, names_to = "Ancestry", values_to = "Proportion") %>% 
  mutate(Cluster = ifelse(Ancestry == "V1", "Mf",
                          ifelse(Ancestry == "V2", "Peninsular", "Mn"))) %>% 
  mutate(Ancestry = Proportion) %>% 
  ggplot(aes(x = Sample, y = Ancestry, fill = Cluster)) +
  geom_col() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
  scale_fill_manual(values = c("#440154FF", "#39568CFF", "#73D055FF"))

# Digging into mixed samples

admixture_table %>% 
  filter(V1 > 0.005 & V2 > 0.005 |
           V1 > 0.005 & V3 > 0.005 |
           V2 > 0.005 & V3 > 0.005) %>% 
  pivot_longer(1:3, names_to = "Ancestry", values_to = "Proportion") %>% 
  mutate(Sample = str_remove(Sample, "_DKD.*")) %>%
  mutate(Cluster = ifelse(Ancestry == "V1", "Mf",
                          ifelse(Ancestry == "V2", "Peninsular", "Mn"))) %>% 
  mutate(Ancestry = Proportion) %>% 
  ggplot(aes(x = Sample, y = Ancestry, fill = Cluster)) +
  geom_col() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_manual(values = c("#440154FF", "#39568CFF", "#73D055FF"))

# What districts are the mixed samples from?

admixture_table %>% 
  filter(V1 > 0.005 & V2 > 0.005 |
           V1 > 0.005 & V3 > 0.005 |
           V2 > 0.005 & V3 > 0.005) %>% 
  mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
  select(Sample) %>% 
  left_join(
    popgen_metadata %>% 
      select(Sample, district) %>% 
      filter(grepl("PK", Sample)) %>% 
      rbind(
        read_csv("data/metadata/Pk_clusters_metadata.csv") %>% 
          select(Sample, area) %>% 
          rename(district = area)
      )
  )

# Map with distribution of samples and clusters

map_data <- popgen_metadata %>% 
  select(Sample, district, Cluster) %>% 
  rename(Location = district) %>% 
  rbind(
    read_csv("data/metadata/cluster_metadata_popgen.csv") %>% 
      filter(!grepl("PK",Sample)) %>% 
      select(Sample, Location, Cluster)
  ) %>% 
  mutate(Location = ifelse(Location == "Sungai_Siuit", "Sungai_Siput", .$Location))

# map_data %>%  # used this to get long an lat
 # select(Location) %>% 
  #unique() %>% 
#write_csv("data/metadata/map_data.csv")
  
map_data <- map_data %>% 
  left_join(
    read_csv("data/metadata/map_data.csv")
  ) %>% 
  na.omit() %>% 
  arrange(Location) %>% 
  left_join(
    admixture_table %>% 
      select(1:4) %>% 
      mutate(Mixed = ifelse(V1 > 0.005 & V2 > 0.005 | V1 > 0.005 & V3 > 0.005 | V2 > 0.005 & V3 > 0.005, "Yes", "No")) %>% 
      mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
      select(Sample, Mixed)
  ) 

world <- ne_countries(scale = "medium", returnclass = "sf", country = c("Malaysia", "Indonesia"))

sf_use_s2(FALSE) # Spherical geometry (s2) switched off

ggplot() +
  geom_sf(data = world, col="transparent") +
  scale_x_continuous(limits = c(95, 120)) +
  coord_sf(expand = FALSE) +
  xlab("Longitude") + ylab("Latitude") +
  theme_linedraw() +
  theme_dark() +
  theme(plot.title = element_text(size=15, hjust = 0.5), plot.margin = unit(c(1,1,1,1), "cm")) +
  geom_jitter(data = map_data, aes(x = Long, y = Lat, colour = Cluster, shape = Mixed), size = 3, width = 0.4, height = 0.4) +
  scale_shape_manual(values = c(3, 17)) 

# Locations of different Sabah samples
##Do the clusters come from specific districts?

Sabah_map_data <- map_data %>% 
  filter(Long > 110 & Lat > 3.5)

ggplot() +
  geom_sf(data = world, col="transparent") +
  scale_x_continuous(limits = c(115, 120)) + 
  scale_y_continuous(limits = c(4.5, 7)) +
  coord_sf(expand = FALSE) +
  xlab("Longitude") + ylab("Latitude") +
  theme_linedraw() +
  theme_dark() +
  theme(plot.title = element_text(size = 15, hjust = 0.5), plot.margin = unit(c(1,1,1,1), "cm")) +
  geom_jitter(data = Sabah_map_data, aes(x = Long, y = Lat, colour = Cluster, shape = Mixed), size = 4, width = 0.05, height = 0.05) +
  scale_shape_manual(values = c(3, 17))


# Combining information on genetic and geographic clusters

admixture_table %>% 
  mutate(Sample = str_remove(Sample, "_DK.*")) %>% 
  left_join(
    map_data %>% 
      select(Sample, Long, Lat)
  ) %>% 
  mutate(Geo_cluster = ifelse(Long < 105, "Peninsular", 
                               ifelse(Long > 105 & Long < 112.2, "Betong", 
                                      ifelse(Long > 112.2 & Long < 115, "Kapit", "Sabah")))) %>% 
  mutate(Geo_cluster = ifelse(is.na(Geo_cluster), "Peninsular", .$Geo_cluster)) %>% 
  pivot_longer(1:3, names_to = "Ancestry", values_to = "Proportion") %>% 
  mutate(Cluster = ifelse(Ancestry == "V1", "Mf",
                          ifelse(Ancestry == "V2", "Peninsular", "Mn"))) %>% 
  mutate(Ancestry = Proportion) %>% 
  arrange(Geo_cluster) %>% 
  ggplot(aes(x = Sample, y = Ancestry, fill = Cluster)) +
    geom_col() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    scale_fill_manual(values = c("#440154FF", "#39568CFF", "#73D055FF")) +
    facet_grid(~Geo_cluster, scales = "free", space = "free")
```

#########################################################################################################

# hmmIBD within genomic clusters (Mf & Mn) and using geographic clusters - Kapit-Betong vs Sabah

Modify exclude text file for hmmIBD input to select specific genomic clusters.

```{bash.eval=F}
mkdir /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters
cp /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD/exclude.txt exclude.txt
```

**NB Geo_Clusters.csv was created on my local and contains information Longitude, Latitude and a simplified version of geographic clusters (Peninsular, Sabah, Kapit, Betong and Sarikei)**

```{R,eval=F}
library(tidyverse)

metadata <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(Sample = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>% 
        as.data.frame() 

metadata <- metadata %>% 
    mutate(Cluster = ifelse(grepl("PK_SB_DNA_011", Sample) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", Sample) | 
                                grepl("PK_SB_DNA_043", Sample) |
                                grepl("PK_SB_DNA_016", Sample) |  
                                grepl("PK_SB_DNA_092", Sample) | 
                                grepl("PK_SB_DNA_030", Sample) | 
                                grepl("PK_SB_DNA_093", Sample) | 
                                grepl("PK_SB_DNA_042", Sample) | 
                                grepl("PK_SB_DNA_063", Sample), "Mn", .$Cluster)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Cluster = ifelse(Cluster == "Sabah", "Mf", .$Cluster)) # if its the remaining Sabah samples, make them Mn, else keep them the same

metadata %>% 
  filter(Cluster != "Mf") %>%
  select(Sample) %>%
  rbind(
    read_table("exclude.txt", col_names = "Sample")
    ) %>%
  unique() %>%
  write_tsv("Mf_exclude.txt", col_names = F)

metadata %>% 
  filter(Cluster != "Mn") %>%
  select(Sample) %>%
    rbind(
    read_table("exclude.txt", col_names = "Sample")
    ) %>%
  unique() %>%
  write_tsv("Mn_exclude.txt", col_names = F)
````

```{bash.eval=F}
#!/bin/bash
#PBS -P pq84
#PBS -q normalbw 
#PBS -N hmmIBD_geo
#PBS -j oe
#PBS -m ae
#PBS -l walltime=24:00:00,mem=40GB,ncpus=5
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Set working env"
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters
export PATH=$PATH:/g/data/pq84/bin/hmmIBD/   
module load R/4.1.0 
export R_LIBS_USER="/g/data/pq84/R"

echo "---------------------------------------"
echo "Run hmmIBD on Mf"
hmmIBD -i /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD/hmmIBD.tsv -b Mf_exclude.txt -o Pk_Mf

echo "---------------------------------------"
echo "Run hmmIBD on Mn"
hmmIBD -i /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD/hmmIBD.tsv -b Mn_exclude.txt -o Pk_Mn

echo "---------------------------------------"
echo "Finished"
```


## Claculate pairwise statistics and plot hmmIBD output 

File: /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/hmmIBD_geo/hmmIBD_stats.R

```{R,eval=F}
# Packages
library(dplyr)
library(readr)
library(stringr)
library(data.table)

# Define Edwins Function
.add.label.to.IBD <-
  function(IBD, metadata, label.col, sample.col = "Sample") {
    metadata <- metadata[, c(sample.col, label.col)]
    
    IBD <- merge(
      IBD,
      metadata,
      by.x = "sample1",
      by.y = sample.col,
      all.x = TRUE,
      sort = FALSE
    )
    colnames(IBD)[length(IBD)] <- paste0(label.col, 1)
    
    IBD <- merge(
      IBD,
      metadata,
      by.x = "sample2",
      by.y = sample.col,
      all.x = TRUE,
      sort = FALSE
    )
    colnames(IBD)[length(IBD)] <- paste0(label.col, 2)
    
    IBD
  }


.create.pairwise.matrix <- function(metadata, label.col) {
  unique.labels <- sort(unique(metadata[, label.col]))
  labels.length <- length(unique.labels)
  matrix(
    nrow = labels.length,
    ncol = labels.length,
    dimnames = list(unique.labels, unique.labels)
  )
}


.populate.pairwise.matrix.with.statistic <-
  function(pairwise.matrix,
           IBD,
           label.col,
           statistic) {
    label.cols1 <- paste0(label.col, 1)
    label.cols2 <- paste0(label.col, 2)
    labels <- rownames(pairwise.matrix)
    
    for (i in 1:nrow(pairwise.matrix)) {
      label1 <- labels[i]
      
      for (j in 1:ncol(pairwise.matrix)) {
        if (i < j)
          break
        label2 <- labels[j]
        
        e11 <- IBD[, label.cols1] == label1
        e12 <- IBD[, label.cols1] == label2
        e21 <- IBD[, label.cols2] == label1
        e22 <- IBD[, label.cols2] == label2
        
        pairwise.matrix[i, j] <-
          statistic(IBD[(e11 &
                           e22) | (e12 & e21), "fract_sites_IBD"]) 
      }
    }
    
    pairwise.matrix
  }


calculate.pairwise.matrix <-
  function(IBD,
           metadata,
           label.col,
           statistic,
           sample.col = "Sample") {
    IBD <-
      .add.label.to.IBD(IBD, metadata, label.col, sample.col = sample.col)
    
    pairwise.matrix <-
      .create.pairwise.matrix(IBD, paste0(label.col, 1))
    
    .populate.pairwise.matrix.with.statistic(pairwise.matrix, IBD, label.col, statistic)
  }


save.pairwise.matrix <- function(pairwise.matrix, file) {
  write.table(pairwise.matrix,
              file,
              quote = FALSE,
              sep = "\t",
              na = "")
}


.add.sentinel.to.upper.triangle <- function(pairwise.matrix) {
  random <- rnorm(1)
  while (any(pairwise.matrix == random, na.rm = TRUE))
    random <- rnorm(1)
  
  pairwise.matrix[upper.tri(pairwise.matrix)] <- random
  
  pairwise.matrix
}


.transform.pairwise.matrix.to.table <- function(pairwise.matrix) {
  pairwise.matrix <- .add.sentinel.to.upper.triangle(pairwise.matrix)
  
  randoms <- pairwise.matrix[upper.tri(pairwise.matrix)]
  random <- randoms[1]
  if (any(is.na(randoms)) ||
      !all(randoms == random))
    warning("Matrix upper triangle does not contain sentinels.")
  
  pairwise.table <- as.data.frame(as.table(pairwise.matrix))
  pairwise.table[!(pairwise.table[, "Freq"] %in% random), ]
}


.concatenate.pairwise.labels <-
  function(pairwise.table, sep = "-") {
    labels <-
      do.call(paste, c(pairwise.table[, c("Var1", "Var2")], sep = sep))
    as.data.frame(cbind(labels, pairwise.table[, "Freq"]))
  }


calculate.pairwise.table <- function(IBD,
                                     metadata,
                                     label.col,
                                     statistics,
                                     sample.col = "Sample") {
  pairwise.tables <- NULL
  # safeguard against non-vector
  statistics <- c(statistics)
  stats <- names(statistics)
  
  
  for (i in seq_along(statistics)) {
    pairwise.matrix <- calculate.pairwise.matrix(IBD,
                                                 metadata,
                                                 label.col,
                                                 statistics[[i]],
                                                 sample.col = sample.col)
    pairwise.table <-
      .transform.pairwise.matrix.to.table(pairwise.matrix)
    
    pairwise.table <- .concatenate.pairwise.labels(pairwise.table)
    if (is.null(stats)) {
      names(pairwise.table) <- c(label.col, LETTERS[i])
    } else {
      names(pairwise.table) <- c(label.col, stats[i])
    }
    
    if (is.null(pairwise.tables)) {
      pairwise.tables <- pairwise.table
      next
    }
    
    pairwise.tables <-
      merge(pairwise.tables,
            pairwise.table,
            by = label.col,
            all = TRUE)
  }
  
  pairwise.tables
}


save.pairwise.table <- function(pairwise.table, file) {
  write.table(
    pairwise.table,
    file,
    quote = FALSE,
    sep = "\t",
    row.names = FALSE
  )
}


statistics <- c(min, function(x)
  quantile(x, probs = 0.25),
  median, mean, function(x)
    quantile(x, probs = 0.75), max)

names(statistics) <- c("Min.", "1st Qu.", "Median", "Mean",
                       "3rd Qu.", "Max.")


# Pk Clusters in Malaysia

## stat medians
metadata1 <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(Sample = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>% 
        as.data.frame()

### add a priori classfication of Sabah clusters based on trees

metadata2 <- metadata1 %>% 
    mutate(Cluster = ifelse(grepl("PK_SB_DNA_011", Sample) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", Sample) | 
                                grepl("PK_SB_DNA_043", Sample) |
                                grepl("PK_SB_DNA_016", Sample) |  
                                grepl("PK_SB_DNA_092", Sample) | 
                                grepl("PK_SB_DNA_030", Sample) | 
                                grepl("PK_SB_DNA_093", Sample) | 
                                grepl("PK_SB_DNA_042", Sample) | 
                                grepl("PK_SB_DNA_063", Sample), "Mn", .$Cluster)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Cluster = ifelse(Cluster == "Sabah", "Mf", .$Cluster)) %>% # if its the remaining Sabah samples, make them Mn, else keep them the same
    mutate(Sample = str_remove(Sample, "_DK.*")) %>%
    left_join(
      read_csv("geo_clusters.csv") %>%
        select(Sample, Geo_cluster)
    ) 

# wrangle the metadata to re-introduce the full sample names (removed in the last step to align with geo clusters) so that they align with the IBD data
metadata <- metadata2 %>%
  left_join(
    metadata1 %>%
      select(1) %>%
      rename(Sample2 = Sample) %>%
      mutate(Sample = str_remove(Sample2, "_DK.*"))
  ) %>% 
  mutate(Sample = Sample2) %>%
  select(-Sample2) 
  


# Mf
Pk.IBD <- read_table("Pk_Mf.hmm_fract.txt") 
Pk.pairwise.median <- calculate.pairwise.matrix(Pk.IBD, metadata, "Geo_cluster", median)
save.pairwise.matrix(Pk.pairwise.median, "Pk_Mf_pairwise_IBD_median.txt")

## stat summary
Pk.pairwise.summary <- calculate.pairwise.table(Pk.IBD, metadata, "Geo_cluster", statistics)
save.pairwise.table(Pk.pairwise.summary, "Pk_Mf_pairwise_summary.txt")


# Mn
Pk.IBD <- read_table("Pk_Mn.hmm_fract.txt") 
Pk.pairwise.median <- calculate.pairwise.matrix(Pk.IBD, metadata, "Geo_cluster", median)
save.pairwise.matrix(Pk.pairwise.median, "Pk_Mn_pairwise_IBD_median.txt")

## stat summary
Pk.pairwise.summary <- calculate.pairwise.table(Pk.IBD, metadata, "Geo_cluster", statistics)
save.pairwise.table(Pk.pairwise.summary, "Pk_Mn_pairwise_summary.txt")
```

Files: /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/hmmIBD_geo/hmmIBD_plots.R

```{R,eval=F}
library(igraph)
library(viridis)

.reorder.metadata <-
  function(IBD, metadata, sample.col = "Sample") {
    samples <- unique(c(IBD[, "sample1"], IBD[, "sample2"]))
    sample.order <- match(metadata[, sample.col], samples)
    actual.length <- length(na.omit(sample.order))
    
    metadata <- metadata[order(sample.order), ]
    metadata[1:actual.length, ]
  }


.get.labels <-
  function(metadata,
           label.col,
           label.palette = NULL) {
    labels <- metadata[, label.col]
    labels[is.na(labels)] <- NaN
    
    labels <- factor(labels)
    
    if (is.null(label.palette)) {
      labels
    } else {
      factor(labels, levels = names(label.palette))
    }
  }


.generate.label.palette <-
  function(labels) {
    label.names <- levels(labels)
    label.palette <-
      viridis(length(label.names), option = ifelse(length(label.names) < 5, "D", "H"))
    names(label.palette) <- label.names
    
    label.palette
  }


.generate.label.colours <-
  function(labels, label.palette)
    label.palette[labels]


.create.edgelist <-
  function(IBD)
    IBD[, c("sample1", "sample2", "fract_sites_IBD")]


.create.vertices <-
  function(metadata, sample.col = "Sample") {
    vertices <- data.frame(metadata[, sample.col])
    names(vertices) <- sample.col
    
    vertices
  }


.internal.plot.IBD <- function(IBD.graph,
                               unlabelled = TRUE,
                               coords = layout_nicely,
                               percent.cutoff = NA) {
  if (unlabelled) {
    plot(
      IBD.graph,
      layout = coords,
      vertex.size = 4,
      vertex.label = NA,
      main = paste0("IBD >=", percent.cutoff, "%")
    )
  } else {
    plot(
      IBD.graph,
      layout = coords,
      vertex.size = 4,
      vertex.label.cex = 0.3,
      main = paste0("IBD >=", percent.cutoff, "%")
    )
  }
}


.plot.IBD <-
  function(file,
           IBD.graph,
           unlabelled = TRUE,
           label.palette = NULL,
           labels = NULL,
           legend.title = NA,
           coords = layout_nicely,
           percent.cutoff = NA) {
    pdf(file)
    par(mar = rep.int(1, 4) + 0.1)
    
    .internal.plot.IBD(
      IBD.graph,
      unlabelled = unlabelled,
      coords = layout_nicely,
      percent.cutoff = percent.cutoff
    )
    
    if (!is.null(label.palette))
      legend(
        "bottomleft",
        legend = levels(labels),
        fill = label.palette,
        title = legend.title,
        cex = 0.6
      )
    
    dev.off()
  }


.plot.IBDs <- function(prefix,
                       IBD.graph,
                       label.palette,
                       labels,
                       legend.title,
                       coords,
                       percent.cutoff) {
  labelled.file <-
    paste0(prefix, "_labelled_IBD", percent.cutoff, ".pdf")
  .plot.IBD(
    labelled.file,
    IBD.graph,
    unlabelled = FALSE,
    label.palette = label.palette,
    labels = labels,
    legend.title = legend.title,
    coords = coords,
    percent.cutoff = percent.cutoff
  )
  
  unlabelled.file <-
    paste0(prefix, "_unlabelled_IBD", percent.cutoff, ".pdf")
  .plot.IBD(
    unlabelled.file,
    IBD.graph,
    unlabelled = TRUE,
    label.palette = label.palette,
    labels = labels,
    legend.title = legend.title,
    coords = coords,
    percent.cutoff = percent.cutoff
  )
}


plot.IBD <- function(file,
                     IBD.cutoffs,
                     IBD,
                     metadata,
                     label.col,
                     unlabelled = TRUE,
                     label.palette = NULL,
                     legend.title = NA,
                     sample.col = "Sample") {
  edgelist <- .create.edgelist(IBD)
  
  metadata <- .reorder.metadata(IBD, metadata)
  
  vertices <- .create.vertices(metadata)
  
  sqrt.n.plots <- ceiling(sqrt(length(IBD.cutoffs)))
  size <- 7 * sqrt.n.plots
  
  legend.position <- sqrt.n.plots ^ 2 - sqrt.n.plots + 1
  legend.plot <- FALSE
  
  pdf(file, width = size, height = size)
  par(mar = rep.int(1, 4) + 0.1,
      mfrow = c(sqrt.n.plots, sqrt.n.plots))
  
  if (is.null(label.palette)) {
    labels <- .get.labels(metadata, label.col)
    label.palette <- .generate.label.palette(labels)
    
  } else {
    labels <-
      .get.labels(metadata, label.col, label.palette = label.palette)
  }
  
  label.colours <-
    .generate.label.colours(labels, label.palette)
  
  for (i in seq_along(IBD.cutoffs)) {
    d <- edgelist[edgelist[, "fract_sites_IBD"] >= IBD.cutoffs[i], ]
    
    IBD.graph <-
      graph_from_data_frame(d, directed = FALSE, vertices = vertices)
    
    coords <- layout_(IBD.graph, nicely())
    percent.cutoff <- round(IBD.cutoffs[i] * 100, digits = 2)
    
    IBD.graph <-
      set_vertex_attr(IBD.graph, "color", value = label.colours)
    
    .internal.plot.IBD(IBD.graph,
                       unlabelled,
                       coords,
                       percent.cutoff)
    
    if (i == legend.position) {
      legend(
        "bottomleft",
        legend = levels(labels),
        fill = label.palette,
        title = legend.title,
        cex = 0.6
      )
      
      legend.plot <- TRUE
    }
  }
  
  if (!legend.plot) {
    while (i != legend.position) {
      plot.new()
      i <- i + 1
    }
    
    legend(
      "bottomleft",
      legend = levels(labels),
      fill = label.palette,
      title = legend.title,
      cex = 0.6
    )
  }
  
  dev.off()
}


plot.IBDs <- function(prefixes,
                      IBD.cutoffs,
                      IBD,
                      metadata,
                      label.cols,
                      label.palettes = NULL,
                      legend.titles = NA,
                      sample.col = "Sample") {
  edgelist <- .create.edgelist(IBD)
  
  metadata <- .reorder.metadata(IBD, metadata)
  
  vertices <- .create.vertices(metadata)
  
  for (IBD.cutoff in IBD.cutoffs) {
    d <- edgelist[edgelist[, "fract_sites_IBD"] >= IBD.cutoff, ]
    
    IBD.graph <-
      graph_from_data_frame(d, directed = FALSE, vertices = vertices)
    
    coords <- layout_(IBD.graph, nicely())
    percent.cutoff <- round(IBD.cutoff * 100, digits = 2)
    
    for (label.col in label.cols) {
      label.palette <- label.palettes[[label.col]]
      
      if (is.null(label.palette)) {
        labels <- .get.labels(metadata, label.col)
        label.palette <- .generate.label.palette(labels)
        
      } else {
        labels <-
          .get.labels(metadata, label.col, label.palette = label.palette)
      }
      
      label.colours <-
        .generate.label.colours(labels, label.palette)
      
      IBD.graph <-
        set_vertex_attr(IBD.graph, "color", value = label.colours)
      
      legend.title <- legend.titles[[label.col]]
      
      prefix <- prefixes[[label.col]]
      
      .plot.IBDs(prefix,
                 IBD.graph,
                 label.palette,
                 labels,
                 legend.title,
                 coords,
                 percent.cutoff)
    }
  }
}


read.pairwise.matrix <- function(file) {
  as.matrix(read.delim(file, check.names = FALSE))
}


IBD.cutoffs <- c(0.125, 0.25, 0.5)

# include +-5% IBD
relaxed.IBD.cutoffs <- c(IBD.cutoffs, 1)
relaxed.IBD.cutoffs <- relaxed.IBD.cutoffs * 0.95


## Pk
library(igraph)
library(tidyverse)
library(MetamapsDB)

metadata1 <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(Sample = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>% 
        as.data.frame()

### add a priori classfication of Sabah clusters based on trees

metadata2 <- metadata1 %>% 
    mutate(Cluster = ifelse(grepl("PK_SB_DNA_011", Sample) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", Sample) | 
                                grepl("PK_SB_DNA_043", Sample) |
                                grepl("PK_SB_DNA_016", Sample) |  
                                grepl("PK_SB_DNA_092", Sample) | 
                                grepl("PK_SB_DNA_030", Sample) | 
                                grepl("PK_SB_DNA_093", Sample) | 
                                grepl("PK_SB_DNA_042", Sample) | 
                                grepl("PK_SB_DNA_063", Sample), "Mn", .$Cluster)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Cluster = ifelse(Cluster == "Sabah", "Mf", .$Cluster)) %>% # if its the remaining Sabah samples, make them Mn, else keep them the same
    mutate(Sample = str_remove(Sample, "_DK.*")) %>%
    left_join(
      read_csv("geo_clusters.csv") %>%
        select(Sample, Geo_cluster)
    ) 

# wrangle the metadata to re-introduce the full sample names (removed in the last step to align with geo clusters) so that they align with the IBD data
metadata <- metadata2 %>%
  left_join(
    metadata1 %>%
      select(1) %>%
      rename(Sample2 = Sample) %>%
      mutate(Sample = str_remove(Sample2, "_DK.*"))
  ) %>% 
  mutate(State= ifelse(Geo_cluster != "Sabah" & Geo_cluster != "Peninsular", "Sarawak", Geo_cluster)) %>% 
  left_join(
    read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters/districts_metadata.csv") %>%
    select(Sample, district)
  ) %>% 
  mutate(district = ifelse(is.na(district), Location, district)) %>%
  mutate(Sample = Sample2) %>%
  select(Sample, Cluster, State, district) 

#Mf
Pk.IBD.file <- "Pk_Mf.hmm_fract.txt"
Pk.IBD <- read.delim(Pk.IBD.file) 


# Plot Fraction of sites that are IBD
IBD_meta_combined <- Pk.IBD %>%
    as.data.frame() %>%
    select(sample1, sample2, fract_sites_IBD) %>%
    left_join(
      metadata %>%
        rename(sample1 = Sample) %>%
        rename(cluster1 = Cluster) %>%
        rename(district1 = district) %>% 
        rename(state1 = State), 
      by = "sample1"
    ) %>% 
    left_join(
      metadata %>%
        rename(sample2 = Sample) %>%
        rename(cluster2 = Cluster) %>%
        rename(district2 = district) %>% 
        rename(state2 = State), 
      by = "sample2" 
    )

## Fraction IBD within and between districts
IBD_fract_plot <- IBD_meta_combined %>%
  unite("Districts", c("district1", "district2"), sep = "-") %>% 
  ggplot(aes(x = Districts, y = fract_sites_IBD, colour = Districts)) +
  geom_boxplot() +
  theme(legend.position = "none", 
    legend.title = element_blank(), 
    panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
  coord_flip() +
  scale_color_viridis_d() +
  ylab("Fraction of IBD sites")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters/Mf/fract_IBD_districts_Mf.png", height = 12, dpi = 300, IBD_fract_plot)

## Fraction IBD within and between states
IBD_fract_plot <- IBD_meta_combined %>%
  unite("States", c("state1", "state2"), sep = "-") %>% 
  ggplot(aes(x = States, y = fract_sites_IBD, colour = States)) +
  geom_boxplot() +
  theme(legend.position = "none", 
    legend.title = element_blank(), 
    panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
  coord_flip() +
  scale_color_viridis_d() +
  ylab("Fraction of IBD sites")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters/Mf/fract_IBD_States_Mf.png", dpi = 300, IBD_fract_plot)


# Base plots - districts
Pk.label.cols <- c("district")
Pk.legend.titles <- list("district" = "district")
Pk.prefixes <- metadata %>% 
  select(district) %>%
  unique() %>% 
  as.list()

plot.IBDs(
  Pk.prefixes,
  IBD.cutoffs,
  Pk.IBD,
  metadata,
  Pk.label.cols,
  legend.titles = Pk.legend.titles
)

Pk.median.file <- "Pk_Mf_pairwise_IBD_median_district.txt"
Pk.median <- read.pairwise.matrix(Pk.median.file)

Pk.IBD.cutoffs <- c(fivenum(Pk.median), relaxed.IBD.cutoffs)

IBDs.plot.file <- "Mf_major_IBDs_district.pdf"

plot.IBD(
  IBDs.plot.file,
  Pk.IBD.cutoffs,
  Pk.IBD,
  metadata,
  "district",
  legend.title = "Districts within Mf"
)


# Base plots - states - might want to change the palette for fewer levels
Pk.label.cols <- c("States")
Pk.legend.titles <- list("States" = "States")
Pk.prefixes <- metadata %>% 
  select(State) %>%
  unique() %>% 
  as.list()

plot.IBDs(
  Pk.prefixes,
  IBD.cutoffs,
  Pk.IBD,
  metadata,
  Pk.label.cols,
  legend.titles = Pk.legend.titles
)

Pk.median.file <- "Pk_Mf_pairwise_IBD_median_states.txt"
Pk.median <- read.pairwise.matrix(Pk.median.file)

Pk.IBD.cutoffs <- c(fivenum(Pk.median), relaxed.IBD.cutoffs)

IBDs.plot.file <- "Mf_major_IBDs_States.pdf"

plot.IBD(
  IBDs.plot.file,
  Pk.IBD.cutoffs,
  Pk.IBD,
  metadata,
  "State",
  legend.title = "States within Mf"
)


# Identify the two outlier Sabah outlier samples
## Filter for state to state, calc mean and arrange in desc order
IBD_meta_combined %>%
  filter(state1 == "Sarawak" & state2 == "Sabah") %>%
  group_by(sample2) %>%
  summarise(mean = mean(fract_sites_IBD)) %>%
  arrange(desc(mean))

IBD_meta_combined %>%
  filter(state1 == "Sarawak" & state2 == "Sabah") %>%
  filter(sample2 == "PK_SB_DNA_028_DKDL210002157-1a_HWHGKDSXY_L4") %>%
  arrange(desc(fract_sites_IBD)) %>%
  head(n=12) %>%
  select(sample1, sample2, fract_sites_IBD)

IBD_meta_combined %>%
  filter(state1 == "Sarawak" & state2 == "Sabah") %>%
  filter(sample2 == "PK_SB_DNA_053_DKDL210002182-1a_HWHGKDSXY_L4") %>%
  arrange(desc(fract_sites_IBD)) %>%
  head(n=12) %>%
  select(sample1, sample2, fract_sites_IBD)

# Where are the regions of IBD in the connected samples? Is it the same as the introgressed regions you've identified?
Pk.IBD_full <- read.delim("Pk_Mf.hmm.txt")

Pk.IBD_full %>%
  filter(sample2 == "PK_SB_DNA_028_DKDL210002157-1a_HWHGKDSXY_L4" | sample2 == "PK_SB_DNA_053_DKDL210002182-1a_HWHGKDSXY_L4") %>% 
  filter(sample1 == "ERR985378" |
    sample2 == "ERR274221" |
    sample2 == "ERR985373" |
    sample2 == "ERR985375" |
    sample2 == "ERR985384" | 
    sample2 == "ERR985372" |
    sample2 == "ERR985379" | 
    sample2 == "ERR985374" |
    sample2 == "ERR985381" |
    sample2 == "ERR985377" |
    sample2 == "ERR274222" |
    sample2 == "ERR985380") %>% 
    filter(different == 0) %>% 
    unite("region", c("start", "end"), sep = "-") %>% write_tsv("shared_regions_for_two_samples.tsv")






#Mn - change wd
Pk.IBD.file <- "Pk_Mn.hmm_fract.txt"

# Plot Fraction of sites that are IBD
IBD_meta_combined <- Pk.IBD %>%
    as.data.frame() %>%
    select(sample1, sample2, fract_sites_IBD) %>%
    left_join(
      metadata %>%
        rename(sample1 = Sample) %>%
        rename(cluster1 = Cluster) %>%
        rename(district1 = district) %>% 
        rename(state1 = State), 
      by = "sample1"
    ) %>% 
    left_join(
      metadata %>%
        rename(sample2 = Sample) %>%
        rename(cluster2 = Cluster) %>%
        rename(district2 = district) %>% 
        rename(state2 = State), 
      by = "sample2" 
    )

## Fraction IBD within and between districts
IBD_fract_plot <- IBD_meta_combined %>%
  unite("Districts", c("district1", "district2"), sep = "-") %>% 
  ggplot(aes(x = Districts, y = fract_sites_IBD, colour = Districts)) +
  geom_boxplot() +
  theme(legend.position = "none", 
    legend.title = element_blank(), 
    panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
  coord_flip() +
  scale_color_viridis_d() +
  ylab("Fraction of IBD sites")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters/Mn/fract_IBD_districts_Mn.png", dpi = 300, height = 12, IBD_fract_plot)

IBD_fract_plot <- IBD_meta_combined %>%
  unite("States", c("state1", "state2"), sep = "-") %>% 
  ggplot(aes(x = States, y = fract_sites_IBD, colour = States)) +
  geom_boxplot() +
  theme(legend.position = "none", 
    legend.title = element_blank(), 
    panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
  coord_flip() +
  scale_color_viridis_d() +
  ylab("Fraction of IBD sites")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters/Mn/fract_IBD_states_Mn.png", dpi = 300, IBD_fract_plot)


# Base plots - districts
Pk.label.cols <- c("district")
Pk.legend.titles <- list("district" = "district")
Pk.prefixes <- metadata %>% 
  select(district) %>%
  unique() %>% 
  as.list()

plot.IBDs(
  Pk.prefixes,
  IBD.cutoffs,
  Pk.IBD,
  metadata,
  Pk.label.cols,
  legend.titles = Pk.legend.titles
)

Pk.median.file <- "Pk_Mn_pairwise_IBD_median_district.txt"
Pk.median <- read.pairwise.matrix(Pk.median.file)

Pk.IBD.cutoffs <- c(fivenum(Pk.median), relaxed.IBD.cutoffs)

IBDs.plot.file <- "Mn_major_IBDs_district.pdf"

plot.IBD(
  IBDs.plot.file,
  Pk.IBD.cutoffs,
  Pk.IBD,
  metadata,
  "district",
  legend.title = "Districts within Mn"
)


# Base plots - states
Pk.label.cols <- c("States")
Pk.legend.titles <- list("States" = "States")
Pk.prefixes <- metadata %>% 
  select(State) %>%
  unique() %>% 
  as.list()

plot.IBDs(
  Pk.prefixes,
  IBD.cutoffs,
  Pk.IBD,
  metadata,
  Pk.label.cols,
  legend.titles = Pk.legend.titles
)

Pk.median.file <- "Pk_Mn_pairwise_IBD_median_states.txt"
Pk.median <- read.pairwise.matrix(Pk.median.file)

Pk.IBD.cutoffs <- c(fivenum(Pk.median), relaxed.IBD.cutoffs)

IBDs.plot.file <- "Mn_major_IBDs_States.pdf"

plot.IBD(
  IBDs.plot.file,
  Pk.IBD.cutoffs,
  Pk.IBD,
  metadata,
  "State",
  legend.title = "States within Mn"
)
```

#########################################################################################################

# Differences between the Betong-Kapit and Sabah clusters within the Mf and Mn genomic clusters

The IBD analysis done above suggests that Kapit and Betong can be grouped together and comapred to Sabah (within clusters) to identify genomic regions that are both driving the differences between the geogrpahic clusters (Fst) and that are under different selection pressures (rehh - iHS etc.).


## Mf

### Here we calculate the distances based on a subset of the data by only including Mf samples.

```{bash,eval=F}
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_Geo/sample_clusters.R
sed 's/,/ /g' sample_clusters.csv > sample_clusters.txt
rm -f sample_clusters.csv
cat sample_clusters.txt | awk '{print $1 " " $2 " 0 0 0 " $3}' > include_samples.txt

plink --bfile cleaned --keep include_samples.txt --cluster --mds-plot 10 --out Pk
plink --bfile cleaned --keep include_samples.txt --pca --out Pk
plink --bfile cleaned --keep include_samples.txt --distance square --out Pk
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_Geo/ordination_plots.R
```

### Fst statistic - within Mf

First create a file that contains the clusters you want used for the Fst. You can ammend the Pk.fam file (or metadata), but the file needs to be structured in PLINKs weird FID-IID-Cluster format.

Steps:
    - Apply filter (as done previously)
    - Attach metadata to fam file (as done previously)
    - Define clusters using metadata
    - Calculate and plot Fst for a-priori clusters (sample_clusters.txt)

```{bash,eval=F}
plink --file Pk --maf 0.05 --geno 0.25 --mind 0.25 --make-bed --out cleaned
sed 's/ /,/g' cleaned.fam > Pk.csv
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/pheno_to_plink_fam.R
sed 's/,/ /g' Pk.csv > Pk.fam

plink --bfile cleaned \
  --within sample_clusters.txt \
  --keep include_samples.txt \
  --fst \
  --allow-no-sex \
  --out Pk

Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_Geo/Fst_plot.R
```

### rehh - within Mf

# Variant selection - between geo-clsuters within Mf

**Need to ammend the sn and xl line manually based on outputs of "cat"**

```{bash}
grep '##contig=<ID=ordered' PK_consensus_no_MOI.vcf| sed 's/##contig=<ID=//g' | sed 's/,.*//g' > rehh_geo/chromosome.txt

cat sample_clusters.txt | awk '{print "-sn " $1}' | tr '\n' ' '

for i in $(cat rehh/chromosome.txt)
    do 
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V PK_consensus_no_MOI.vcf \
        -L $i \
        -sn ERR274221 -sn ERR274222 -sn ERR366426 -sn ERR985372 -sn ERR985373 -sn ERR985374 -sn ERR985375 -sn ERR985376 -sn ERR985377 -sn ERR985378 -sn ERR985379 -sn ERR985380 -sn ERR985381 -sn ERR985382 -sn ERR985383 -sn ERR985384 -sn ERR985385 -sn ERR985387 -sn ERR985388 -sn ERR985389 -sn ERR985390 -sn ERR985391 -sn ERR985392 -sn ERR985393 -sn ERR985394 -sn ERR985398 -sn ERR985399 -sn ERR985400 -sn ERR985401 -sn ERR985402 -sn ERR985403 -sn ERR985404 -sn ERR985406 -sn ERR985407 -sn ERR985408 -sn ERR985409 -sn PK_SB_DNA_001_DKDL210002130-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_002_DKDL210002131-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_003_DKDL210002132-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_005_DKDL210002134-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_008_DKDL210002137-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_009_DKDL210002138-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_010_DKDL210002139-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_013_DKDL210002142-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_014_DKDL210002143-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_015_DKDL210002144-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_021_DKDL210002150-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_023_DKDL210002152-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_028_DKDL210002157-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_029_DKDL210002158-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_031_DKDL210002160-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_034_DKDL210002163-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_035_DKDL210002164-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_036_DKDL210002165-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_038_DKDL210002167-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_040_DKDL210002169-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_044_DKDL210002173-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_045_DKDL210002174-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_046_DKDL210002175-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_047_DKDL210002176-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_049_DKDL210002178-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_052_DKDL210002181-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_053_DKDL210002182-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_054_DKDL210002183-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_057_DKDL210002186-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_067_DKDL210002196-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_068_DKDL210002197-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_069_DKDL210002198-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_070_DKDL210002199-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_072_DKDL210002201-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_073_DKDL210002202-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_074_DKDL210002203-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_077_DKDL210002206-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_081_DKDL210002210-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_082_DKDL210002211-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_086_DKDL210002215-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_087_DKDL210002216-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_088_DKDL210002217-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_089_DKDL210002218-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_090_DKDL210002219-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_094_DKDL210002223-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o rehh_geo/${i}.vcf.gz
    done 

cat sample_clusters.txt | grep 'Sabah' | awk '{print "-sn " $1}' | tr '\n' ' '
cat sample_clusters.txt | grep 'Sabah' | awk '{print "-xl_sn " $1}' | tr '\n' ' '

cd rehh_geo
mkdir Mf_plots
mkdir Mn_plots
ls *vcf.gz > vcf_list.txt

for i in $(cat vcf_list.txt)
    do  
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V $i \
        -sn PK_SB_DNA_001_DKDL210002130-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_002_DKDL210002131-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_003_DKDL210002132-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_005_DKDL210002134-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_008_DKDL210002137-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_009_DKDL210002138-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_010_DKDL210002139-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_013_DKDL210002142-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_014_DKDL210002143-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_015_DKDL210002144-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_021_DKDL210002150-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_023_DKDL210002152-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_028_DKDL210002157-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_029_DKDL210002158-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_031_DKDL210002160-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_034_DKDL210002163-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_035_DKDL210002164-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_036_DKDL210002165-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_038_DKDL210002167-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_040_DKDL210002169-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_044_DKDL210002173-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_045_DKDL210002174-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_046_DKDL210002175-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_047_DKDL210002176-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_049_DKDL210002178-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_052_DKDL210002181-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_053_DKDL210002182-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_054_DKDL210002183-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_057_DKDL210002186-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_067_DKDL210002196-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_068_DKDL210002197-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_069_DKDL210002198-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_070_DKDL210002199-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_072_DKDL210002201-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_073_DKDL210002202-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_074_DKDL210002203-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_077_DKDL210002206-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_081_DKDL210002210-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_082_DKDL210002211-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_086_DKDL210002215-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_087_DKDL210002216-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_088_DKDL210002217-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_089_DKDL210002218-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_090_DKDL210002219-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_094_DKDL210002223-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o ${i%.vcf.gz}_Sabah.vcf.gz
    done

for i in $(cat vcf_list.txt)
    do  
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V $i \
        -xl_sn PK_SB_DNA_001_DKDL210002130-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_002_DKDL210002131-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_003_DKDL210002132-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_005_DKDL210002134-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_008_DKDL210002137-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_009_DKDL210002138-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_010_DKDL210002139-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_013_DKDL210002142-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_014_DKDL210002143-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_015_DKDL210002144-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_021_DKDL210002150-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_023_DKDL210002152-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_028_DKDL210002157-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_029_DKDL210002158-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_031_DKDL210002160-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_034_DKDL210002163-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_035_DKDL210002164-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_036_DKDL210002165-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_038_DKDL210002167-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_040_DKDL210002169-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_044_DKDL210002173-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_045_DKDL210002174-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_046_DKDL210002175-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_047_DKDL210002176-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_049_DKDL210002178-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_052_DKDL210002181-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_053_DKDL210002182-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_054_DKDL210002183-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_057_DKDL210002186-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_067_DKDL210002196-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_068_DKDL210002197-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_069_DKDL210002198-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_070_DKDL210002199-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_072_DKDL210002201-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_073_DKDL210002202-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_074_DKDL210002203-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_077_DKDL210002206-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_081_DKDL210002210-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_082_DKDL210002211-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_086_DKDL210002215-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_087_DKDL210002216-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_088_DKDL210002217-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_089_DKDL210002218-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_090_DKDL210002219-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_094_DKDL210002223-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o ${i%.vcf.gz}_Betong-Kapit-Sarikei.vcf.gz
    done


Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mf_Geo/rehh.R
mv *vcf.gz* Mf_plots/
```


## Mn

### Here we calculate the distances based on a subset of the data by only including Mn samples.

```{bash,eval=F}
cdd 1
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mn_Geo/sample_clusters.R
sed 's/,/ /g' sample_clusters.csv > sample_clusters.txt
rm -f sample_clusters.csv
cat sample_clusters.txt | awk '{print $1 " " $2 " 0 0 0 " $3}' > include_samples.txt

plink --bfile cleaned --keep include_samples.txt --cluster --mds-plot 10 --out Pk
plink --bfile cleaned --keep include_samples.txt --pca --out Pk
plink --bfile cleaned --keep include_samples.txt --distance square --out Pk
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mn_Geo/ordination_plots.R
```

### Fst statistic - within Mn

```{bash,eval=F}
plink --file Pk --maf 0.05 --geno 0.25 --mind 0.25 --make-bed --out cleaned
sed 's/ /,/g' cleaned.fam > Pk.csv
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/pheno_to_plink_fam.R
sed 's/,/ /g' Pk.csv > Pk.fam

plink --bfile cleaned \
  --within sample_clusters.txt \
  --keep include_samples.txt \
  --fst \
  --allow-no-sex \
  --out Pk

Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mn_Geo/Fst_plot.R
```

### rehh - within Mn

# Variant selection - between geo-clsuters within Mn

**Need to ammend the sn and xl line manually based on outputs of "cat"**

```{bash}
grep '##contig=<ID=ordered' PK_consensus_no_MOI.vcf| sed 's/##contig=<ID=//g' | sed 's/,.*//g' > rehh_geo/chromosome.txt

cat sample_clusters.txt | awk '{print "-sn " $1}' | tr '\n' ' '

for i in $(cat rehh/chromosome.txt)
    do 
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V PK_consensus_no_MOI.vcf \
        -L $i \
        -sn ERR2214837 -sn ERR2214838 -sn ERR2214839 -sn ERR2214840 -sn ERR2214841 -sn ERR2214842 -sn ERR2214843 -sn ERR2214844 -sn ERR2214845 -sn ERR2214846 -sn ERR2214847 -sn ERR2214848 -sn ERR2214849 -sn ERR2214851 -sn ERR2214852 -sn ERR2214853 -sn ERR2214854 -sn ERR2214855 -sn ERR2214857 -sn ERR274224 -sn ERR274225 -sn ERR366425 -sn ERR985411 -sn ERR985412 -sn ERR985413 -sn ERR985415 -sn ERR985416 -sn ERR985418 -sn ERR985419 -sn PK_SB_DNA_011_DKDL210002140-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_016_DKDL210002145-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_030_DKDL210002159-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_043_DKDL210002172-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_091_DKDL210002220-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_092_DKDL210002221-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o rehh_geo/${i}.vcf.gz
    done 

cat sample_clusters.txt | grep 'Sabah' | awk '{print "-sn " $1}' | tr '\n' ' '
cat sample_clusters.txt | grep 'Sabah' | awk '{print "-xl_sn " $1}' | tr '\n' ' '

cd rehh_geo
ls *vcf.gz > vcf_list.txt

for i in $(cat vcf_list.txt)
    do  
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V $i \
        -sn PK_SB_DNA_011_DKDL210002140-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_016_DKDL210002145-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_030_DKDL210002159-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_043_DKDL210002172-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_091_DKDL210002220-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_092_DKDL210002221-1a_HWHGKDSXY_L4 -sn PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o ${i%.vcf.gz}_Sabah.vcf.gz
    done

for i in $(cat vcf_list.txt)
    do  
        java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
        -T SelectVariants \
        -R $FASTA/strain_A1_H.1.Icor.fasta \
        -V $i \
        -xl_sn PK_SB_DNA_011_DKDL210002140-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_016_DKDL210002145-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_030_DKDL210002159-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_042_DKDL210002171-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_043_DKDL210002172-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_063_DKDL210002192-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_091_DKDL210002220-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_092_DKDL210002221-1a_HWHGKDSXY_L4 -xl_sn PK_SB_DNA_093_DKDL210002222-1a_HWHGKDSXY_L4 \
        -XL $REGIONS_TO_MASK \
        -o ${i%.vcf.gz}_Betong-Kapit-Sarikei.vcf.gz
    done


Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/Mn_Geo/rehh.R

mv *vcf.gz* Mn_plots/
```


#########################################################################################################

# hmmIBD with variants subset to those that differ between geographic clusters 

Modify exclude text file for hmmIBD input to select specific genomic clusters.

```{bash.eval=F}
mkdir /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters
cp /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD/exclude.txt exclude.txt
```

## Need to first subset the hmmIBD input file

```{bash,eval=F}
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD/
```

```{R,eval=F}
library(tidyverse)

# Metadata
metadata <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(Sample = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>% 
        as.data.frame() 

metadata <- metadata %>% 
    mutate(Cluster = ifelse(grepl("PK_SB_DNA_011", Sample) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", Sample) | 
                                grepl("PK_SB_DNA_043", Sample) |
                                grepl("PK_SB_DNA_016", Sample) |  
                                grepl("PK_SB_DNA_092", Sample) | 
                                grepl("PK_SB_DNA_030", Sample) | 
                                grepl("PK_SB_DNA_093", Sample) | 
                                grepl("PK_SB_DNA_042", Sample) | 
                                grepl("PK_SB_DNA_063", Sample), "Mn", .$Cluster)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Cluster = ifelse(Cluster == "Sabah", "Mf", .$Cluster))  # if its the remaining Sabah samples, make them Mn, else keep them the same

metadata_Mn <- metadata %>%
  filter(Cluster == "Mn")

metadata_Mf <- metadata %>%
  filter(Cluster == "Mf") 

IBD <- read_tsv("hmmIBD.tsv") 

IBD %>% 
  select(1,2, metadata_Mn$Sample) %>% # Mn
  filter(!if_all(3:ncol(.), ~ . == -1)) %>%
  filter(!if_all(3:ncol(.), ~ . == 0)) %>%
  filter(!if_all(3:ncol(.), ~ . == 1)) %>%
  filter(!if_all(3:ncol(.), ~ . == 2)) %>%
  filter(!if_all(3:ncol(.), ~ . == 3)) %>%
  left_join(
    IBD %>% 
      select(1,2, metadata_Mf$Sample) %>% # Mf
      filter(!if_all(3:ncol(.), ~ . == -1)) %>%
      filter(!if_all(3:ncol(.), ~ . == 0)) %>%
      filter(!if_all(3:ncol(.), ~ . == 1)) %>%
      filter(!if_all(3:ncol(.), ~ . == 2)) %>%
      filter(!if_all(3:ncol(.), ~ . == 3))
    ) %>%
    na.omit() %>%
    write_tsv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters/hmmIBD_subset.tsv")
```

```{bash.eval=F}
#!/bin/bash
#PBS -P pq84
#PBS -q normalbw 
#PBS -N hmmIBD_geo
#PBS -j oe
#PBS -m ae
#PBS -l walltime=24:00:00,mem=40GB,ncpus=5
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Set working env"
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD_geo_clusters
export PATH=$PATH:/g/data/pq84/bin/hmmIBD/   
module load R/4.1.0 
export R_LIBS_USER="/g/data/pq84/R"

echo "---------------------------------------"
echo "Run hmmIBD on Mf"
hmmIBD -i hmmIBD_subset.tsv -b Mf_exclude.txt -o SNP_subset/Pk_Mf

echo "---------------------------------------"
echo "Run hmmIBD on Mn"
hmmIBD -i hmmIBD_subset.tsv -b Mn_exclude.txt -o SNP_subset/Pk_Mn

echo "---------------------------------------"
echo "Finished"
```

#########################################################################################################

# Search for drug resistance markers (orthologues) from other Plasmodium species

Using this [list](https://www.nature.com/articles/s41467-021-23422-3/tables/2) of Pv orthologues compiled by Diaz et al, we use a genotype file (subset) previously created for hmmIBD and add annotations inforamtion. By adding the metadata we can identify differences in SNP frequency between different clusters.


```{bash,eval=F}
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/hmmIBD
Rscript  /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/drug_resistance/drug_resistance_freq.R
```
**NB. More subsetting options are available in the script /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/drug_resistance_freq.R**

```{R,eval=F}
# Packages
library(tidyverse)
library(janitor)
library(dplyr)
library(readr)
library(stringr)
library(data.table)

# Read in genotype file previously created and subset to orthologue genes
genotype_file <- read_tsv("hmmIBD.tsv")
orthologue_genes <- genotype_file %>% 
  filter(CHROM == "01" & POS > 381348 & POS < 384824) %>%
  rbind(genotype_file %>% filter(CHROM == "05" & POS > 445338 & POS < 447218)) %>%
  rbind(genotype_file %>% filter(CHROM == "08" & POS > 1408170 & POS < 1411360)) %>%
  rbind(genotype_file %>% filter(CHROM == "10" & POS > 448974 & POS < 453368)) %>%
  rbind(genotype_file %>% filter(CHROM == "13" & POS > 2302686 & POS < 2304038)) %>%
  rbind(genotype_file %>% filter(CHROM == "14" & POS > 1308259 & POS < 1310954)) %>%
  rbind(genotype_file %>% filter(CHROM == "14" & POS > 2116423 & POS < 2122473))

# Read in annotation text file, left join to subset genotype file & subset to non-synomynous
annotation <- read_table("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/PK_consensus_filtered_annotation.txt", skip = 82) %>%
  rename(Variation = `#Uploaded_variation`) %>%
  separate(Variation, sep = "_", c("ordered", "PKNH", "CHROM", "v2", "POS", "Variant")) %>%
  select(-c(ordered, PKNH, v2, Allele, Location, Variant, Feature, Feature_type, cDNA_position, CDS_position, Codons, Existing_variation)) %>%
  mutate(POS = as.numeric(POS))

# Left join annotation & genotype file & filter out synomynous variants 
annotated_orthologue_genes <- orthologue_genes %>%
  left_join(annotation) %>%
  filter(!grepl("synonymous_variant", Consequence)) %>%
  filter(!grepl("downstream_gene_variant", Consequence)) %>%
  filter(!grepl("upstream_gene_variant", Consequence)) %>%
  filter(!grepl("intron_variant", Consequence)) %>%
  filter(!grepl("stop_retained_variant", Consequence))

annotated_orthologue_genes %>%
  filter(Gene == "PKA1H_140035100") %>% 
  pivot_longer(3:187, names_to = "Sample", values_to = "SNP") %>% 
  select(POS, Consequence, SNP, Sample) %>%
  group_by(POS, Consequence, Sample) %>% 
  summarise(SNP = mean(SNP)) %>%
  group_by(POS, Consequence) %>% 
  summarise(total_snp_prop = sum(SNP == 1)/nrow(annotated_orthologue_genes %>% 
    filter(Gene == "PKA1H_140035100") %>% 
    pivot_longer(3:187, names_to = "Sample", values_to = "SNP") %>%
    select(Sample) %>% 
    unique())) #%>% write_tsv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/drug_resistance/prop_pvdhps_snps.tsv")

# double check dominant
annotated_orthologue_genes %>%
  filter(Gene == "PKA1H_140035100") %>% 
  pivot_longer(3:187, names_to = "Sample", values_to = "SNP") %>% 
  select(POS, Consequence, SNP, Sample) %>% 
  filter(POS == 1309752) %>% 
  group_by(Sample) %>% 
  summarise(mean = mean(SNP)) %>% 
  summarise(sum(mean == 1)/nrow(.))

# calculate for clusters
annotated_orthologue_genes %>%
  filter(Gene == "PKA1H_140035100") %>% 
  pivot_longer(3:187, names_to = "Sample", values_to = "SNP") %>% 
  select(POS, Consequence, SNP, Sample) %>%
  filter(POS == 1309410) %>% # overlap with Pv
  group_by(Sample) %>% 
  summarise(mean = mean(SNP)) %>%
  mutate(Sample = str_remove(Sample, "_DK.*")) %>%
  left_join(Metadata) %>% 
  group_by(Cluster) %>%
  summarise(clust_prop = sum(mean == 1)/n()*100)

annotated_orthologue_genes %>%
  filter(Gene == "PKA1H_140035100") %>% 
  pivot_longer(3:187, names_to = "Sample", values_to = "SNP") %>% 
  group_by(Sample) %>%
  summarise(total_snp = sum(SNP == 1)) %>%
  summary(total_snp)

annotated_orthologue_genes %>%
  filter(Gene == "PKA1H_050015200") %>% 
  pivot_longer(3:187, names_to = "Sample", values_to = "SNP") %>% 
  group_by(Sample) %>%
  summarise(mean = mean(SNP)) 

# Export Pk variants
annotated_orthologue_genes %>% 
  mutate(CHROM = as.numeric(CHROM)) %>%
  arrange(CHROM, POS) %>%
  select(1:2, 188:ncol(.)) %>%
  write_tsv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/drug_resistance/orthologues_in_Pk.tsv") 


# plot

Metadata <- readxl::read_xlsx("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/PK_Sabah_Sample_naming_indexes.xlsx") %>% 
      select(1) %>%
      add_column(Location = "Sabah") %>%
      rename(Sample = sampleid) %>%
      add_column(Cluster = "Sabah") %>% 
      rbind(
        read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_metadata.csv") %>%
        select(1, 3, 5) %>% 
        rename(Location = area) %>%
        mutate(Cluster = str_remove(Group, "-Pk")) %>% 
        select(-Group) 
      ) %>%
      rbind(
        read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_peninsular_metadata.csv") %>%
        select(2, 8) %>%
        rename(Sample = ENA_accession_no_ES) %>%
        add_column(Cluster = "Peninsular") 
      ) %>%
    mutate(Location = str_replace(Location, " ", "_")) %>%
    mutate(Cluster = ifelse(grepl("PK_SB_DNA_011", Sample) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", Sample) | 
                                grepl("PK_SB_DNA_043", Sample) |
                                grepl("PK_SB_DNA_016", Sample) |  
                                grepl("PK_SB_DNA_092", Sample) | 
                                grepl("PK_SB_DNA_030", Sample) | 
                                grepl("PK_SB_DNA_093", Sample) | 
                                grepl("PK_SB_DNA_042", Sample) | 
                                grepl("PK_SB_DNA_063", Sample), "Mn", .$Cluster)) %>% # if not, just use values from Cluster
    mutate(Cluster = ifelse(Cluster == "Sabah", "Mf", .$Cluster))

SNP_Frequency <- annotated_orthologue_genes %>%
  pivot_longer(3:187, names_to = "Sample", values_to = "Allele")  %>%
  mutate(Sample = str_remove(Sample, "_DK.*")) %>%
  #filter(grepl("PK", Sample)) %>%
  left_join(Metadata) %>%
  filter(Allele != -1) %>% # filter out samples with missing calles
  group_by(Cluster, CHROM, POS) %>%
  summarise(SNP_freq = sum(Allele >= 1)/n()*100) %>%
  na.omit()

SNP_Frequency_plot <- SNP_Frequency %>%
  filter(CHROM == "01") %>%
  ggplot(aes(x = POS, y = SNP_freq, colour = Cluster)) +
  geom_point(size = 3) +
  geom_line(aes(group = POS), colour = "grey") +
  scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF")) +
  ylab("SNP Frequency") +
  xlab("Chromosome Postion")

ggsave("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/drug_resistance/SNP_Frequency_plot_01.png", dpi = 600, width = 12, SNP_Frequency_plot)
```

## RSCB PDB 

For variants that are "close calls" in AA position to their orhoglues in Pv/Pf, we searched for the area of the resulting protein structures using RSCB PDB.

## Subset to orthologue genes and build trees

IBS to be caclculated separately for each of the genes:

ordered_PKNH_01_v2  381348  384824  Extract 
ordered_PKNH_05_v2  445338  447218  Extract
ordered_PKNH_08_v2  1408170 1411360 Extract 
ordered_PKNH_10_v2  448974  453368  Extract
ordered_PKNH_13_v2  2302686 2304038 Extract
ordered_PKNH_14_v2  1308259 1310954 Extract
ordered_PKNH_14_v2  2116423 2122473 Extract

Firstly, you need to use plink to create a new plink dataset (cleaned), that can be passed into IBS plotting.
So, you need to repeat the chunk below (or write a loop) for each of the genes described above, by substituing them in the drug_resistant_genes.txt file.

```{bash,eval=F}
plink --bcf PK_consensus_filtered_pass.bcf \
  --keep-allele-order \
  --vcf-idspace-to _ \
  --double-id \
  --allow-extra-chr 0 \
  --make-bed \
  --indiv-sort file sample_order.txt \
  --extract 'range' drug_resistant_genes.txt \
  --allow-no-sex \
  --out Pk

plink --bfile Pk --recode --out Pk
plink --file Pk --maf 0.05 --geno 0.25 --mind 0.25 --make-bed --out cleaned

plink --bfile cleaned \
  --distance square \
  --out drug_resistance/DR_14_2
```

```{bash,eval=F}
cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/drug_resistance
Rscript  /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/drug_resistance/ordination_plots.R
```

```{R,eval=F}
# Load packages
library(tidyverse)
library(ape)
library(ggtree)

# Neighbour-joining tree

## Create distance matrix from PLINK data and build NJT
NJT_ID <- read_table("DR_01.dist.id", col_names=F) %>%
    as.data.frame() %>%
    mutate_all(~str_remove(., "_DKD.*")) 

NJT_matrix <- read_table("DR_01.dist", col_names=NJT_ID$X1) %>%
    as.data.frame() %>%
    add_column(Row_Names = NJT_ID$X1) %>%
    column_to_rownames("Row_Names") %>%
    as.matrix()

NJT_tree <- nj(NJT_matrix)

# Read in initial metadata
NJT_metadata <- read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(Sample = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>% 
        as.data.frame() 

NJT_metadata <- NJT_metadata %>% 
    mutate(Cluster = ifelse(grepl("PK_SB_DNA_011", Sample) | # does the ID match any of these
                                grepl("PK_SB_DNA_091", Sample) | 
                                grepl("PK_SB_DNA_043", Sample) |
                                grepl("PK_SB_DNA_016", Sample) |  
                                grepl("PK_SB_DNA_092", Sample) | 
                                grepl("PK_SB_DNA_030", Sample) | 
                                grepl("PK_SB_DNA_093", Sample) | 
                                grepl("PK_SB_DNA_042", Sample) | 
                                grepl("PK_SB_DNA_063", Sample), "Mn", .$Cluster)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Cluster = ifelse(Cluster == "Sabah", "Mf", .$Cluster)) %>% # if its the remaining Sabah samples, make them Mn, else keep them the same
    mutate_all(~str_remove(., "_DKD.*")) 

# Plot tree
## options(ignore.negative.edge=TRUE)

options(ignore.negative.edge=TRUE)

# 01
# Cluster 
NJT_tree_plot <- ggtree(NJT_tree, layout="daylight", size = 0.5, aes(colour = Cluster)) %<+% NJT_metadata +
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank()) +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF"))
    
ggsave("NJT_unrooted_01.png", dpi = 600, height = 15, width = 15, limitsize = FALSE, NJT_tree_plot)

#Labels
NJT_tree_plot <- ggtree(NJT_tree, layout="daylight", size = 0.5, aes(colour = Cluster)) %<+% NJT_metadata +
    theme(legend.position = "right", 
    legend.title = element_blank(), 
    legend.key = element_blank()) +
    geom_tiplab(size = 2) +
    scale_color_manual(values = c("#440154FF", "#39568CFF", "#1F968BFF"))

ggsave("NJT_labelled_unrooted_01.png", dpi = 600, height = 20, width = 20, limitsize = FALSE, NJT_tree_plot)
```

## Align protein sequences of Pk against Pv orthologues 

This was done with PlasmoDB. Then using data read in above, we can subset to variants of interest the obtain their SNP frequencies and hypothesise reasoning for their presence.

```{R,eval=F}
SNP_Frequency %>% 
  filter(CHROM == "14" & (POS == 1309410 | POS == 2116751 | POS == 2118627 | POS == 2120614 | POS == 2121527 | POS == 2121528 | POS == 2121722 | POS == 2121725)) %>%
  arrange(POS) %>%
  write_tsv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/drug_resistance/SNP_freq_for_variant_orthologues.tsv")
```


## 







# Introgression 



## NJT on introgressed regions

The regions_to_include_NJT.txt file was altered dependent on what window(s) needed to be plotted. A 20K buffer was added to either side of the 10k window



```{bash,eval=F}
while read WINDOW CHROM start end; do
  echo "$CHROM  $start  $end  Include" > Introgression/NJT_data/${WINDOW}.txt
done <Introgression/tree_windows.tsv

cd Introgression/NJT_data/
rm -f WINDOW.txt

ANALYSES="/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses/"

for i in *.txt
  do
    plink --bcf $ANALYSES/PK_consensus_filtered_pass.bcf \
    --keep-allele-order \
    --vcf-idspace-to _ \
    --double-id \
    --allow-extra-chr 0 \
    --make-bed \
    --indiv-sort file $ANALYSES/sample_order.txt \
    --extract 'range' $i \
    --allow-no-sex \
    --out Pk

    plink --bfile Pk --recode --out Pk
    plink --file Pk --maf 0.05 --geno 0.25 --mind 0.25 --make-bed --out cleaned
    plink --bfile cleaned --distance square --out ${i%.txt}.NJT
done






plink --bcf PK_consensus_filtered_pass.bcf \
  --keep-allele-order \
  --vcf-idspace-to _ \
  --double-id \
  --allow-extra-chr 0 \
  --make-bed \
  --indiv-sort file sample_order.txt \
  --extract 'range' regions_to_include_NJT.txt \
  --allow-no-sex \
  --out Pk

plink --bfile Pk --recode --out Pk
plink --file Pk --maf 0.05 --geno 0.25 --mind 0.25 --make-bed --out cleaned
plink --bfile cleaned --distance square --out Pk
```
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/introgression_NJT_plots.R


















R script used above for manually curating the clusters - **script dependent on comparison/clusters**

### Mf, Mn and Peninsular

```{R,eval=F}
library(tidyverse)

read_csv("Pk.csv", col_names = F) %>% 
    mutate(Fst_cluster = ifelse(grepl("PK_SB_DNA_021", X1) | grepl("PK_SB_DNA_088", X1) | grepl("PK_SB_DNA_049", X1) | grepl("PK_SB_DNA_094", X1) | grepl("PK_SB_DNA_046", X1) | # does the ID match any of these
                    grepl("PK_SB_DNA_054", X1) | grepl("PK_SB_DNA_040", X1) | grepl("PK_SB_DNA_005", X1) | grepl("PK_SB_DNA_090", X1) | grepl("PK_SB_DNA_014", X1) |
                    grepl("PK_SB_DNA_036", X1) | grepl("PK_SB_DNA_029", X1) | grepl("PK_SB_DNA_044", X1) | grepl("PK_SB_DNA_072", X1) | grepl("PK_SB_DNA_067", X1) |
                    grepl("PK_SB_DNA_082", X1) | grepl("PK_SB_DNA_070", X1) | grepl("PK_SB_DNA_052", X1) | grepl("PK_SB_DNA_002", X1) | grepl("PK_SB_DNA_009", X1) |
                    grepl("PK_SB_DNA_035", X1) | grepl("PK_SB_DNA_001", X1) | grepl("PK_SB_DNA_031", X1) | grepl("PK_SB_DNA_084", X1) | grepl("PK_SB_DNA_077", X1) |
                    grepl("PK_SB_DNA_074", X1) | grepl("PK_SB_DNA_015", X1) | grepl("PK_SB_DNA_023", X1) | grepl("PK_SB_DNA_013", X1) | grepl("PK_SB_DNA_047", X1) |
                    grepl("PK_SB_DNA_038", X1) | grepl("PK_SB_DNA_081", X1) | grepl("PK_SB_DNA_003", X1) | grepl("PK_SB_DNA_086", X1) | grepl("PK_SB_DNA_089", X1) |
                    grepl("PK_SB_DNA_034", X1) | grepl("PK_SB_DNA_057", X1) | grepl("PK_SB_DNA_045", X1), "Mf", .$X7)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Fst_cluster = ifelse(Fst_cluster == "Sabah", "Mn", .$Fst_cluster)) %>% # if its the remaining Sabah samples, make them Mn, else keep them the same
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)
```

### Newly identified 4th cluster vs everything else

```{R,eval=F}
library(tidyverse)

read_csv("Pk.csv", col_names = F) %>% 
    mutate(Fst_cluster = ifelse(grepl("PK_SB_DNA_019", X1) | grepl("PK_SB_DNA_080", X1) | grepl("PK_SB_DNA_022", X1) | grepl("PK_SB_DNA_041", X1) | # does the ID match any of these
                        grepl("PK_SB_DNA_004", X1) | grepl("PK_SB_DNA_020", X1) | grepl("PK_SB_DNA_064", X1) | grepl("PK_SB_DNA_061", X1) |
                        grepl("SRR3135172", X1) | grepl("SRR2225467", X1) | grepl("SRR2222335", X1), "New_cluster", "Other_clusters")) %>% # if not, just use values from X7 - clusters and Sabah
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)
```

### 4th cluster vs peninsular, 4th cluster vs Mf, & 4th vs Mn 

```{R,eval=F}
Clusters <- read_csv("Pk.csv", col_names = F) %>% 
    mutate(Fst_cluster = ifelse(grepl("PK_SB_DNA_021", X1) | grepl("PK_SB_DNA_088", X1) | grepl("PK_SB_DNA_049", X1) | grepl("PK_SB_DNA_094", X1) | grepl("PK_SB_DNA_046", X1) | # does the ID match any of these
                    grepl("PK_SB_DNA_054", X1) | grepl("PK_SB_DNA_040", X1) | grepl("PK_SB_DNA_005", X1) | grepl("PK_SB_DNA_090", X1) | grepl("PK_SB_DNA_014", X1) |
                    grepl("PK_SB_DNA_036", X1) | grepl("PK_SB_DNA_029", X1) | grepl("PK_SB_DNA_044", X1) | grepl("PK_SB_DNA_072", X1) | grepl("PK_SB_DNA_067", X1) |
                    grepl("PK_SB_DNA_082", X1) | grepl("PK_SB_DNA_070", X1) | grepl("PK_SB_DNA_052", X1) | grepl("PK_SB_DNA_002", X1) | grepl("PK_SB_DNA_009", X1) |
                    grepl("PK_SB_DNA_035", X1) | grepl("PK_SB_DNA_001", X1) | grepl("PK_SB_DNA_031", X1) | grepl("PK_SB_DNA_084", X1) | grepl("PK_SB_DNA_077", X1) |
                    grepl("PK_SB_DNA_074", X1) | grepl("PK_SB_DNA_015", X1) | grepl("PK_SB_DNA_023", X1) | grepl("PK_SB_DNA_013", X1) | grepl("PK_SB_DNA_047", X1) |
                    grepl("PK_SB_DNA_038", X1) | grepl("PK_SB_DNA_081", X1) | grepl("PK_SB_DNA_003", X1) | grepl("PK_SB_DNA_086", X1) | grepl("PK_SB_DNA_089", X1) |
                    grepl("PK_SB_DNA_034", X1) | grepl("PK_SB_DNA_057", X1) | grepl("PK_SB_DNA_045", X1), "Mf", .$X7)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Fst_cluster = ifelse(Fst_cluster == "Sabah", "Mn", .$Fst_cluster)) %>% # if its the remaining Sabah samples, make them Mn, else keep them the same
    mutate(Fst_cluster = ifelse(grepl("PK_SB_DNA_019", X1) | grepl("PK_SB_DNA_080", X1) | grepl("PK_SB_DNA_022", X1) | grepl("PK_SB_DNA_041", X1) | # does the ID match any of these
                        grepl("PK_SB_DNA_004", X1) | grepl("PK_SB_DNA_020", X1) | grepl("PK_SB_DNA_064", X1) | grepl("PK_SB_DNA_061", X1) |
                        grepl("SRR3135172", X1) | grepl("SRR2225467", X1) | grepl("SRR2222335", X1), "New_cluster", .$Fst_cluster))

# 4th cluster vs peninsular
Clusters %>%
    filter(Fst_cluster == "Peninsular" | Fst_cluster == "New_cluster") %>%
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)

# 4th cluster vs Mf
Clusters %>%
    filter(Fst_cluster == "Mf" | Fst_cluster == "New_cluster") %>%
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)

# 4th cluster vs Mn
Clusters %>%
    filter(Fst_cluster == "Mn" | Fst_cluster == "New_cluster") %>%
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)
```