# Using plink to for explotaroy analyses

This markdown describes the population genetics analyses, inlcuding pre-processing steps.

Helpful links:
PLINK: https://www.biostars.org/p/271694/
PLINK: https://plink.readthedocs.io/en/latest/plink_ex/
MOI: https://bahlolab.github.io/moimix/vignettes/introduction.html#estimating-moi-with-f_ws

## Required tools/libraries for Analyses

```{bash}
module load bcftools/1.12
module load java/jdk-8.40 
module load R/4.1.0 
export R_LIBS_USER="/g/data/pq84/R"
export PATH=$PATH:/g/data/pq84/bin/plink2/

FASTA="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/ref_genomes/PKA1H1/fasta"
VCF="/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/04_Variant_calling/filtered/"
GATK="/g/data/pq84/bin/GenomeAnalysisTK-3.8-1-0/GenomeAnalysisTK.jar"

cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses
```

## Filter for variants that pass filters and select chromosomes 1 to 14

```{bash,eval=F}
java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
    -T SelectVariants \
    -R $FASTA/strain_A1_H.1.Icor.fasta \
    -V $VCF/FILTERED_SNPs.vcf  \
    --excludeFiltered \
    --excludeIntervals PKNH_MIT_v2 --excludeIntervals new_API_strain_A1_H.1 \
    -o PK_consensus_filtered_pass.vcf
```

############################################################################################################# WORK IN PROGRESS
# TO Complete
## Calculate multiplicity of infection using moimix

To install MCMCpack on Gadi: https://opus.nci.org.au/display/Help/R 

Steps:
 - Convert VCF to GDS
 - Estimate BAF matrix 
    - estimate B-allele freq for each isolate. 
    - resulting bafmatrix stores estimates and genomic coordinates
 - Estimating MOI with binommix
    - estimated on per-isolate basis using binomial mixture model
    - to fit model - need a matrix of read counts and to set the value of k=2c where c is the number of colnal infections
 - Estimating MOI with Fws
    - alternative way of assessing clonality is to estmate Fws, with Fws < 0.95 indicating clonal infection (low MOI)

**NB.** Clean up VCF.
Add annotation the ID column of your VCF file.
bcftools annotate --threads 10 --set-id +'%CHROM\_%POS\_%REF\_%FIRST_ALT' PK_consensus_filtered_pass.vcf > moimix/PK_consensus_filtered_pass.vcf

bcftools annotate --threads 10 --set-id +'%CHROM\_%POS\_%REF\_%FIRST_ALT' PK_consensus_filtered_pass.vcf | sed 's/Number=A/Number=./g' | bcftools view --max-ac :alt1 - > moimix/PK_consensus_filtered_pass_test.vcf



Run below in R
```{R,eval=F}
# Converting VCF to GDS for estMOI
library(SeqArray)
seqVCF2GDS("moimix/PK_consensus_filtered_pass.vcf", "moimix/PK_consensus_filtered_pass.gds")
isolates <- seqOpen("moimix/PK_consensus_filtered_pass.gds") # read into R
seqSummary(isolates)
sample.id <- seqGetData(isolates, "sample.id")
seqCheck(isolates)
str(seqGetData(isolates, "annotation/info/AC"))

head(seqGetData(isolates, "sample.id"))

# moimix analysis
library(moimix)
coords <- getCoordinates(isolates) # get genomic coordinates of all variants

## estimating BAF matrix 
isolate_baf <- bafMatrix(isolates)

## Plot BAF matrix and export ###########################

## Estimate MOI with binommix
set.seed(2022)
counts <- alleleCounts(isolates)
m1 <- binommix(counts, samplie.id = "ERR2214837", k = 2)
summary(m1)
param.estimates <- getTheta(m1)
param.estimates

## Plot MOI and export ######################

## Estiamting MOI with Fws
fws_all <- getFws(isolates)

## Plot MOI distribution and export ######################

## identify isolates with high MOI
library(tidyverse)
fws_all %>%
    filter() %>%
    write_tsv("estMOI/high_MOI_samples.tsv")


```

############################################################################################################# WORK IN PROGRESS

## Normalise VCF and convert filtered vcf to bcf

 - Ensure that multi-allelic calls are split (1st part)
 - Left-align indels (2nd part)
 - Set the ID field to a unique value: CHROM:POS:REF:ALT (3rd part)
 
 --indiv-sort - specify how samples will be sorted - use to be consistent when adding metadata downstream 

```{bash,eval=F}
bcftools view PK_consensus_filtered_pass.vcf | \
  sed 's/AD,Number=R/AD,Number=./' | \
  bcftools norm -m-any | \
  bcftools norm --check-ref w -f $FASTA/strain_A1_H.1.Icor.fasta | \
  bcftools annotate -Ob -x ID -I +'%CHROM:%POS:%REF:%ALT:%ID' \
  > PK_consensus_filtered_pass.bcf
```

## Convert to PLINK format 

Generic PLINK function to produce generic PLINK files
First create sample_order file

```{bash,eval=F}
bcftools view PK_consensus_filtered_pass.bcf | sed -n '/#CHROM/,$p' | head -n 1 | tr '\t' '\n' |  sed -n '/ERR2214837/,$p' | awk '{print $1 "\t" $1}' > sample_order.txt

plink --bcf PK_consensus_filtered_pass.bcf \
  --keep-allele-order \
  --vcf-idspace-to _ \
  --double-id \
  --allow-extra-chr 0 \
  --make-bed \
  --indiv-sort file sample_order.txt \
  --allow-no-sex \
  --out Pk
```

## Add metadata to PLINK fam

*Had to write an R script to manually add metadata as PLINK was being painful*

```{bash,eval=F}
sed 's/ /,/g' Pk.fam > Pk.csv
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/pheno_to_plink_fam.R
sed 's/,/ /g' Pk.csv > Pk.fam
```

### pheno_to_plink_fam.R

```{R,eval=F}
# Packages
library(tidyverse)
library(janitor)
library(data.table)

# Meatdata
Location_data <- readxl::read_xlsx("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/PK_Sabah_Sample_naming_indexes.xlsx") %>% 
  select(1) %>%
  add_column(Location = "Sabah") %>%
  rename(Sample = sampleid) %>%
  add_column(Cluster = "Sabah") %>% 
  rbind(
    read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_metadata.csv") %>%
      select(1, 3, 5) %>% 
      rename(Location = area) %>%
      mutate(Cluster = str_remove(Group, "-Pk")) %>% 
      select(-Group) 
  ) %>%
  rbind(
    read_csv("/g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/data/metadata/Pk_clusters_peninsular_metadata.csv") %>%
      select(2, 8) %>%
      rename(Sample = ENA_accession_no_ES) %>%
      add_column(Cluster = "Peninsular") 
  ) %>%
  mutate(Location = str_replace(Location, " ", "_"))


Plink_fam <- read_csv("Pk.csv", col_names=FALSE) %>%
  select(-ncol(.)) %>%
  mutate(Sample = str_remove(X2, "_DKD.*")) %>% 
  left_join(Location_data) %>% 
  select(-Sample) 

write_csv(Plink_fam, "Pk.csv", col_names = FALSE)
```

## Remove duplicates

```{bash,eval=F}
cut -f 2 Pk.bim | sort | uniq -d > Pk.dups
```

## Calculate minor allele frequency, genotypic missingness

```{eval=F}
plink --bfile Pk \
  --freq \
  --missing \
  --allow-no-sex \
  --out Pk
```

 - N_MISS =Number of missing SNPs
 - N_GENO =Number of non-obligatory missing genotypes
 - F_MISS =Proportion of missing SNPs

## Plot distribution of MAF and genotypic missingness

```{bash,eval=F}
Rscript /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/Exploratory_analyses/PLINK/maf_dist.R
Rscript /g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/Exploratory_analyses/PLINK/miss_dist.R
```

### maf_dist.R

```{R,eval=F}
# Load packages
library(tidyverse)

# Read in data
pk_maf <- read_table("Pk.frq", col_names=T) %>%
    mutate(CHR = str_remove(SNP, ":.*"))

# Plot MAF by SNP
pk_maf_SNP <- pk_maf  %>%
    ggplot(aes(x = SNP, y = MAF, fill = CHR)) +
    geom_col() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    geom_hline(yintercept = 0.01, colour = "#1F968BFF") +
    theme(axis.text.x = element_text(angle = 45), legend.position = "none") +
    scale_fill_viridis_d() 

ggsave("Pk_maf_plot.png", dpi = 600, pk_maf_SNP)

# Density of MAF by SNP
pk_maf_density <- pk_maf  %>%
    ggplot(aes(x = MAF)) +
    geom_density(alpha=.3) +
    geom_vline(xintercept = 0.01, colour = "#1F968BFF") 

ggsave("Pk_maf_density.png", dpi = 600, pk_maf_density)

# Plot MAF by chr
pk_maf_chr <- pk_maf %>% 
    mutate(SNP = str_remove(SNP, ":.*"),
        SNP = str_remove(SNP, "ordered_"),
        SNP = str_remove(SNP, "new_"),
        SNP = str_remove(SNP, "_v2"),
        SNP = str_remove(SNP, "A1.*")) %>%
    group_by(SNP) %>%
    summarise(MAF = mean(MAF)) %>%
    ggplot(aes(x = SNP, y = MAF)) +
    geom_col() +
    theme(axis.text.x = element_text(angle = 45), legend.position = "none") +
    scale_fill_viridis_d() 

ggsave("Pk_maf_plot_chr.png", dpi = 600, pk_maf_chr)

# Summary stats

pk_maf %>% 
    mutate(SNP = str_remove(SNP, ":.*")) %>%
    group_by(SNP) %>%
    summarise(MAF_mean = mean(MAF),
            MAF_SD = sd(MAF),
            MAF_SE = (sd(MAF))/sqrt(n())) %>%
rbind(
    pk_maf %>%
        summarise(MAF_mean = mean(MAF),
                MAF_SD = sd(MAF),
                MAF_SE = (sd(MAF))/sqrt(n())) %>%
        add_column(SNP = "Total")
    ) %>%
    write_csv(col_names = T, "pk_maf_summary.csv")
```

### miss_dist.R

```{R,eval=F}
# Load packages
library(tidyverse)
library(data.table)

# Sample-based missing data
## Read in data
sample_mis <- read_table("Pk.imiss", col_names=T) 

# Density of miss by sample
sample_miss_density <- sample_mis %>% 
    rename(Sample = IID,
        Geno_Miss = F_MISS) %>%
    mutate(Sample = str_remove(Sample, "DKD.*"),
            Sample = str_remove(Sample, "PK_SB_DNA_")) %>%
    select(2:ncol(.)) %>%
    ggplot(aes(x = Geno_Miss)) +
    geom_density(alpha=.3) +
    geom_vline(xintercept = 0.25, colour = "#1F968BFF") 

ggsave("sample_miss_density_plot.png", width = 12, dpi = 600, sample_miss_density)

# Variant-based missing data
## Read in data
var_miss <- read_table("Pk.lmiss", col_names=T) %>%
    mutate(CHR = str_remove(SNP, ":.*")) %>%
    rename(Geno_Miss = F_MISS,
            Chr = CHR)

## Plot dist by variant and coloured by chromosome
var_miss_plot <- var_miss %>% 
    ggplot(aes(x = SNP, y = Geno_Miss, fill =  Chr)) +
    geom_col() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    scale_fill_viridis_d(option = "D") 

ggsave("var_miss_plot.png", width = 12, dpi = 600, var_miss_plot)

# Density of miss by variant
var_miss_density <- var_miss %>% 
    ggplot(aes(x = Geno_Miss)) +
    geom_density(alpha=.3) +
    geom_vline(xintercept = 0.25, colour = "#1F968BFF") 

ggsave("variant_miss_density_plot.png", width = 12, dpi = 600, var_miss_density)

# Summary stats

miss_summary <- var_miss %>%
    summarise(Miss_mean = mean(Geno_Miss),
                Miss_SD = sd(Geno_Miss),
                Miss_SE = (sd(Geno_Miss))/sqrt(n()))

write_csv(miss_summary, "pk_miss_summary.csv")
```

## Apply filters based on MAF, genotypic missingess and Fsw

 - first recode creates the intial .map and .ped files
 - recode generates new .ped and .map files
 - --make-bed generates .bed, .bim and .fam files
 - -mind filters per sample (missingness)
 - -geno filters per variant (missingness

**FILTERED:**
 - exclude variants with > 25% missing alleles
 - exclude variants with minor allele frequencies > 5%

MAF (=Minor Allele Frequency) = P(geno=1, Aa)x0.5 + P(geno=2, aa)

```{bash,eval=F}
plink --bfile Pk --recode --out Pk
plink --file Pk --maf 0.05 --geno 0.25 --make-bed --out cleaned
```

Result:

--maf 0.01 --geno 0.25

102081 variants removed due to missing genotype data
731000 variants removed due to minor allele threshold(s)
842207 variants and 202 people pass filters and QC

--maf 0.05 --geno 0.25

102081 variants removed due to missing genotype data (--geno).
1244443 variants removed due to minor allele threshold(s)
(--maf/--max-maf/--mac/--max-mac).
328764 variants and 202 people pass filters and QC.


## Use PLINK to create PCA and MDS data, and a distance matrix (identity-by-state)

The first few lines is based on non-filtered data, wherease the lines below these are for the filtered data generated above.

```{bash,eval=F}
plink --bfile cleaned --cluster --mds-plot 10 --out Pk
plink --bfile cleaned --pca --out Pk
plink --bfile cleaned --distance square --out Pk
Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/ordination_plots.R
```

### ordination_plots.R

```{R,eval=F}
# Load packages
library(tidyverse)
library(ape)
library(ggtree)

# PCA

## Read in data
pca <- read_table("Pk.eigenvec", col_names=F) %>%
    select(-1) %>%
    mutate(X2 = str_remove(X2, "_DK.*")) %>%
    rename("sampleid" = "X2") %>%
    rename_at(vars(starts_with("X")), ~str_replace(., "X.*", paste0("PC", seq_along(.))))

eigenval <- scan("Pk.eigenval")

## Add metadata 
pca <- pca %>%
    left_join(read_csv("Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(sampleid = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>%
        mutate(sampleid = str_remove(sampleid, "_DK.*")))


## Percentage variance explained by each PC
pve_plot <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100) %>%
    ggplot(aes(PC, pve)) + 
    geom_bar(stat = "identity") + 
    ylab("Percentage variance explained") + 
    theme_light()

ggsave("Percentage_variance_explained.png", dpi=600, pve_plot)

## Plot PCA
pve <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100)

pca_plot <- ggplot(pca, aes(PC1, PC2, colour = Cluster)) + 
    geom_point(size = 3) +
    scale_color_viridis_d() +
    coord_equal() + 
    theme_light() + 
    xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + 
    ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))

ggsave("PCA.png", dpi = 600, pca_plot)


# MDS

## Read in data
mds <- read_table("Pk.mds", col_names=T) %>%
    select(-c("FID", "X14")) %>%
    mutate(IID = str_remove(IID, "_DK.*")) %>%
    rename("sampleid" = "IID") %>%
    rename_at(vars(starts_with("C")), ~str_replace(., "C", "MDS"))

## Add metadata 
mds <- mds %>%
    left_join(read_csv("Pk.csv", col_names = FALSE) %>% 
        select(1, 6, 7) %>%
        rename(sampleid = X1) %>%
        rename(Location = X6) %>%
        rename(Cluster = X7) %>%
        mutate(sampleid = str_remove(sampleid, "_DK.*")))

## Plot MDS
mds_plot <- ggplot(mds, aes(MDS1, MDS2, colour = Cluster)) + 
    geom_point(size = 3) +
    scale_color_viridis_d() +
    coord_equal() + 
    theme_light()

ggsave("MDS.png", dpi=600, mds_plot)


# Neighbour-joining tree

## Create distance matrix from PLINK data and build NJT
NJT_ID <- read_table("Pk.dist.id", col_names=F) %>%
    as.data.frame() %>%
    mutate_all(~str_remove(., "_DKD.*")) 

NJT_matrix <- read_table("Pk.dist", col_names=NJT_ID$X1) %>%
    as.data.frame() %>%
    add_column(Row_Names = NJT_ID$X1) %>%
    column_to_rownames("Row_Names") %>%
    as.matrix()

NJT_tree <- nj(NJT_matrix)

## Plot tree
# options(ignore.negative.edge=TRUE)

# Read in initial metadata
NJT_metadata <- read_csv("Pk.csv", col_names = FALSE) %>%
    select(1, 6, 7) %>%
    rename(Sample = X1) %>%
    rename(Location = X6) %>%
    rename(Cluster = X7) %>%
    mutate(Sample = str_remove(Sample, "_DKD.*")) %>%
    mutate_if(is.character, as.factor) %>%
    mutate(Region = ifelse(Location == "Sabah" | Location == "Betong" | Location == "Kapit" | Location == "Sarikei", "Borneo", "Peninsular")) 

# Update metadata with more detailed location - districts - where we split up the Sabah samples (which represent OUR data), and also the clinic samples
NJT_metadata <- NJT_metadata %>%
    left_join(
    readxl::read_xlsx("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/PK_Sabah_Sample_naming_indexes.xlsx") %>% # index data
        mutate(subjectid = as.character(subjectid)) %>%
        mutate(subjectid = ifelse(str_length(subjectid) == 1, paste0("00", subjectid), # if subjectid length = 1 paste 00
                     ifelse(str_length(subjectid) == 2, paste0("0", subjectid), # if not, then if subjectid length = 2 paste 0
                     .$subjectid))) %>% # if not, keep value the same
    mutate(studycode = paste0(group, subjectid)) %>% # create studycode column that aligns with metadata
    left_join( # left join the epi data to the index data so that we only get epi data for the samples that we have
    read_tsv("/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/metadata/Epi_data_ZB_clean.tsv")) %>%
    select(sampleid, district) %>%
    rename(Sample = sampleid) %>%
    rename(District = district) %>%
    mutate(District = str_replace(District, " ", "_"))
    ) %>% 
    mutate_if(is.factor, as.character) %>%
    mutate(District = ifelse(is.na(District), .$Location, .$District)) %>% 
    mutate(District = ifelse(Sample == "SRR2221468", "Hackeri", 
        ifelse(Sample == "SRR2222335", "H(AW)", 
        ifelse(Sample == "SRR2225467", "Malayan", 
        ifelse(Sample == "SRR2225571", "MR4-H", 
        ifelse(Sample == "SRR2225573", "Philippine",
        ifelse(Sample == "SRR3135172", "YH1", .$District))))))) %>%
    mutate(District = str_replace(District, "Kinaalu", "Kinabalu"))

options(ignore.negative.edge=TRUE)

# Location - Our Sabah samples vs OTHER
NJT_tree_plot <- ggtree(NJT_tree, layout = "circular", size = 0.5, aes(colour = Location)) %<+% NJT_metadata +
    geom_tippoint(aes(colour = Location, shape = Region)) + 
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank()) 
    
ggsave("NJT_tree_location.png", dpi = 600, height = 8, width = 16, NJT_tree_plot)

# Cluster - Our Sabah samples vs previously defined clusters
NJT_tree_plot <- ggtree(NJT_tree, layout="circular", size = 0.5, aes(colour = Cluster)) %<+% NJT_metadata +
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank()) 
    
ggsave("NJT_tree_cluster_rooted.png", dpi = 600, height = 8, width = 16, NJT_tree_plot)

NJT_tree_plot <- ggtree(NJT_tree, layout="daylight", size = 0.5, aes(colour = Cluster)) %<+% NJT_metadata +
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank())
    
ggsave("NJT_tree_cluster_unrooted.png", dpi = 600, height = 15, width = 15, limitsize = FALSE, NJT_tree_plot)

# District - MORE RESOLUTION - split our Sabah samples by district, and the clinic samples
NJT_tree_plot <- ggtree(NJT_tree, layout="daylight", size = 0.5, aes(colour = District)) %<+% NJT_metadata +
    geom_tippoint(aes(colour = District, shape = Cluster)) + 
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank())
    
ggsave("NJT_tree_district_unrooted.png", dpi = 600, height = 15, width = 15, limitsize = FALSE, NJT_tree_plot)

NJT_tree_plot <- ggtree(NJT_tree, layout="circular", size = 0.5, aes(colour = District)) %<+% NJT_metadata +
    geom_tippoint(aes(colour = District, shape = Cluster)) + 
    theme(legend.position = "right", 
        legend.title = element_blank(), 
        legend.key = element_blank()) 
    
ggsave("NJT_tree_district_rooted.png", dpi = 600, height = 8, width = 16, NJT_tree_plot)

#Labels
NJT_tree_plot <- ggtree(NJT_tree, layout="circular", size = 0.5, aes(colour = Cluster)) %<+% NJT_metadata +
   theme(legend.position = "right", 
   legend.title = element_blank(), 
   legend.key = element_blank()) +
   geom_tiplab(size = 2)
    
ggsave("NJT_tree_labelled_rooted.png", dpi = 600, height = 20, width = 20, limitsize = FALSE, NJT_tree_plot)
```


## Fst statistic

First create a file that contains the clusters you want to used for the Fst. You can ammend the Pk.fam file (or metadata), but the file needs to be structured in PLINKs weird FID-IID-Cluster format.

I aslo tighten up the filter to get rid of any variant that has > 10% genotypic missingness

```{bash,eval=F}
plink --file Pk --maf 0.05 --geno 0.10 --make-bed --out cleaned

Rscript /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/scripts/06_Analysis/sample_clusters.R
sed 's/,/ /g' sample_clusters.csv > sample_clusters.txt
rm -f sample_clusters.csv

plink --bfile cleaned \
  --within sample_clusters.txt \
  --fst \
  --allow-no-sex \
  --out Pk
```

R script used above for manually curating the clusters - **script dependent on comparison/clusters**

### Mf, Mn and Peninsular

```{R,eval=F}
library(tidyverse)

read_csv("Pk.csv", col_names = F) %>% 
    mutate(Fst_cluster = ifelse(grepl("PK_SB_DNA_021", X1) | grepl("PK_SB_DNA_088", X1) | grepl("PK_SB_DNA_049", X1) | grepl("PK_SB_DNA_094", X1) | grepl("PK_SB_DNA_046", X1) | # does the ID match any of these
                    grepl("PK_SB_DNA_054", X1) | grepl("PK_SB_DNA_040", X1) | grepl("PK_SB_DNA_005", X1) | grepl("PK_SB_DNA_090", X1) | grepl("PK_SB_DNA_014", X1) |
                    grepl("PK_SB_DNA_036", X1) | grepl("PK_SB_DNA_029", X1) | grepl("PK_SB_DNA_044", X1) | grepl("PK_SB_DNA_072", X1) | grepl("PK_SB_DNA_067", X1) |
                    grepl("PK_SB_DNA_082", X1) | grepl("PK_SB_DNA_070", X1) | grepl("PK_SB_DNA_052", X1) | grepl("PK_SB_DNA_002", X1) | grepl("PK_SB_DNA_009", X1) |
                    grepl("PK_SB_DNA_035", X1) | grepl("PK_SB_DNA_001", X1) | grepl("PK_SB_DNA_031", X1) | grepl("PK_SB_DNA_084", X1) | grepl("PK_SB_DNA_077", X1) |
                    grepl("PK_SB_DNA_074", X1) | grepl("PK_SB_DNA_015", X1) | grepl("PK_SB_DNA_023", X1) | grepl("PK_SB_DNA_013", X1) | grepl("PK_SB_DNA_047", X1) |
                    grepl("PK_SB_DNA_038", X1) | grepl("PK_SB_DNA_081", X1) | grepl("PK_SB_DNA_003", X1) | grepl("PK_SB_DNA_086", X1) | grepl("PK_SB_DNA_089", X1) |
                    grepl("PK_SB_DNA_034", X1) | grepl("PK_SB_DNA_057", X1) | grepl("PK_SB_DNA_045", X1), "Mf", .$X7)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Fst_cluster = ifelse(Fst_cluster == "Sabah", "Mn", .$Fst_cluster)) %>% # if its the remaining Sabah samples, make them Mn, else keep them the same
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)
```

### Newly identified 4th cluster vs everything else

```{R,eval=F}
library(tidyverse)

read_csv("Pk.csv", col_names = F) %>% 
    mutate(Fst_cluster = ifelse(grepl("PK_SB_DNA_019", X1) | grepl("PK_SB_DNA_080", X1) | grepl("PK_SB_DNA_022", X1) | grepl("PK_SB_DNA_041", X1) | # does the ID match any of these
                        grepl("PK_SB_DNA_004", X1) | grepl("PK_SB_DNA_020", X1) | grepl("PK_SB_DNA_064", X1) | grepl("PK_SB_DNA_061", X1) |
                        grepl("SRR3135172", X1) | grepl("SRR2225467", X1) | grepl("SRR2222335", X1), "New_cluster", "Other_clusters")) %>% # if not, just use values from X7 - clusters and Sabah
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)
```

### 4th cluster vs peninsular, 4th cluster vs Mf, & 4th vs Mn 

```{R,eval=F}
Clusters <- read_csv("Pk.csv", col_names = F) %>% 
    mutate(Fst_cluster = ifelse(grepl("PK_SB_DNA_021", X1) | grepl("PK_SB_DNA_088", X1) | grepl("PK_SB_DNA_049", X1) | grepl("PK_SB_DNA_094", X1) | grepl("PK_SB_DNA_046", X1) | # does the ID match any of these
                    grepl("PK_SB_DNA_054", X1) | grepl("PK_SB_DNA_040", X1) | grepl("PK_SB_DNA_005", X1) | grepl("PK_SB_DNA_090", X1) | grepl("PK_SB_DNA_014", X1) |
                    grepl("PK_SB_DNA_036", X1) | grepl("PK_SB_DNA_029", X1) | grepl("PK_SB_DNA_044", X1) | grepl("PK_SB_DNA_072", X1) | grepl("PK_SB_DNA_067", X1) |
                    grepl("PK_SB_DNA_082", X1) | grepl("PK_SB_DNA_070", X1) | grepl("PK_SB_DNA_052", X1) | grepl("PK_SB_DNA_002", X1) | grepl("PK_SB_DNA_009", X1) |
                    grepl("PK_SB_DNA_035", X1) | grepl("PK_SB_DNA_001", X1) | grepl("PK_SB_DNA_031", X1) | grepl("PK_SB_DNA_084", X1) | grepl("PK_SB_DNA_077", X1) |
                    grepl("PK_SB_DNA_074", X1) | grepl("PK_SB_DNA_015", X1) | grepl("PK_SB_DNA_023", X1) | grepl("PK_SB_DNA_013", X1) | grepl("PK_SB_DNA_047", X1) |
                    grepl("PK_SB_DNA_038", X1) | grepl("PK_SB_DNA_081", X1) | grepl("PK_SB_DNA_003", X1) | grepl("PK_SB_DNA_086", X1) | grepl("PK_SB_DNA_089", X1) |
                    grepl("PK_SB_DNA_034", X1) | grepl("PK_SB_DNA_057", X1) | grepl("PK_SB_DNA_045", X1), "Mf", .$X7)) %>% # if not, just use values from X7 - clusters and Sabah
    mutate(Fst_cluster = ifelse(Fst_cluster == "Sabah", "Mn", .$Fst_cluster)) %>% # if its the remaining Sabah samples, make them Mn, else keep them the same
    mutate(Fst_cluster = ifelse(grepl("PK_SB_DNA_019", X1) | grepl("PK_SB_DNA_080", X1) | grepl("PK_SB_DNA_022", X1) | grepl("PK_SB_DNA_041", X1) | # does the ID match any of these
                        grepl("PK_SB_DNA_004", X1) | grepl("PK_SB_DNA_020", X1) | grepl("PK_SB_DNA_064", X1) | grepl("PK_SB_DNA_061", X1) |
                        grepl("SRR3135172", X1) | grepl("SRR2225467", X1) | grepl("SRR2222335", X1), "New_cluster", .$Fst_cluster))

# 4th cluster vs peninsular
Clusters %>%
    filter(Fst_cluster == "Peninsular" | Fst_cluster == "New_cluster") %>%
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)

# 4th cluster vs Mf
Clusters %>%
    filter(Fst_cluster == "Mf" | Fst_cluster == "New_cluster") %>%
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)

# 4th cluster vs Mn
Clusters %>%
    filter(Fst_cluster == "Mn" | Fst_cluster == "New_cluster") %>%
    select(1, 2, 8) %>%
    write_csv("sample_clusters.csv", col_names = F)

```

### Plot Fst 

```{R,eval=F}
library(tidyverse)

Fst_matrix <- read_table("Pk.fst", col_names=T) %>%
    mutate(CHR = str_remove(SNP, "ordered_PKNH_")) %>%
    mutate(CHR = str_remove(CHR, "_v2.*"))

# Sliding window
window_size <- 10000

Fst_plot_window <- Fst_matrix %>%
    filter(FST != "nan") %>%
    mutate(FST = as.numeric(FST)) %>%
    mutate(Window = (floor(POS/window_size) * window_size)+ (window_size/2)) %>%
    group_by(CHR, Window) %>%
    summarise(Window_Fst = mean(FST), count = n()) %>% # caculate the mean Fst for each window in each chr
    filter(count > 10) %>% # remove windows with low counts
    ggplot(aes(x = Window, y = Window_Fst, colour = CHR)) +
    geom_point() +
    facet_wrap(~CHR) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
    ylab("Fst") +
    xlab("Windows (10000)") +
    scale_color_discrete("Chr") +
    geom_hline(yintercept = 0.4, colour = "#1F968BFF")

ggsave("Fst_sliding_window_plot.png", dpi = 600, width = 14, Fst_plot_window)


# Per SNP
Fst_plot <- Fst_matrix %>% 
    filter(CHR == "14") %>%
    ggplot(aes(x = POS, y = FST, colour = CHR)) +
    geom_point()  +
    geom_hline(yintercept = 0.4, colour = "#1F968BFF")
    #theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) 

ggsave("Fst_SNP_plot.png", dpi = 600, width = 14, Fst_plot)
```













## Linkage-disequilibrium

If you want an LD report: https://plink.readthedocs.io/en/latest/plink_ex/
*Note* Requires a lot of resources (for example, when applying to my clean dataset of ~45K variants PLINK tried to use 47 threads and 128GB of ram)
plink --bfile cleaned --allow-no-sex --r2 dprime inter-chr with-freqs --ld-window-r2 0.005

