#!/bin/bash
#PBS -P pq84
#PBS -q normalbw
#PBS -N Annotate_variants
#PBS -j oe
#PBS -m ae
#PBS -l walltime=12:00:00,mem=40GB,ncpus=5
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and load modules"
export PATH=$PATH:/g/data/pq84/bin/ensembl-vep

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
GFF="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/ref_genomes/PKA1H1/gff/updated_version"
FASTA="/g/data/pq84/malaria/Parasite_and_human_genetic_risk_factors_for_Pk_malaria/data/ref_genomes/PKA1H1/fasta/strain_A1_H.1.Icor.fasta"

cd /g/data/pq84/malaria/Pk_Malaysian_Population_Genetics/outputs/05_Analyses

echo "---------------------------------------------------------------------------------------------------------------------"
echo "Annotate SNPs"
vep -i PK_consensus_filtered_pass.vcf --gff $GFF/PlasmoDB-55_PknowlesiA1H1.gff.gz --fasta $FASTA -o PK_consensus_filtered_annotation.txt --force_overwrite --everything

echo "---------------------------------------"
echo "FINISHED"